// igniteui-webcomponents/components/common/mixins/forms/associated.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { property } from 'lit/decorators.js';
function BaseFormAssociated(base) {
    class BaseFormAssociatedElement extends base {
        get __validators() {
            return [];
        }
        set disabled(value) {
            this._disabled = value;
            this.toggleAttribute('disabled', Boolean(this._disabled));
        }
        get disabled() {
            return this._disabled;
        }
        set invalid(value) {
            this._invalid = value;
            this.toggleAttribute('invalid', Boolean(this._invalid));
        }
        get invalid() {
            return this._invalid;
        }
        get form() {
            return this.__internals.form;
        }
        get validity() {
            return this.__internals.validity;
        }
        get validationMessage() {
            return this.__internals.validationMessage;
        }
        get willValidate() {
            return this.__internals.willValidate;
        }
        constructor(...args) {
            super(args);
            this._disabled = false;
            this._invalid = false;
            this._dirty = false;
            this.__internals = this.attachInternals();
            this.addEventListener('invalid', this._handleInvalid);
        }
        connectedCallback() {
            super.connectedCallback();
            this._dirty = false;
            this._updateValidity();
            if (!this.hasUpdated) {
                this._setInitialDefaultValue();
            }
        }
        _handleInvalid(event) {
            event.preventDefault();
            this.invalid = true;
        }
        _setInvalidState() {
            if (this.hasUpdated || this._dirty) {
                this.invalid = !this.checkValidity();
            }
        }
        _setInitialDefaultValue() { }
        _setDefaultValue(_prev, _current) { }
        _restoreDefaultValue() { }
        _validate(message) {
            this._updateValidity(message);
            this._setInvalidState();
        }
        _updateValidity(message) {
            const validity = {};
            let validationMessage = '';
            for (const validator of this.__validators) {
                const isValid = validator.isValid(this);
                validity[validator.key] = !isValid;
                if (!isValid) {
                    validationMessage =
                        typeof validator.message === 'function'
                            ? validator.message(this)
                            : validator.message;
                }
            }
            if (message) {
                validity.customError = true;
                validationMessage = message;
            }
            this.__internals.setValidity(validity, validationMessage);
        }
        _setFormValue(value, state) {
            this.__internals.setFormValue(value, state);
        }
        formAssociatedCallback(_form) { }
        formDisabledCallback(state) {
            this._disabled = state;
            this.requestUpdate();
        }
        formResetCallback() {
            this._restoreDefaultValue();
            this._dirty = false;
            this.performUpdate();
            this.invalid = false;
        }
        formStateRestoreCallback(_state, _mode) { }
        reportValidity() {
            return this.__internals.reportValidity();
        }
        checkValidity() {
            return this.__internals.checkValidity();
        }
        setCustomValidity(message) {
            this._updateValidity(message);
            this.invalid = !this.checkValidity();
        }
    }
    BaseFormAssociatedElement.formAssociated = true;
    __decorate([
        property({ reflect: true })
    ], BaseFormAssociatedElement.prototype, "name", void 0);
    __decorate([
        property({ type: Boolean, reflect: true })
    ], BaseFormAssociatedElement.prototype, "disabled", null);
    __decorate([
        property({ type: Boolean, reflect: true })
    ], BaseFormAssociatedElement.prototype, "invalid", null);
    return BaseFormAssociatedElement;
}
export function FormAssociatedMixin(base) {
    class FormAssociatedElement extends BaseFormAssociated(base) {
        attributeChangedCallback(name, prev, current) {
            super.attributeChangedCallback(name, prev, current);
            if (name === 'value') {
                this._setDefaultValue(prev, current);
            }
        }
        _setDefaultValue(_, current) {
            this._defaultValue = current;
        }
        _restoreDefaultValue() {
            if ('value' in this) {
                this.value = this._defaultValue;
            }
        }
        _setInitialDefaultValue() {
            if ('_value' in this) {
                this._defaultValue = this._value;
            }
        }
    }
    return FormAssociatedElement;
}
export function FormAssociatedCheckboxMixin(base) {
    class FormAssociatedCheckboxElement extends BaseFormAssociated(base) {
        constructor() {
            super(...arguments);
            this._defaultChecked = false;
        }
        attributeChangedCallback(name, prev, current) {
            super.attributeChangedCallback(name, prev, current);
            if (name === 'checked') {
                this._setDefaultValue(prev, current);
            }
        }
        _setDefaultValue(_, current) {
            this._defaultChecked = current !== null;
        }
        _restoreDefaultValue() {
            if ('checked' in this) {
                this.checked = this._defaultChecked;
            }
        }
        _setInitialDefaultValue() {
            if ('checked' in this) {
                this._defaultChecked = this.checked;
            }
        }
    }
    return FormAssociatedCheckboxElement;
}
