// igniteui-react/esm2015/lib/Accordion.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { BaseElement } from "./BaseElement";
import { WCNativeHelper } from "./WCNativeHelper";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let Accordion = /*@__PURE__*/ (() => {
    class Accordion extends BaseElement {
        constructor() {
            super(...arguments);
            this.a = new WCNativeHelper();
            this.d = null;
        }
        get b() {
            return this.a;
        }
        get nativeElement() {
            return this.d;
        }
        set nativeElement(a) {
            this.d = a;
            this.a.o = this.d;
        }
        setNativeElement(a) {
            this.nativeElement = a;
        }
        get c() {
            let ret_ = this.b.n("singleExpand");
            return ret_;
        }
        set c(a) {
            let value_ = a;
            this.b.w("singleExpand", value_);
        }
        f() {
            this.nativeElement.hideAll();
        }
        h() {
            this.nativeElement.showAll();
        }
    }
    Accordion.$t = /*@__PURE__*/ markType(Accordion, 'Accordion', BaseElement.$);
    return Accordion;
})();
