// igniteui-react/esm2015/lib/ActiveStepChangingArgsDetail.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { WCNativeHelper } from "./WCNativeHelper";
/**
 * @hidden
 */
export let ActiveStepChangingArgsDetail = /*@__PURE__*/ (() => {
    class ActiveStepChangingArgsDetail extends Base {
        constructor() {
            super(...arguments);
            this.a = new WCNativeHelper();
            this.e = null;
        }
        get d() {
            let ret_ = this.b.n("oldIndex");
            return ret_;
        }
        set d(a) {
            let value_ = a;
            this.b.w("oldIndex", value_);
        }
        get c() {
            let ret_ = this.b.n("newIndex");
            return ret_;
        }
        set c(a) {
            let value_ = a;
            this.b.w("newIndex", value_);
        }
        get b() {
            return this.a;
        }
        get nativeElement() {
            return this.e;
        }
        set nativeElement(a) {
            this.e = a;
            this.a.o = this.e;
        }
    }
    ActiveStepChangingArgsDetail.$t = /*@__PURE__*/ markType(ActiveStepChangingArgsDetail, 'ActiveStepChangingArgsDetail');
    return ActiveStepChangingArgsDetail;
})();
