// igniteui-react/esm2015/lib/ActivePaneEventArgsDetail.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { WCNativeHelper } from "./WCNativeHelper";
import { ContentPane } from "./ContentPane";
/**
 * @hidden
 */
export let ActivePaneEventArgsDetail = /*@__PURE__*/ (() => {
    class ActivePaneEventArgsDetail extends Base {
        constructor() {
            super(...arguments);
            this.c = new WCNativeHelper();
            this.e = null;
        }
        get d() {
            return this.c;
        }
        get nativeElement() {
            return this.e;
        }
        set nativeElement(a) {
            this.e = a;
            this.c.o = this.e;
        }
        setNativeElement(a) {
            this.nativeElement = a;
        }
        get a() {
            let ret_ = this.d.m("newPane", (a) => new ContentPane());
            return ret_;
        }
        set a(a) {
            let value_ = a;
            this.d.w("newPane", value_);
        }
        get b() {
            let ret_ = this.d.m("oldPane", (a) => new ContentPane());
            return ret_;
        }
        set b(a) {
            let value_ = a;
            this.d.w("oldPane", value_);
        }
    }
    ActivePaneEventArgsDetail.$t = /*@__PURE__*/ markType(ActivePaneEventArgsDetail, 'ActivePaneEventArgsDetail');
    return ActivePaneEventArgsDetail;
})();
