// igniteui-react/esm2015/lib/ActiveStepChangedArgsDetail.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { WCNativeHelper } from "./WCNativeHelper";
/**
 * @hidden
 */
export let ActiveStepChangedArgsDetail = /*@__PURE__*/ (() => {
    class ActiveStepChangedArgsDetail extends Base {
        constructor() {
            super(...arguments);
            this.a = new WCNativeHelper();
            this.d = null;
        }
        get c() {
            let ret_ = this.b.n("index");
            return ret_;
        }
        set c(a) {
            let value_ = a;
            this.b.w("index", value_);
        }
        get b() {
            return this.a;
        }
        get nativeElement() {
            return this.d;
        }
        set nativeElement(a) {
            this.d = a;
            this.a.o = this.d;
        }
    }
    ActiveStepChangedArgsDetail.$t = /*@__PURE__*/ markType(ActiveStepChangedArgsDetail, 'ActiveStepChangedArgsDetail');
    return ActiveStepChangedArgsDetail;
})();
