// igniteui-react/esm2015/lib/Banner.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { BaseElement } from "./BaseElement";
import { WCNativeHelper } from "./WCNativeHelper";
import { markType } from "igniteui-react-core";
import { VoidEventArgs } from "./VoidEventArgs";
/**
 * @hidden
 */
export let Banner = /*@__PURE__*/ (() => {
    class Banner extends BaseElement {
        constructor() {
            super(...arguments);
            this.c = new WCNativeHelper();
            this.i = null;
            this.closing = null;
            this.closed = null;
        }
        get d() {
            return this.c;
        }
        get nativeElement() {
            return this.i;
        }
        set nativeElement(a) {
            this.i = a;
            this.c.o = this.i;
        }
        setNativeElement(a) {
            this.nativeElement = a;
        }
        get f() {
            let ret_ = this.d.n("open");
            return ret_;
        }
        set f(a) {
            let value_ = a;
            this.d.w("open", value_);
        }
        closingChanged() {
            if (this.closing != null) {
                this.d.s("closing", (a) => this.closing(this, VoidEventArgs.b(a)));
            }
            else {
                this.d.v("closing");
            }
        }
        closedChanged() {
            if (this.closed != null) {
                this.d.s("closed", (a) => this.closed(this, VoidEventArgs.b(a)));
            }
            else {
                this.d.v("closed");
            }
        }
        g() {
            return (this.nativeElement.show());
        }
        e() {
            return (this.nativeElement.hide());
        }
        h() {
            return (this.nativeElement.toggle());
        }
    }
    Banner.$t = /*@__PURE__*/ markType(Banner, 'Banner', BaseElement.$);
    return Banner;
})();
