// igniteui-react/esm2015/lib/BaseOptionLike.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { BaseElement } from "./BaseElement";
import { WCNativeHelper } from "./WCNativeHelper";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let BaseOptionLike = /*@__PURE__*/ (() => {
    class BaseOptionLike extends BaseElement {
        constructor() {
            super(...arguments);
            this.a = new WCNativeHelper();
            this.f = null;
        }
        get b() {
            return this.a;
        }
        get nativeElement() {
            return this.f;
        }
        set nativeElement(a) {
            this.f = a;
            this.a.o = this.f;
        }
        setNativeElement(a) {
            this.nativeElement = a;
        }
        get c() {
            let ret_ = this.b.n("active");
            return ret_;
        }
        set c(a) {
            let value_ = a;
            this.b.w("active", value_);
        }
        get d() {
            let ret_ = this.b.n("disabled");
            return ret_;
        }
        set d(a) {
            let value_ = a;
            this.b.w("disabled", value_);
        }
        get e() {
            let ret_ = this.b.n("selected");
            return ret_;
        }
        set e(a) {
            let value_ = a;
            this.b.w("selected", value_);
        }
        get h() {
            let ret_ = this.b.n("value");
            return ret_;
        }
        set h(a) {
            let value_ = a;
            this.b.w("value", value_);
        }
        i() {
            this.nativeElement.connectedCallback();
        }
    }
    BaseOptionLike.$t = /*@__PURE__*/ markType(BaseOptionLike, 'BaseOptionLike', BaseElement.$);
    return BaseOptionLike;
})();
