// igniteui-react/esm2015/lib/BaseComboBoxLike.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { BaseElement } from "./BaseElement";
import { WCNativeHelper } from "./WCNativeHelper";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let BaseComboBoxLike = /*@__PURE__*/ (() => {
    class BaseComboBoxLike extends BaseElement {
        constructor() {
            super(...arguments);
            this.a = new WCNativeHelper();
            this.i = null;
        }
        get b() {
            return this.a;
        }
        get nativeElement() {
            return this.i;
        }
        set nativeElement(a) {
            this.i = a;
            this.a.o = this.i;
        }
        setNativeElement(a) {
            this.nativeElement = a;
        }
        get j() {
            let ret_ = this.b.n("emitEvent");
            return ret_;
        }
        set j(a) {
            let value_ = a;
            this.b.w("emitEvent", value_);
        }
        get e() {
            let ret_ = this.b.n("keepOpenOnSelect");
            return ret_;
        }
        set e(a) {
            let value_ = a;
            this.b.w("keepOpenOnSelect", value_);
        }
        get d() {
            let ret_ = this.b.n("keepOpenOnOutsideClick");
            return ret_;
        }
        set d(a) {
            let value_ = a;
            this.b.w("keepOpenOnOutsideClick", value_);
        }
        get f() {
            let ret_ = this.b.n("open");
            return ret_;
        }
        set f(a) {
            let value_ = a;
            this.b.w("open", value_);
        }
        g() {
            return (this.nativeElement.show());
        }
        c() {
            return (this.nativeElement.hide());
        }
        h() {
            return (this.nativeElement.toggle());
        }
    }
    BaseComboBoxLike.$t = /*@__PURE__*/ markType(BaseComboBoxLike, 'BaseComboBoxLike', BaseElement.$);
    return BaseComboBoxLike;
})();
