// igniteui-react/esm2015/lib/ButtonVariant.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
export var ButtonVariant = /*@__PURE__*/ (function (ButtonVariant) {
    ButtonVariant[ButtonVariant["Flat"] = 0] = "Flat";
    ButtonVariant[ButtonVariant["Contained"] = 1] = "Contained";
    ButtonVariant[ButtonVariant["Outlined"] = 2] = "Outlined";
    ButtonVariant[ButtonVariant["Fab"] = 3] = "Fab";
    return ButtonVariant;
})({});
/**
 * @hidden
 */
export let ButtonVariant_$type = /*@__PURE__*/ markEnum('ButtonVariant', 'Flat,0|Contained,1|Outlined,2|Fab,3');
