// igniteui-react/esm2015/lib/CalendarResourceStrings.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let CalendarResourceStrings = /*@__PURE__*/ (() => {
    class CalendarResourceStrings extends Base {
        constructor() {
            super(...arguments);
            this.j = null;
            this.l = null;
            this.h = null;
            this.k = null;
            this.i = null;
            this.m = null;
            this.a = null;
            this.e = null;
            this.b = null;
            this.f = null;
            this.c = null;
            this.g = null;
            this.d = null;
            this.n = null;
        }
    }
    CalendarResourceStrings.$t = /*@__PURE__*/ markType(CalendarResourceStrings, 'CalendarResourceStrings');
    return CalendarResourceStrings;
})();
