// igniteui-react/esm2015/lib/CardContent.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { BaseElement } from "./BaseElement";
import { WCNativeHelper } from "./WCNativeHelper";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let CardContent = /*@__PURE__*/ (() => {
    class CardContent extends BaseElement {
        constructor() {
            super(...arguments);
            this.a = new WCNativeHelper();
            this.c = null;
        }
        get b() {
            return this.a;
        }
        get nativeElement() {
            return this.c;
        }
        set nativeElement(a) {
            this.c = a;
            this.a.o = this.c;
        }
        setNativeElement(a) {
            this.nativeElement = a;
        }
    }
    CardContent.$t = /*@__PURE__*/ markType(CardContent, 'CardContent', BaseElement.$);
    return CardContent;
})();
