// igniteui-react/esm2015/lib/CardActions.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { BaseElement } from "./BaseElement";
import { WCNativeHelper } from "./WCNativeHelper";
import { enumGetBox, EnumUtil, markType } from "igniteui-react-core";
import { CardActionsOrientation_$type } from "./CardActionsOrientation";
/**
 * @hidden
 */
export let CardActions = /*@__PURE__*/ (() => {
    class CardActions extends BaseElement {
        constructor() {
            super(...arguments);
            this.b = new WCNativeHelper();
            this.d = null;
        }
        get c() {
            return this.b;
        }
        get nativeElement() {
            return this.d;
        }
        set nativeElement(a) {
            this.d = a;
            this.b.o = this.d;
        }
        setNativeElement(a) {
            this.nativeElement = a;
        }
        get a() {
            let ret_ = this.c.n("orientation");
            if ((ret_ === undefined && this.___rawMode)) {
                return undefined;
            }
            ret_ = enumGetBox(CardActionsOrientation_$type, EnumUtil.getEnumValue(CardActionsOrientation_$type, this.c.f(CardActionsOrientation_$type, ret_)));
            return EnumUtil.getEnumValue(CardActionsOrientation_$type, ret_);
        }
        set a(a) {
            let value_ = enumGetBox(CardActionsOrientation_$type, a);
            value_ = WCNativeHelper.camel(value_.toString());
            this.c.w("orientation", value_);
        }
    }
    CardActions.$t = /*@__PURE__*/ markType(CardActions, 'CardActions', BaseElement.$);
    return CardActions;
})();
