// igniteui-react/esm2015/lib/ComboChangeEventArgsDetail.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, enumGetBox, EnumUtil, markType } from "igniteui-react-core";
import { WCNativeHelper } from "./WCNativeHelper";
import { ComboChangeType_$type } from "./ComboChangeType";
/**
 * @hidden
 */
export let ComboChangeEventArgsDetail = /*@__PURE__*/ (() => {
    class ComboChangeEventArgsDetail extends Base {
        constructor() {
            super(...arguments);
            this.d = new WCNativeHelper();
            this.f = null;
        }
        get b() {
            let ret_ = this.e.n("newValue");
            return ret_;
        }
        set b(a) {
            let value_ = a;
            this.e.w("newValue", value_);
        }
        get a() {
            let ret_ = this.e.n("items");
            return ret_;
        }
        set a(a) {
            let value_ = a;
            this.e.w("items", value_);
        }
        get c() {
            let ret_ = this.e.n("type");
            if ((ret_ === undefined && this.___rawMode)) {
                return undefined;
            }
            ret_ = enumGetBox(ComboChangeType_$type, EnumUtil.getEnumValue(ComboChangeType_$type, this.e.f(ComboChangeType_$type, ret_)));
            return EnumUtil.getEnumValue(ComboChangeType_$type, ret_);
        }
        set c(a) {
            let value_ = enumGetBox(ComboChangeType_$type, a);
            value_ = WCNativeHelper.camel(value_.toString());
            this.e.w("type", value_);
        }
        get e() {
            return this.d;
        }
        get nativeElement() {
            return this.f;
        }
        set nativeElement(a) {
            this.f = a;
            this.d.o = this.f;
        }
        setNativeElement(a) {
            this.nativeElement = a;
        }
    }
    ComboChangeEventArgsDetail.$t = /*@__PURE__*/ markType(ComboChangeEventArgsDetail, 'ComboChangeEventArgsDetail');
    return ComboChangeEventArgsDetail;
})();
