// igniteui-react/esm2015/lib/DatePart.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
export var DatePart = /*@__PURE__*/ (function (DatePart) {
    DatePart[DatePart["Month"] = 0] = "Month";
    DatePart[DatePart["Year"] = 1] = "Year";
    DatePart[DatePart["Date"] = 2] = "Date";
    DatePart[DatePart["Hours"] = 3] = "Hours";
    DatePart[DatePart["Minutes"] = 4] = "Minutes";
    DatePart[DatePart["Seconds"] = 5] = "Seconds";
    DatePart[DatePart["AmPm"] = 6] = "AmPm";
    return DatePart;
})({});
/**
 * @hidden
 */
export let DatePart_$type = /*@__PURE__*/ markEnum('DatePart', 'Month:month,0|Year:year,1|Date:date,2|Hours:hours,3|Minutes:minutes,4|Seconds:seconds,5|AmPm:amPm,6');
