// igniteui-react/esm2015/lib/Divider.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { BaseElement } from "./BaseElement";
import { WCNativeHelper } from "./WCNativeHelper";
import { enumGetBox, EnumUtil, markType } from "igniteui-react-core";
import { DividerType_$type } from "./DividerType";
/**
 * @hidden
 */
export let Divider = /*@__PURE__*/ (() => {
    class Divider extends BaseElement {
        constructor() {
            super(...arguments);
            this.b = new WCNativeHelper();
            this.f = null;
        }
        get c() {
            return this.b;
        }
        get nativeElement() {
            return this.f;
        }
        set nativeElement(a) {
            this.f = a;
            this.b.o = this.f;
        }
        setNativeElement(a) {
            this.nativeElement = a;
        }
        get e() {
            let ret_ = this.c.n("vertical");
            return ret_;
        }
        set e(a) {
            let value_ = a;
            this.c.w("vertical", value_);
        }
        get d() {
            let ret_ = this.c.n("middle");
            return ret_;
        }
        set d(a) {
            let value_ = a;
            this.c.w("middle", value_);
        }
        get a() {
            let ret_ = this.c.n("type");
            if ((ret_ === undefined && this.___rawMode)) {
                return undefined;
            }
            ret_ = enumGetBox(DividerType_$type, EnumUtil.getEnumValue(DividerType_$type, this.c.f(DividerType_$type, ret_)));
            return EnumUtil.getEnumValue(DividerType_$type, ret_);
        }
        set a(a) {
            let value_ = enumGetBox(DividerType_$type, a);
            value_ = WCNativeHelper.camel(value_.toString());
            this.c.w("type", value_);
        }
    }
    Divider.$t = /*@__PURE__*/ markType(Divider, 'Divider', BaseElement.$);
    return Divider;
})();
