// igniteui-react/esm2015/lib/DockManagerPoint.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { WCNativeHelper } from "./WCNativeHelper";
/**
 * @hidden
 */
export let DockManagerPoint = /*@__PURE__*/ (() => {
    class DockManagerPoint extends Base {
        constructor() {
            super(...arguments);
            this.a = new WCNativeHelper();
            this.e = null;
        }
        get b() {
            return this.a;
        }
        get nativeElement() {
            return this.e;
        }
        set nativeElement(a) {
            this.e = a;
            this.a.o = this.e;
        }
        setNativeElement(a) {
            this.nativeElement = a;
        }
        get c() {
            let ret_ = this.b.n("x");
            return ret_;
        }
        set c(a) {
            let value_ = a;
            this.b.w("x", value_);
        }
        get d() {
            let ret_ = this.b.n("y");
            return ret_;
        }
        set d(a) {
            let value_ = a;
            this.b.w("y", value_);
        }
    }
    DockManagerPoint.$t = /*@__PURE__*/ markType(DockManagerPoint, 'DockManagerPoint');
    return DockManagerPoint;
})();
