// igniteui-react/esm2015/lib/DockingIndicator.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, enumGetBox, EnumUtil, markType } from "igniteui-react-core";
import { WCNativeHelper } from "./WCNativeHelper";
import { DockingIndicatorPosition_$type } from "./DockingIndicatorPosition";
/**
 * @hidden
 */
export let DockingIndicator = /*@__PURE__*/ (() => {
    class DockingIndicator extends Base {
        constructor() {
            super(...arguments);
            this.b = new WCNativeHelper();
            this.e = null;
        }
        get c() {
            return this.b;
        }
        get nativeElement() {
            return this.e;
        }
        set nativeElement(a) {
            this.e = a;
            this.b.o = this.e;
        }
        setNativeElement(a) {
            this.nativeElement = a;
        }
        get a() {
            let ret_ = this.c.n("position");
            if ((ret_ === undefined && this.___rawMode)) {
                return undefined;
            }
            ret_ = enumGetBox(DockingIndicatorPosition_$type, EnumUtil.getEnumValue(DockingIndicatorPosition_$type, this.c.f(DockingIndicatorPosition_$type, ret_)));
            return EnumUtil.getEnumValue(DockingIndicatorPosition_$type, ret_);
        }
        set a(a) {
            let value_ = enumGetBox(DockingIndicatorPosition_$type, a);
            value_ = WCNativeHelper.camel(value_.toString());
            this.c.w("position", value_);
        }
        get d() {
            let ret_ = this.c.n("isRoot");
            return ret_;
        }
        set d(a) {
            let value_ = a;
            this.c.w("isRoot", value_);
        }
        get g() {
            let ret_ = this.c.n("direction");
            return ret_;
        }
        set g(a) {
            let value_ = a;
            this.c.w("direction", value_);
        }
    }
    DockingIndicator.$t = /*@__PURE__*/ markType(DockingIndicator, 'DockingIndicator');
    return DockingIndicator;
})();
