// igniteui-react/esm2015/lib/DockPaneAction.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { PaneDragAction } from "./PaneDragAction";
import { WCNativeHelper } from "./WCNativeHelper";
import { enumGetBox, EnumUtil, markType } from "igniteui-react-core";
import { PaneDragActionType_$type } from "./PaneDragActionType";
import { DockingIndicator } from "./DockingIndicator";
import { DockManagerPane } from "./DockManagerPane";
/**
 * @hidden
 */
export let DockPaneAction = /*@__PURE__*/ (() => {
    class DockPaneAction extends PaneDragAction {
        constructor() {
            super(...arguments);
            this.d = new WCNativeHelper();
            this.f = null;
        }
        get e() {
            return this.d;
        }
        get nativeElement() {
            return this.f;
        }
        set nativeElement(a) {
            this.f = a;
            this.d.o = this.f;
        }
        setNativeElement(a) {
            this.nativeElement = a;
        }
        get c() {
            let ret_ = this.e.n("type");
            if ((ret_ === undefined && this.___rawMode)) {
                return undefined;
            }
            ret_ = enumGetBox(PaneDragActionType_$type, EnumUtil.getEnumValue(PaneDragActionType_$type, this.e.f(PaneDragActionType_$type, ret_)));
            return EnumUtil.getEnumValue(PaneDragActionType_$type, ret_);
        }
        set c(a) {
            let value_ = enumGetBox(PaneDragActionType_$type, a);
            value_ = WCNativeHelper.camel(value_.toString());
            this.e.w("type", value_);
        }
        get a() {
            let ret_ = this.e.m("dockingIndicator", (a) => new DockingIndicator());
            return ret_;
        }
        set a(a) {
            let value_ = a;
            this.e.w("dockingIndicator", value_);
        }
        get b() {
            let ret_ = this.e.m("targetPane", (a) => DockManagerPane.createFromNative(a));
            return ret_;
        }
        set b(a) {
            let value_ = a;
            this.e.w("targetPane", value_);
        }
    }
    DockPaneAction.$t = /*@__PURE__*/ markType(DockPaneAction, 'DockPaneAction', PaneDragAction.$);
    return DockPaneAction;
})();
