// igniteui-react/esm2015/lib/DockManagerLayout.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { WCNativeHelper } from "./WCNativeHelper";
import { SplitPane } from "./SplitPane";
import { SplitPaneCollection } from "./SplitPaneCollection";
/**
 * @hidden
 */
export let DockManagerLayout = /*@__PURE__*/ (() => {
    class DockManagerLayout extends Base {
        constructor() {
            super(...arguments);
            this.___mustPassByValue = true;
            this.c = new WCNativeHelper();
            this.f = null;
        }
        get h() {
            let ret_ = this.d.n("name");
            return ret_;
        }
        set h(a) {
            let value_ = a;
            this.d.w("name", value_);
        }
        get d() {
            return this.c;
        }
        get nativeElement() {
            return this.f;
        }
        set nativeElement(a) {
            this.f = a;
            this.c.o = this.f;
        }
        setNativeElement(a) {
            this.nativeElement = a;
        }
        get a() {
            let ret_ = this.d.m("rootPane", (a) => new SplitPane());
            return ret_;
        }
        set a(a) {
            let value_ = a;
            this.d.w("rootPane", value_);
        }
        get b() {
            let ret_ = this.d.k("floatingPanes", (a) => new SplitPaneCollection(), (a) => new SplitPane());
            return ret_;
        }
    }
    DockManagerLayout.$t = /*@__PURE__*/ markType(DockManagerLayout, 'DockManagerLayout');
    return DockManagerLayout;
})();
