// igniteui-react/esm2015/lib/FloatingPaneResizeEventArgsDetail.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, enumGetBox, EnumUtil, markType } from "igniteui-react-core";
import { WCNativeHelper } from "./WCNativeHelper";
import { SplitPane } from "./SplitPane";
import { ResizerLocation_$type } from "./ResizerLocation";
/**
 * @hidden
 */
export let FloatingPaneResizeEventArgsDetail = /*@__PURE__*/ (() => {
    class FloatingPaneResizeEventArgsDetail extends Base {
        constructor() {
            super(...arguments);
            this.c = new WCNativeHelper();
            this.e = null;
        }
        get d() {
            return this.c;
        }
        get nativeElement() {
            return this.e;
        }
        set nativeElement(a) {
            this.e = a;
            this.c.o = this.e;
        }
        setNativeElement(a) {
            this.nativeElement = a;
        }
        get b() {
            let ret_ = this.d.m("sourcePane", (a) => new SplitPane());
            return ret_;
        }
        set b(a) {
            let value_ = a;
            this.d.w("sourcePane", value_);
        }
        get a() {
            let ret_ = this.d.n("resizerLocation");
            if ((ret_ === undefined && this.___rawMode)) {
                return undefined;
            }
            ret_ = enumGetBox(ResizerLocation_$type, EnumUtil.getEnumValue(ResizerLocation_$type, this.d.f(ResizerLocation_$type, ret_)));
            return EnumUtil.getEnumValue(ResizerLocation_$type, ret_);
        }
        set a(a) {
            let value_ = enumGetBox(ResizerLocation_$type, a);
            value_ = WCNativeHelper.camel(value_.toString());
            this.d.w("resizerLocation", value_);
        }
    }
    FloatingPaneResizeEventArgsDetail.$t = /*@__PURE__*/ markType(FloatingPaneResizeEventArgsDetail, 'FloatingPaneResizeEventArgsDetail');
    return FloatingPaneResizeEventArgsDetail;
})();
