// igniteui-react/esm2015/lib/FloatingPaneResizeMoveEventArgsDetail.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { FloatingPaneResizeEventArgsDetail } from "./FloatingPaneResizeEventArgsDetail";
import { markType } from "igniteui-react-core";
import { DockManagerPoint } from "./DockManagerPoint";
/**
 * @hidden
 */
export let FloatingPaneResizeMoveEventArgsDetail = /*@__PURE__*/ (() => {
    class FloatingPaneResizeMoveEventArgsDetail extends FloatingPaneResizeEventArgsDetail {
        get m() {
            let ret_ = this.d.n("oldWidth");
            return ret_;
        }
        set m(a) {
            let value_ = a;
            this.d.w("oldWidth", value_);
        }
        get k() {
            let ret_ = this.d.n("newWidth");
            return ret_;
        }
        set k(a) {
            let value_ = a;
            this.d.w("newWidth", value_);
        }
        get l() {
            let ret_ = this.d.n("oldHeight");
            return ret_;
        }
        set l(a) {
            let value_ = a;
            this.d.w("oldHeight", value_);
        }
        get j() {
            let ret_ = this.d.n("newHeight");
            return ret_;
        }
        set j(a) {
            let value_ = a;
            this.d.w("newHeight", value_);
        }
        get i() {
            let ret_ = this.d.m("oldLocation", (a) => new DockManagerPoint());
            return ret_;
        }
        set i(a) {
            let value_ = a;
            this.d.w("oldLocation", value_);
        }
        get h() {
            let ret_ = this.d.m("newLocation", (a) => new DockManagerPoint());
            return ret_;
        }
        set h(a) {
            let value_ = a;
            this.d.w("newLocation", value_);
        }
    }
    FloatingPaneResizeMoveEventArgsDetail.$t = /*@__PURE__*/ markType(FloatingPaneResizeMoveEventArgsDetail, 'FloatingPaneResizeMoveEventArgsDetail', FloatingPaneResizeEventArgsDetail.$);
    return FloatingPaneResizeMoveEventArgsDetail;
})();
