// igniteui-react/esm2015/lib/igr-component-bool-value-changed-event-args.js
import { ComponentBoolValueChangedEventArgs as ComponentBoolValueChangedEventArgs_internal } from "./ComponentBoolValueChangedEventArgs";
import { ensureBool } from "igniteui-react-core";
export class IgrComponentBoolValueChangedEventArgs {
    createImplementation() {
        return new ComponentBoolValueChangedEventArgs_internal();
    }
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    get detail() {
        return this.i.b;
    }
    set detail(v) {
        this.i.b = ensureBool(v);
    }
}
