// igniteui-react/esm2015/lib/igr-calendar-format-options.js
import { CalendarFormatOptions as CalendarFormatOptions_internal } from "./CalendarFormatOptions";
export class IgrCalendarFormatOptions {
    createImplementation() {
        return new CalendarFormatOptions_internal();
    }
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    get weekday() {
        return this.i.weekday;
    }
    set weekday(v) {
        this.i.weekday = v;
    }
    get month() {
        return this.i.month;
    }
    set month(v) {
        this.i.month = v;
    }
    findByName(name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        return null;
    }
}
