// igniteui-react/esm2015/lib/igr-active-step-changed-args-detail.js
import { ActiveStepChangedArgsDetail as ActiveStepChangedArgsDetail_internal } from "./ActiveStepChangedArgsDetail";
export class IgrActiveStepChangedArgsDetail {
    createImplementation() {
        return new ActiveStepChangedArgsDetail_internal();
    }
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    get index() {
        return this.i.c;
    }
    set index(v) {
        this.i.c = +v;
    }
    findByName(name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        return null;
    }
}
