// igniteui-react/esm2015/lib/igr-button-group.js
import * as React from 'react';
import { delegateCombine, delegateRemove } from "igniteui-react-core";
import { ButtonGroupAlignment_$type } from "./ButtonGroupAlignment";
import { ButtonGroupSelection_$type } from "./ButtonGroupSelection";
import { IgrComponentValueChangedEventArgs } from "./igr-component-value-changed-event-args";
import { ButtonGroup } from "./ButtonGroup";
import { TypeRegistrar } from "igniteui-react-core";
import { ReactRenderer, PortalManager } from "igniteui-react-core";
import { IgcButtonGroupComponent, defineComponents } from "igniteui-webcomponents";
import { NamePatcher, getModifiedProps, isValidProp, ensureBool, ensureEnum, toSpinal, initializePropertiesFromCss } from "igniteui-react-core";
/**
 * The `igc-button-group` groups a series of `igc-toggle-button`s together, exposing features such as layout and selection.
*/
export class IgrButtonGroup extends React.Component {
    createImplementation() {
        return new ButtonGroup();
    }
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    } /**
                                 * @hidden
                                 */
    static _createFromInternal(internal) {
        if (!internal) {
            return null;
        }
        if (!internal.$type) {
            return null;
        }
        let name = internal.$type.name;
        let externalName = "Igr" + name;
        if (!TypeRegistrar.isRegistered(externalName)) {
            return null;
        }
        return TypeRegistrar.create(externalName);
    }
    onImplementationCreated() {
    }
    constructor(props) {
        super(props);
        this.mounted = false;
        this.__p = null;
        this._hasUserValues = new Set();
        this._stylingContainer = null;
        this._stylingParent = null;
        this._inStyling = false;
        this._select = null;
        this._select_wrapped = null;
        this._deselect = null;
        this._deselect_wrapped = null;
        if (this._styling) {
            NamePatcher.ensureStylablePatched(Object.getPrototypeOf(this));
        }
        this._getMainRef = this._getMainRef.bind(this);
        defineComponents(IgcButtonGroupComponent);
        this._implementation = this.createImplementation();
        this._portalManager = new PortalManager("templates", () => {
            if (this.mounted) {
                this.setState({});
            }
        });
        if (typeof window !== 'undefined' && typeof document !== 'undefined') {
            this._renderer = new ReactRenderer(this._implementation.nativeElement, document, false, null, this._portalManager);
        }
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    componentDidMount() {
        this.i.nativeElement = this._elRef;
        this.mounted = true;
        for (const p of Object.keys(this.props)) {
            if (isValidProp(this, p)) {
                this[p] = this.props[p];
            }
        }
    }
    shouldComponentUpdate(nextProps, nextState) {
        const mod = getModifiedProps(this.props, nextProps);
        for (const p of Object.keys(mod)) {
            if (isValidProp(this, p)) {
                this[p] = mod[p];
            }
        }
        return true;
    }
    render() {
        const nativePropsName = Object.keys(this.props).filter(prop => !isValidProp(this, prop) && prop !== "originalRef");
        const nativeProps = {};
        nativePropsName.forEach(propName => {
            nativeProps[propName] = this.props[propName];
        });
        let propChildren = this.props.children;
        let children = [];
        React.Children.forEach(propChildren, (ch) => {
            children.push(React.cloneElement(ch));
        });
        this._portalManager.onRender(children);
        let style = {};
        if (this.props.style) {
            style = this.props.style;
        }
        let div = React.createElement("igc-button-group", Object.assign(Object.assign({}, nativeProps), { ref: this._getMainRef, id: this.props.id, class: this.props.className, style: style, children: children }));
        return div;
    }
    _getMainRef(ref) {
        this._elRef = ref;
    }
    /**
     * Disables all buttons inside the group.
    */
    get disabled() {
        return this.i.h;
    }
    set disabled(v) {
        this.i.h = ensureBool(v);
    }
    /**
     * Sets the orientation of the buttons in the group.
    */
    get alignment() {
        return this.i.b;
    }
    set alignment(v) {
        this.i.b = ensureEnum(ButtonGroupAlignment_$type, v);
    }
    /**
     * Controls the mode of selection for the button group.
    */
    get selection() {
        return this.i.c;
    }
    set selection(v) {
        this.i.c = ensureEnum(ButtonGroupSelection_$type, v);
    }
    get selectedItems() {
        return this.i.a;
    }
    set selectedItems(v) {
        if (v && !Array.isArray(v) && typeof (v) == "string") {
            const re = /\s*(?:,|\s|$)\s*/gm;
            v = v.split(re);
        }
        this.i.a = v;
    }
    findByName(name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        return null;
    }
    get hasUserValues() {
        return this._hasUserValues;
    }
    __m(propertyName) {
        if (!this._inStyling) {
            this._hasUserValues.add(propertyName);
        }
    }
    _styling(container, component, parent) {
        if (this._inStyling) {
            return;
        }
        this._inStyling = true;
        this._stylingContainer = container;
        this._stylingParent = component;
        let genericPrefix = "";
        let typeName = this.i.$type.name;
        if (typeName.indexOf("Xam") === 0) {
            typeName = typeName.substring(3);
        }
        genericPrefix = toSpinal("ButtonGroup");
        let additionalPrefixes = [];
        let prefix = toSpinal(typeName);
        additionalPrefixes.push(prefix + "-");
        let b = this.i.$type.baseType;
        while (b && b.name != "Object" &&
            b.name != "Base" &&
            b.name != "Control" &&
            b.Name != "DependencyObject" &&
            b.Name != "FrameworkElement") {
            typeName = b.name;
            if (typeName.indexOf("Xam") === 0) {
                typeName = typeName.substring(3);
            }
            let basePrefix = toSpinal(typeName);
            additionalPrefixes.push(basePrefix + "-");
            b = b.baseType;
        }
        if (parent) {
            let parentTypeName = parent.i.$type.name;
            if (parentTypeName.indexOf("Xam") === 0) {
                parentTypeName = parentTypeName.substring(3);
            }
            let parentPrefix = toSpinal(parentTypeName);
            additionalPrefixes.push(parentPrefix + "-" + genericPrefix + "-");
            additionalPrefixes.push(parentPrefix + "-" + prefix + "-");
        }
        initializePropertiesFromCss(container, this, genericPrefix + "-", this.hasUserValues, false, additionalPrefixes);
        if (this._otherStyling) {
            this._otherStyling(container, component, parent);
        }
        this._inStyling = false;
    }
    setNativeElement(element) {
        this.i.setNativeElement(element);
    }
    get select() {
        return this._select;
    }
    set select(ev) {
        if (this._select_wrapped !== null) {
            this.i.select = delegateRemove(this.i.select, this._select_wrapped);
            this._select_wrapped = null;
            this._select = null;
        }
        this._select = ev;
        this._select_wrapped = (o, e) => {
            let outerArgs = new IgrComponentValueChangedEventArgs();
            outerArgs._provideImplementation(e);
            if (this.beforeSelect) {
                this.beforeSelect(this, outerArgs);
            }
            if (this._select) {
                this._select(this, outerArgs);
            }
        };
        this.i.select = delegateCombine(this.i.select, this._select_wrapped);
        if (this.i.selectChanged) {
            this.i.selectChanged();
        }
        ;
    }
    get deselect() {
        return this._deselect;
    }
    set deselect(ev) {
        if (this._deselect_wrapped !== null) {
            this.i.deselect = delegateRemove(this.i.deselect, this._deselect_wrapped);
            this._deselect_wrapped = null;
            this._deselect = null;
        }
        this._deselect = ev;
        this._deselect_wrapped = (o, e) => {
            let outerArgs = new IgrComponentValueChangedEventArgs();
            outerArgs._provideImplementation(e);
            if (this.beforeDeselect) {
                this.beforeDeselect(this, outerArgs);
            }
            if (this._deselect) {
                this._deselect(this, outerArgs);
            }
        };
        this.i.deselect = delegateCombine(this.i.deselect, this._deselect_wrapped);
        if (this.i.deselectChanged) {
            this.i.deselectChanged();
        }
        ;
    }
}
