// igniteui-react/esm2015/lib/igr-date-part-deltas.js
import { DatePartDeltas as DatePartDeltas_internal } from "./DatePartDeltas";
export class IgrDatePartDeltas {
    createImplementation() {
        return new DatePartDeltas_internal();
    }
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    get date() {
        return this.i.a;
    }
    set date(v) {
        this.i.a = +v;
    }
    get month() {
        return this.i.d;
    }
    set month(v) {
        this.i.d = +v;
    }
    get year() {
        return this.i.f;
    }
    set year(v) {
        this.i.f = +v;
    }
    get hours() {
        return this.i.b;
    }
    set hours(v) {
        this.i.b = +v;
    }
    get minutes() {
        return this.i.c;
    }
    set minutes(v) {
        this.i.c = +v;
    }
    get seconds() {
        return this.i.e;
    }
    set seconds(v) {
        this.i.e = +v;
    }
    findByName(name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        return null;
    }
}
