// igniteui-react/esm2015/lib/igr-dock-pane-action.js
import { PaneDragActionType_$type } from "./PaneDragActionType";
import { IgrDockingIndicator } from "./igr-docking-indicator";
import { IgrDockManagerPane } from "./igr-dock-manager-pane";
import { IgrPaneDragAction } from "./igr-pane-drag-action";
import { DockPaneAction } from "./DockPaneAction";
import { ensureEnum } from "igniteui-react-core";
export class IgrDockPaneAction extends IgrPaneDragAction {
    createImplementation() {
        return new DockPaneAction();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor(props) {
        super(props);
    }
    get type() {
        return this.i.c;
    }
    set type(v) {
        this.i.c = ensureEnum(PaneDragActionType_$type, v);
    }
    get dockingIndicator() {
        const r = this.i.a;
        if (r == null) {
            return null;
        }
        if (!r.externalObject) {
            let e = new IgrDockingIndicator();
            if (r.$type) {
                e._implementation = r;
            }
            else {
                if (e.i.setNativeElement) {
                    e.i.setNativeElement(r);
                }
            }
            r.externalObject = e;
        }
        return r.externalObject;
    }
    set dockingIndicator(v) {
        v == null ? this.i.a = null : this.i.a = v.i;
    }
    get targetPane() {
        const r = this.i.b;
        if (r == null) {
            return null;
        }
        if (!r.externalObject) {
            let e = IgrDockManagerPane._createFromInternal(r);
            if (e) {
                e._implementation = r;
            }
            r.externalObject = e;
        }
        return r.externalObject;
    }
    set targetPane(v) {
        if (v != null && this._stylingContainer && v._styling)
            v._styling(this._stylingContainer, this, this);
        v == null ? this.i.b = null : this.i.b = v.i;
    }
    findByName(name) {
        var baseResult = super.findByName(name);
        if (baseResult) {
            return baseResult;
        }
        if (this.dockingIndicator && this.dockingIndicator.name && this.dockingIndicator.name == name) {
            return this.dockingIndicator;
        }
        if (this.targetPane && this.targetPane.name && this.targetPane.name == name) {
            return this.targetPane;
        }
        return null;
    }
    _styling(container, component, parent) {
        super._styling(container, component, parent);
        this._inStyling = true;
        if (this.targetPane && this.targetPane._styling) {
            this.targetPane._styling(container, component, this);
        }
        this._inStyling = false;
    }
    setNativeElement(element) {
        this.i.setNativeElement(element);
    }
}
