// igniteui-react/esm2015/lib/igr-dock-manager-point.js
import { DockManagerPoint as DockManagerPoint_internal } from "./DockManagerPoint";
export class IgrDockManagerPoint {
    createImplementation() {
        return new DockManagerPoint_internal();
    }
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    get x() {
        return this.i.c;
    }
    set x(v) {
        this.i.c = +v;
    }
    get y() {
        return this.i.d;
    }
    set y(v) {
        this.i.d = +v;
    }
    findByName(name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        return null;
    }
    setNativeElement(element) {
        this.i.setNativeElement(element);
    }
}
