// igniteui-react/esm2015/lib/igr-button.js
import * as React from 'react';
import { ButtonVariant_$type } from "./ButtonVariant";
import { IgrButtonBase } from "./igr-button-base";
import { Button } from "./Button";
import { IgcButtonComponent, defineComponents } from "igniteui-webcomponents";
import { isValidProp, ensureEnum } from "igniteui-react-core";
/**
 * Represents a clickable button, used to submit forms or anywhere in a
 * document for accessible, standard button functionality.
*/
export class IgrButton extends IgrButtonBase {
    createImplementation() {
        return new Button();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor(props) {
        super(props);
        this._getMainRef = this._getMainRef.bind(this);
        defineComponents(IgcButtonComponent);
    }
    componentDidMount() {
        this.i.nativeElement = this._elRef;
        this.mounted = true;
        super.componentDidMount();
    }
    render() {
        const nativePropsName = Object.keys(this.props).filter(prop => !isValidProp(this, prop) && prop !== "originalRef");
        const nativeProps = {};
        nativePropsName.forEach(propName => {
            nativeProps[propName] = this.props[propName];
        });
        let propChildren = this.props.children;
        let children = [];
        React.Children.forEach(propChildren, (ch) => {
            children.push(React.cloneElement(ch));
        });
        this._portalManager.onRender(children);
        let style = {};
        if (this.props.style) {
            style = this.props.style;
        }
        let div = React.createElement("igc-button", Object.assign(Object.assign({}, nativeProps), { ref: this._getMainRef, id: this.props.id, class: this.props.className, style: style, children: children }));
        return div;
    }
    _getMainRef(ref) {
        this._elRef = ref;
    }
    /**
     * Sets the variant of the button.
    */
    get variant() {
        return this.i.v;
    }
    set variant(v) {
        this.i.v = ensureEnum(ButtonVariant_$type, v);
    }
}
