// igniteui-react/esm2015/lib/igr-checkbox-change-event-args-detail.js
import { CheckboxChangeEventArgsDetail as CheckboxChangeEventArgsDetail_internal } from "./CheckboxChangeEventArgsDetail";
import { ensureBool } from "igniteui-react-core";
export class IgrCheckboxChangeEventArgsDetail {
    createImplementation() {
        return new CheckboxChangeEventArgsDetail_internal();
    }
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    get checked() {
        return this.i.c;
    }
    set checked(v) {
        this.i.c = ensureBool(v);
    }
    get value() {
        return this.i.f;
    }
    set value(v) {
        this.i.f = v;
    }
    findByName(name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        return null;
    }
    setNativeElement(element) {
        this.i.setNativeElement(element);
    }
}
