// igniteui-react/esm2015/lib/igr-expansion-panel-component-event-args.js
import { IgrExpansionPanel } from "./igr-expansion-panel";
import { ExpansionPanelComponentEventArgs as ExpansionPanelComponentEventArgs_internal } from "./ExpansionPanelComponentEventArgs";
export class IgrExpansionPanelComponentEventArgs {
    createImplementation() {
        return new ExpansionPanelComponentEventArgs_internal();
    }
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    get detail() {
        const r = this.i.a;
        if (r == null) {
            return null;
        }
        if (!r.externalObject) {
            let e = new IgrExpansionPanel({});
            if (r.$type) {
                e._implementation = r;
            }
            else {
                if (e.i.setNativeElement) {
                    e.i.setNativeElement(r);
                }
            }
            r.externalObject = e;
        }
        return r.externalObject;
    }
    set detail(v) {
        v == null ? this.i.a = null : this.i.a = v.i;
    }
}
