// igniteui-react/esm2015/lib/igr-calendar-resource-strings.js
import { CalendarResourceStrings as CalendarResourceStrings_internal } from "./CalendarResourceStrings";
export class IgrCalendarResourceStrings {
    createImplementation() {
        return new CalendarResourceStrings_internal();
    }
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    get selectMonth() {
        return this.i.j;
    }
    set selectMonth(v) {
        this.i.j = v;
    }
    get selectYear() {
        return this.i.l;
    }
    set selectYear(v) {
        this.i.l = v;
    }
    get selectDate() {
        return this.i.h;
    }
    set selectDate(v) {
        this.i.h = v;
    }
    get selectRange() {
        return this.i.k;
    }
    set selectRange(v) {
        this.i.k = v;
    }
    get selectedDate() {
        return this.i.i;
    }
    set selectedDate(v) {
        this.i.i = v;
    }
    get startDate() {
        return this.i.m;
    }
    set startDate(v) {
        this.i.m = v;
    }
    get endDate() {
        return this.i.a;
    }
    set endDate(v) {
        this.i.a = v;
    }
    get previousMonth() {
        return this.i.e;
    }
    set previousMonth(v) {
        this.i.e = v;
    }
    get nextMonth() {
        return this.i.b;
    }
    set nextMonth(v) {
        this.i.b = v;
    }
    get previousYear() {
        return this.i.f;
    }
    set previousYear(v) {
        this.i.f = v;
    }
    get nextYear() {
        return this.i.c;
    }
    set nextYear(v) {
        this.i.c = v;
    }
    get previousYears() {
        return this.i.g;
    }
    set previousYears(v) {
        this.i.g = v;
    }
    get nextYears() {
        return this.i.d;
    }
    set nextYears(v) {
        this.i.d = v;
    }
    get weekLabel() {
        return this.i.n;
    }
    set weekLabel(v) {
        this.i.n = v;
    }
    findByName(name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        return null;
    }
}
