// igniteui-react/esm2015/lib/igr-pane-header-connection-event-args-detail.js
import { IgrContentPane } from "./igr-content-pane";
import { IgrPaneHeaderElement } from "./igr-pane-header-element";
import { PaneHeaderConnectionEventArgsDetail as PaneHeaderConnectionEventArgsDetail_internal } from "./PaneHeaderConnectionEventArgsDetail";
export class IgrPaneHeaderConnectionEventArgsDetail {
    createImplementation() {
        return new PaneHeaderConnectionEventArgsDetail_internal();
    }
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    /**
     * The pane whose header is being connected/disconnected.
    */
    get pane() {
        const r = this.i.a;
        if (r == null) {
            return null;
        }
        if (!r.externalObject) {
            let e = IgrContentPane._createFromInternal(r);
            if (e) {
                e._implementation = r;
            }
            r.externalObject = e;
        }
        return r.externalObject;
    }
    set pane(v) {
        v == null ? this.i.a = null : this.i.a = v.i;
    }
    /**
     * The header element that is being connected/disconnected.
    */
    get element() {
        const r = this.i.b;
        if (r == null) {
            return null;
        }
        if (!r.externalObject) {
            let e = new IgrPaneHeaderElement();
            if (r.$type) {
                e._implementation = r;
            }
            else {
                if (e.i.setNativeElement) {
                    e.i.setNativeElement(r);
                }
            }
            r.externalObject = e;
        }
        return r.externalObject;
    }
    set element(v) {
        v == null ? this.i.b = null : this.i.b = v.i;
    }
    findByName(name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        if (this.pane && this.pane.name && this.pane.name == name) {
            return this.pane;
        }
        if (this.element && this.element.name && this.element.name == name) {
            return this.element;
        }
        return null;
    }
    setNativeElement(element) {
        this.i.setNativeElement(element);
    }
}
