// igniteui-react/esm2015/lib/igr-chip.js
import * as React from 'react';
import { delegateCombine, delegateRemove } from "igniteui-react-core";
import { StyleVariant_$type } from "./StyleVariant";
import { IgrComponentBoolValueChangedEventArgs } from "./igr-component-bool-value-changed-event-args";
import { Chip } from "./Chip";
import { TypeRegistrar } from "igniteui-react-core";
import { ReactRenderer, PortalManager } from "igniteui-react-core";
import { IgcChipComponent, defineComponents } from "igniteui-webcomponents";
import { NamePatcher, getModifiedProps, isValidProp, ensureBool, ensureEnum, toSpinal, initializePropertiesFromCss } from "igniteui-react-core";
/**
 * Chips help people enter information, make selections, filter content, or trigger actions.
*/
export class IgrChip extends React.Component {
    createImplementation() {
        return new Chip();
    }
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    } /**
                                 * @hidden
                                 */
    static _createFromInternal(internal) {
        if (!internal) {
            return null;
        }
        if (!internal.$type) {
            return null;
        }
        let name = internal.$type.name;
        let externalName = "Igr" + name;
        if (!TypeRegistrar.isRegistered(externalName)) {
            return null;
        }
        return TypeRegistrar.create(externalName);
    }
    onImplementationCreated() {
    }
    constructor(props) {
        super(props);
        this.mounted = false;
        this.__p = null;
        this._hasUserValues = new Set();
        this._stylingContainer = null;
        this._stylingParent = null;
        this._inStyling = false;
        this._remove = null;
        this._remove_wrapped = null;
        this._select = null;
        this._select_wrapped = null;
        if (this._styling) {
            NamePatcher.ensureStylablePatched(Object.getPrototypeOf(this));
        }
        this._getMainRef = this._getMainRef.bind(this);
        defineComponents(IgcChipComponent);
        this._implementation = this.createImplementation();
        this._portalManager = new PortalManager("templates", () => {
            if (this.mounted) {
                this.setState({});
            }
        });
        if (typeof window !== 'undefined' && typeof document !== 'undefined') {
            this._renderer = new ReactRenderer(this._implementation.nativeElement, document, false, null, this._portalManager);
        }
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    componentDidMount() {
        this.i.nativeElement = this._elRef;
        this.mounted = true;
        for (const p of Object.keys(this.props)) {
            if (isValidProp(this, p)) {
                this[p] = this.props[p];
            }
        }
    }
    shouldComponentUpdate(nextProps, nextState) {
        const mod = getModifiedProps(this.props, nextProps);
        for (const p of Object.keys(mod)) {
            if (isValidProp(this, p)) {
                this[p] = mod[p];
            }
        }
        return true;
    }
    render() {
        const nativePropsName = Object.keys(this.props).filter(prop => !isValidProp(this, prop) && prop !== "originalRef");
        const nativeProps = {};
        nativePropsName.forEach(propName => {
            nativeProps[propName] = this.props[propName];
        });
        let propChildren = this.props.children;
        let children = [];
        React.Children.forEach(propChildren, (ch) => {
            children.push(React.cloneElement(ch));
        });
        this._portalManager.onRender(children);
        let style = {};
        if (this.props.style) {
            style = this.props.style;
        }
        let div = React.createElement("igc-chip", Object.assign(Object.assign({}, nativeProps), { ref: this._getMainRef, id: this.props.id, class: this.props.className, style: style, children: children }));
        return div;
    }
    _getMainRef(ref) {
        this._elRef = ref;
    }
    /**
     * Sets the disabled state for the chip.
    */
    get disabled() {
        return this.i.f;
    }
    set disabled(v) {
        this.i.f = ensureBool(v);
    }
    /**
     * Defines if the chip is removable or not.
    */
    get removable() {
        return this.i.g;
    }
    set removable(v) {
        this.i.g = ensureBool(v);
    }
    /**
     * Defines if the chip is selectable or not.
    */
    get selectable() {
        return this.i.h;
    }
    set selectable(v) {
        this.i.h = ensureBool(v);
    }
    /**
     * Defines if the chip is selected or not.
    */
    get selected() {
        return this.i.i;
    }
    set selected(v) {
        this.i.i = ensureBool(v);
    }
    /**
     * A property that sets the color variant of the chip component.
    */
    get variant() {
        return this.i.c;
    }
    set variant(v) {
        this.i.c = ensureEnum(StyleVariant_$type, v);
    }
    findByName(name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        return null;
    }
    get hasUserValues() {
        return this._hasUserValues;
    }
    __m(propertyName) {
        if (!this._inStyling) {
            this._hasUserValues.add(propertyName);
        }
    }
    _styling(container, component, parent) {
        if (this._inStyling) {
            return;
        }
        this._inStyling = true;
        this._stylingContainer = container;
        this._stylingParent = component;
        let genericPrefix = "";
        let typeName = this.i.$type.name;
        if (typeName.indexOf("Xam") === 0) {
            typeName = typeName.substring(3);
        }
        genericPrefix = toSpinal("Chip");
        let additionalPrefixes = [];
        let prefix = toSpinal(typeName);
        additionalPrefixes.push(prefix + "-");
        let b = this.i.$type.baseType;
        while (b && b.name != "Object" &&
            b.name != "Base" &&
            b.name != "Control" &&
            b.Name != "DependencyObject" &&
            b.Name != "FrameworkElement") {
            typeName = b.name;
            if (typeName.indexOf("Xam") === 0) {
                typeName = typeName.substring(3);
            }
            let basePrefix = toSpinal(typeName);
            additionalPrefixes.push(basePrefix + "-");
            b = b.baseType;
        }
        if (parent) {
            let parentTypeName = parent.i.$type.name;
            if (parentTypeName.indexOf("Xam") === 0) {
                parentTypeName = parentTypeName.substring(3);
            }
            let parentPrefix = toSpinal(parentTypeName);
            additionalPrefixes.push(parentPrefix + "-" + genericPrefix + "-");
            additionalPrefixes.push(parentPrefix + "-" + prefix + "-");
        }
        initializePropertiesFromCss(container, this, genericPrefix + "-", this.hasUserValues, false, additionalPrefixes);
        if (this._otherStyling) {
            this._otherStyling(container, component, parent);
        }
        this._inStyling = false;
    }
    setNativeElement(element) {
        this.i.setNativeElement(element);
    }
    get remove() {
        return this._remove;
    }
    set remove(ev) {
        if (this._remove_wrapped !== null) {
            this.i.remove = delegateRemove(this.i.remove, this._remove_wrapped);
            this._remove_wrapped = null;
            this._remove = null;
        }
        this._remove = ev;
        this._remove_wrapped = (o, e) => {
            let outerArgs = new IgrComponentBoolValueChangedEventArgs();
            outerArgs._provideImplementation(e);
            if (this.beforeRemove) {
                this.beforeRemove(this, outerArgs);
            }
            if (this._remove) {
                this._remove(this, outerArgs);
            }
        };
        this.i.remove = delegateCombine(this.i.remove, this._remove_wrapped);
        if (this.i.removeChanged) {
            this.i.removeChanged();
        }
        ;
    }
    get select() {
        return this._select;
    }
    set select(ev) {
        if (this._select_wrapped !== null) {
            this.i.select = delegateRemove(this.i.select, this._select_wrapped);
            this._select_wrapped = null;
            this._select = null;
        }
        this._select = ev;
        this._select_wrapped = (o, e) => {
            let outerArgs = new IgrComponentBoolValueChangedEventArgs();
            outerArgs._provideImplementation(e);
            if (this.beforeSelect) {
                this.beforeSelect(this, outerArgs);
            }
            if (this._select) {
                this._select(this, outerArgs);
            }
        };
        this.i.select = delegateCombine(this.i.select, this._select_wrapped);
        if (this.i.selectChanged) {
            this.i.selectChanged();
        }
        ;
    }
}
