// igniteui-react/esm2015/lib/igr-tab-group-pane.js
import * as React from 'react';
import { DockManagerPaneType_$type } from "./DockManagerPaneType";
import { IgrContentPaneCollection } from "./igr-content-pane-collection";
import { IgrDockManagerPane } from "./igr-dock-manager-pane";
import { TabGroupPane } from "./TabGroupPane";
import { defineCustomElements } from "igniteui-dockmanager/loader";
import { isValidProp, ensureEnum, ensureBool } from "igniteui-react-core";
import { ContentPaneCollection as ContentPaneCollection_internal } from "./ContentPaneCollection";
import { ContentPane } from "./ContentPane";
import { SyncableObservableCollection$1 } from "igniteui-react-core";
export class IgrTabGroupPane extends IgrDockManagerPane {
    createImplementation() {
        return new TabGroupPane();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor(props) {
        super(props);
        this._panes = null;
        this._getMainRef = this._getMainRef.bind(this);
        defineCustomElements();
    }
    componentDidMount() {
        this.i.nativeElement = this._elRef;
        this.mounted = true;
        super.componentDidMount();
    }
    render() {
        const nativePropsName = Object.keys(this.props).filter(prop => !isValidProp(this, prop) && prop !== "originalRef");
        const nativeProps = {};
        nativePropsName.forEach(propName => {
            nativeProps[propName] = this.props[propName];
        });
        let propChildren = this.props.children;
        let children = [];
        React.Children.forEach(propChildren, (ch) => {
            children.push(React.cloneElement(ch));
        });
        this._portalManager.onRender(children);
        let style = {};
        if (this.props.style) {
            style = this.props.style;
        }
        let div = React.createElement("", Object.assign(Object.assign({}, nativeProps), { ref: this._getMainRef, class: this.props.className, style: style, children: children }));
        return div;
    }
    _getMainRef(ref) {
        this._elRef = ref;
    }
    /**
     * The id of the pane. If not set the Dock Manager generates it automatically.
    */
    get id() {
        return this.i.o;
    }
    set id(v) {
        this.i.o = v;
    }
    /**
     * The type of the pane.
    */
    get type() {
        return this.i.f;
    }
    set type(v) {
        this.i.f = ensureEnum(DockManagerPaneType_$type, v);
    }
    /**
     * The child content panes of the tab group.
    */
    get panes() {
        if (this._panes === null) {
            let coll = new IgrContentPaneCollection();
            let innerColl = this.i.e;
            if (!innerColl) {
                innerColl = new ContentPaneCollection_internal();
            }
            this._panes = coll._fromInner(innerColl);
        }
        return this._panes;
    }
    set panes(v) {
        if (this._panes !== null) {
            this._panes._setSyncTarget(null);
            this._panes = null;
        }
        let coll = new IgrContentPaneCollection();
        this._panes = coll._fromOuter(v);
        let syncColl = new SyncableObservableCollection$1(ContentPane.$type);
        let innerColl = this.i.e;
        if (!innerColl) {
            innerColl = new ContentPaneCollection_internal();
        }
        syncColl._inner = innerColl;
        syncColl.clear();
        this._panes._setSyncTarget(syncColl);
    }
    /**
     * The size of the pane relative to its sibling panes' sizes. Defaults to 100.
    */
    get size() {
        return this.i.l;
    }
    set size(v) {
        this.i.l = +v;
    }
    /**
     * The index of the selected tab.
    */
    get selectedIndex() {
        return this.i.k;
    }
    set selectedIndex(v) {
        this.i.k = +v;
    }
    /**
     * Determines whether the pane should present in the UI when empty.
    */
    get allowEmpty() {
        return this.i.i;
    }
    set allowEmpty(v) {
        this.i.i = ensureBool(v);
    }
    /**
     * Determines whether a tab group is maximized or not. Defaults to false.
    */
    get isMaximized() {
        return this.i.j;
    }
    set isMaximized(v) {
        this.i.j = ensureBool(v);
    }
    findByName(name) {
        var baseResult = super.findByName(name);
        if (baseResult) {
            return baseResult;
        }
        if (this.panes != null && this.panes.findByName && this.panes.findByName(name)) {
            return this.panes.findByName(name);
        }
        return null;
    }
    setNativeElement(element) {
        this.i.setNativeElement(element);
    }
}
