// igniteui-react/esm2015/lib/igr-float-pane-action.js
import { PaneDragActionType_$type } from "./PaneDragActionType";
import { IgrDockManagerPoint } from "./igr-dock-manager-point";
import { IgrPaneDragAction } from "./igr-pane-drag-action";
import { FloatPaneAction } from "./FloatPaneAction";
import { ensureEnum } from "igniteui-react-core";
export class IgrFloatPaneAction extends IgrPaneDragAction {
    createImplementation() {
        return new FloatPaneAction();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor(props) {
        super(props);
    }
    get type() {
        return this.i.b;
    }
    set type(v) {
        this.i.b = ensureEnum(PaneDragActionType_$type, v);
    }
    get location() {
        const r = this.i.a;
        if (r == null) {
            return null;
        }
        if (!r.externalObject) {
            let e = new IgrDockManagerPoint();
            if (r.$type) {
                e._implementation = r;
            }
            else {
                if (e.i.setNativeElement) {
                    e.i.setNativeElement(r);
                }
            }
            r.externalObject = e;
        }
        return r.externalObject;
    }
    set location(v) {
        v == null ? this.i.a = null : this.i.a = v.i;
    }
    get width() {
        return this.i.f;
    }
    set width(v) {
        this.i.f = +v;
    }
    get height() {
        return this.i.e;
    }
    set height(v) {
        this.i.e = +v;
    }
    findByName(name) {
        var baseResult = super.findByName(name);
        if (baseResult) {
            return baseResult;
        }
        if (this.location && this.location.name && this.location.name == name) {
            return this.location;
        }
        return null;
    }
    setNativeElement(element) {
        this.i.setNativeElement(element);
    }
}
