// igniteui-react/esm2015/lib/igr-dock-manager-resource-strings.js
import { DockManagerResourceStrings as DockManagerResourceStrings_internal } from "./DockManagerResourceStrings";
/**
 * Describes dock manager resource strings.
*/
export class IgrDockManagerResourceStrings {
    createImplementation() {
        return new DockManagerResourceStrings_internal();
    }
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    get close() {
        return this.i.e;
    }
    set close(v) {
        this.i.e = v;
    }
    get pin() {
        return this.i.l;
    }
    set pin(v) {
        this.i.l = v;
    }
    get unpin() {
        return this.i.m;
    }
    set unpin(v) {
        this.i.m = v;
    }
    get maximize() {
        return this.i.g;
    }
    set maximize(v) {
        this.i.g = v;
    }
    get minimize() {
        return this.i.h;
    }
    set minimize(v) {
        this.i.h = v;
    }
    get moreOptions() {
        return this.i.i;
    }
    set moreOptions(v) {
        this.i.i = v;
    }
    get moreTabs() {
        return this.i.j;
    }
    set moreTabs(v) {
        this.i.j = v;
    }
    get panes() {
        return this.i.k;
    }
    set panes(v) {
        this.i.k = v;
    }
    get documents() {
        return this.i.f;
    }
    set documents(v) {
        this.i.f = v;
    }
    findByName(name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        return null;
    }
    setNativeElement(element) {
        this.i.setNativeElement(element);
    }
}
