// igniteui-react/esm2015/lib/igr-calendar-date.js
import { CalendarDate as CalendarDate_internal } from "./CalendarDate";
import { ensureBool } from "igniteui-react-core";
export class IgrCalendarDate {
    createImplementation() {
        return new CalendarDate_internal();
    }
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    get date() {
        return this.i.d;
    }
    set date(v) {
        this.i.d = v;
    }
    get isCurrentMonth() {
        return this.i.a;
    }
    set isCurrentMonth(v) {
        this.i.a = ensureBool(v);
    }
    get isPrevMonth() {
        return this.i.c;
    }
    set isPrevMonth(v) {
        this.i.c = ensureBool(v);
    }
    get isNextMonth() {
        return this.i.b;
    }
    set isNextMonth(v) {
        this.i.b = ensureBool(v);
    }
    findByName(name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        return null;
    }
}
