// igniteui-react/esm2015/lib/igr-docking-indicator.js
import { DockingIndicatorPosition_$type } from "./DockingIndicatorPosition";
import { DockingIndicator as DockingIndicator_internal } from "./DockingIndicator";
import { ensureEnum, ensureBool } from "igniteui-react-core";
export class IgrDockingIndicator {
    createImplementation() {
        return new DockingIndicator_internal();
    }
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    /**
     * Gets the position of the docking indicator.
    */
    get position() {
        return this.i.a;
    }
    set position(v) {
        this.i.a = ensureEnum(DockingIndicatorPosition_$type, v);
    }
    /**
     * Gets a value indicating whether the docking indicator is a root one.
    */
    get isRoot() {
        return this.i.d;
    }
    set isRoot(v) {
        this.i.d = ensureBool(v);
    }
    /**
     * Gets a value indicating the direction of the docking indicator.
    */
    get direction() {
        return this.i.g;
    }
    set direction(v) {
        this.i.g = v;
    }
    findByName(name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        return null;
    }
    setNativeElement(element) {
        this.i.setNativeElement(element);
    }
}
