// igniteui-react/esm2015/lib/igr-range-slider-value.js
import { RangeSliderValue as RangeSliderValue_internal } from "./RangeSliderValue";
export class IgrRangeSliderValue {
    createImplementation() {
        return new RangeSliderValue_internal();
    }
    get nativeElement() {
        return this._implementation.nativeElement;
    }
    /**
     * @hidden
     */
    get i() {
        return this._implementation;
    }
    onImplementationCreated() {
    }
    constructor() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    _provideImplementation(i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    get lower() {
        return this.i.a;
    }
    set lower(v) {
        this.i.a = +v;
    }
    get upper() {
        return this.i.b;
    }
    set upper(v) {
        this.i.b = +v;
    }
    findByName(name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        return null;
    }
}
