// igniteui-react/esm2015/lib/igr-linear-progress.js
import * as React from 'react';
import { LinearProgressLabelAlign_$type } from "./LinearProgressLabelAlign";
import { IgrProgressBase } from "./igr-progress-base";
import { LinearProgress } from "./LinearProgress";
import { IgcLinearProgressComponent, defineComponents } from "igniteui-webcomponents";
import { isValidProp, ensureBool, ensureEnum } from "igniteui-react-core";
/**
 * A linear progress indicator used to express unspecified wait time or display
 * the length of a process.
*/
export class IgrLinearProgress extends IgrProgressBase {
    createImplementation() {
        return new LinearProgress();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor(props) {
        super(props);
        this._getMainRef = this._getMainRef.bind(this);
        defineComponents(IgcLinearProgressComponent);
    }
    componentDidMount() {
        this.i.nativeElement = this._elRef;
        this.mounted = true;
        super.componentDidMount();
    }
    render() {
        const nativePropsName = Object.keys(this.props).filter(prop => !isValidProp(this, prop) && prop !== "originalRef");
        const nativeProps = {};
        nativePropsName.forEach(propName => {
            nativeProps[propName] = this.props[propName];
        });
        let propChildren = this.props.children;
        let children = [];
        React.Children.forEach(propChildren, (ch) => {
            children.push(React.cloneElement(ch));
        });
        this._portalManager.onRender(children);
        let style = {};
        if (this.props.style) {
            style = this.props.style;
        }
        let div = React.createElement("igc-linear-progress", Object.assign(Object.assign({}, nativeProps), { ref: this._getMainRef, id: this.props.id, class: this.props.className, style: style, children: children }));
        return div;
    }
    _getMainRef(ref) {
        this._elRef = ref;
    }
    /**
     * Sets the striped look of the control.
    */
    get striped() {
        return this.i.o;
    }
    set striped(v) {
        this.i.o = ensureBool(v);
    }
    /**
     * The position for the default label of the control.
    */
    get labelAlign() {
        return this.i.n;
    }
    set labelAlign(v) {
        this.i.n = ensureEnum(LinearProgressLabelAlign_$type, v);
    }
}
