// igniteui-react/esm2015/lib/MoveTabAction.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { PaneDragAction } from "./PaneDragAction";
import { WCNativeHelper } from "./WCNativeHelper";
import { enumGetBox, EnumUtil, markType } from "igniteui-react-core";
import { PaneDragActionType_$type } from "./PaneDragActionType";
/**
 * @hidden
 */
export let MoveTabAction = /*@__PURE__*/ (() => {
    class MoveTabAction extends PaneDragAction {
        constructor() {
            super(...arguments);
            this.b = new WCNativeHelper();
            this.f = null;
        }
        get c() {
            return this.b;
        }
        get nativeElement() {
            return this.f;
        }
        set nativeElement(a) {
            this.f = a;
            this.b.o = this.f;
        }
        setNativeElement(a) {
            this.nativeElement = a;
        }
        get a() {
            let ret_ = this.c.n("type");
            if ((ret_ === undefined && this.___rawMode)) {
                return undefined;
            }
            ret_ = enumGetBox(PaneDragActionType_$type, EnumUtil.getEnumValue(PaneDragActionType_$type, this.c.f(PaneDragActionType_$type, ret_)));
            return EnumUtil.getEnumValue(PaneDragActionType_$type, ret_);
        }
        set a(a) {
            let value_ = enumGetBox(PaneDragActionType_$type, a);
            value_ = WCNativeHelper.camel(value_.toString());
            this.c.w("type", value_);
        }
        get e() {
            let ret_ = this.c.n("oldIndex");
            return ret_;
        }
        set e(a) {
            let value_ = a;
            this.c.w("oldIndex", value_);
        }
        get d() {
            let ret_ = this.c.n("newIndex");
            return ret_;
        }
        set d(a) {
            let value_ = a;
            this.c.w("newIndex", value_);
        }
    }
    MoveTabAction.$t = /*@__PURE__*/ markType(MoveTabAction, 'MoveTabAction', PaneDragAction.$);
    return MoveTabAction;
})();
