// igniteui-react/esm2015/lib/NavDrawerItem.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { BaseElement } from "./BaseElement";
import { WCNativeHelper } from "./WCNativeHelper";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let NavDrawerItem = /*@__PURE__*/ (() => {
    class NavDrawerItem extends BaseElement {
        constructor() {
            super(...arguments);
            this.a = new WCNativeHelper();
            this.e = null;
        }
        get b() {
            return this.a;
        }
        get nativeElement() {
            return this.e;
        }
        set nativeElement(a) {
            this.e = a;
            this.a.o = this.e;
        }
        setNativeElement(a) {
            this.nativeElement = a;
        }
        get d() {
            let ret_ = this.b.n("disabled");
            return ret_;
        }
        set d(a) {
            let value_ = a;
            this.b.w("disabled", value_);
        }
        get c() {
            let ret_ = this.b.n("active");
            return ret_;
        }
        set c(a) {
            let value_ = a;
            this.b.w("active", value_);
        }
    }
    NavDrawerItem.$t = /*@__PURE__*/ markType(NavDrawerItem, 'NavDrawerItem', BaseElement.$);
    return NavDrawerItem;
})();
