// igniteui-react/esm2015/lib/NavDrawerPosition.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
export var NavDrawerPosition = /*@__PURE__*/ (function (NavDrawerPosition) {
    NavDrawerPosition[NavDrawerPosition["Bottom"] = 0] = "Bottom";
    NavDrawerPosition[NavDrawerPosition["Top"] = 1] = "Top";
    NavDrawerPosition[NavDrawerPosition["Start"] = 2] = "Start";
    NavDrawerPosition[NavDrawerPosition["End"] = 3] = "End";
    NavDrawerPosition[NavDrawerPosition["Relative"] = 4] = "Relative";
    return NavDrawerPosition;
})({});
/**
 * @hidden
 */
export let NavDrawerPosition_$type = /*@__PURE__*/ markEnum('NavDrawerPosition', 'Bottom,0|Top,1|Start,2|End,3|Relative,4');
