// igniteui-react/esm2015/lib/NavDrawer.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { BaseElement } from "./BaseElement";
import { WCNativeHelper } from "./WCNativeHelper";
import { enumGetBox, EnumUtil, markType } from "igniteui-react-core";
import { NavDrawerPosition_$type } from "./NavDrawerPosition";
/**
 * @hidden
 */
export let NavDrawer = /*@__PURE__*/ (() => {
    class NavDrawer extends BaseElement {
        constructor() {
            super(...arguments);
            this.b = new WCNativeHelper();
            this.h = null;
        }
        get c() {
            return this.b;
        }
        get nativeElement() {
            return this.h;
        }
        set nativeElement(a) {
            this.h = a;
            this.b.o = this.h;
        }
        setNativeElement(a) {
            this.nativeElement = a;
        }
        get a() {
            let ret_ = this.c.n("position");
            if ((ret_ === undefined && this.___rawMode)) {
                return undefined;
            }
            ret_ = enumGetBox(NavDrawerPosition_$type, EnumUtil.getEnumValue(NavDrawerPosition_$type, this.c.f(NavDrawerPosition_$type, ret_)));
            return EnumUtil.getEnumValue(NavDrawerPosition_$type, ret_);
        }
        set a(a) {
            let value_ = enumGetBox(NavDrawerPosition_$type, a);
            value_ = WCNativeHelper.camel(value_.toString());
            this.c.w("position", value_);
        }
        get e() {
            let ret_ = this.c.n("open");
            return ret_;
        }
        set e(a) {
            let value_ = a;
            this.c.w("open", value_);
        }
        f() {
            return (this.nativeElement.show());
        }
        d() {
            return (this.nativeElement.hide());
        }
        g() {
            return (this.nativeElement.toggle());
        }
    }
    NavDrawer.$t = /*@__PURE__*/ markType(NavDrawer, 'NavDrawer', BaseElement.$);
    return NavDrawer;
})();
