// igniteui-react/esm2015/lib/PaneScrollEventArgsDetail.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { WCNativeHelper } from "./WCNativeHelper";
import { ContentPane } from "./ContentPane";
/**
 * @hidden
 */
export let PaneScrollEventArgsDetail = /*@__PURE__*/ (() => {
    class PaneScrollEventArgsDetail extends Base {
        constructor() {
            super(...arguments);
            this.b = new WCNativeHelper();
            this.d = null;
        }
        get c() {
            return this.b;
        }
        get nativeElement() {
            return this.d;
        }
        set nativeElement(a) {
            this.d = a;
            this.b.o = this.d;
        }
        setNativeElement(a) {
            this.nativeElement = a;
        }
        get a() {
            let ret_ = this.c.m("pane", (a) => new ContentPane());
            return ret_;
        }
        set a(a) {
            let value_ = a;
            this.c.w("pane", value_);
        }
    }
    PaneScrollEventArgsDetail.$t = /*@__PURE__*/ markType(PaneScrollEventArgsDetail, 'PaneScrollEventArgsDetail');
    return PaneScrollEventArgsDetail;
})();
