// igniteui-react/esm2015/lib/PaneDragActionType.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
export var PaneDragActionType = /*@__PURE__*/ (function (PaneDragActionType) {
    PaneDragActionType[PaneDragActionType["FloatPane"] = 0] = "FloatPane";
    PaneDragActionType[PaneDragActionType["MoveFloatingPane"] = 1] = "MoveFloatingPane";
    PaneDragActionType[PaneDragActionType["DockPane"] = 2] = "DockPane";
    PaneDragActionType[PaneDragActionType["MoveTab"] = 3] = "MoveTab";
    return PaneDragActionType;
})({});
/**
 * @hidden
 */
export let PaneDragActionType_$type = /*@__PURE__*/ markEnum('PaneDragActionType', 'FloatPane,0|MoveFloatingPane,1|DockPane,2|MoveTab,3');
