// igniteui-react/esm2015/lib/PaneDragOverEventArgsDetail.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { WCNativeHelper } from "./WCNativeHelper";
import { DockManagerPane } from "./DockManagerPane";
import { PaneDragAction } from "./PaneDragAction";
/**
 * @hidden
 */
export let PaneDragOverEventArgsDetail = /*@__PURE__*/ (() => {
    class PaneDragOverEventArgsDetail extends Base {
        constructor() {
            super(...arguments);
            this.d = new WCNativeHelper();
            this.g = null;
        }
        get e() {
            return this.d;
        }
        get nativeElement() {
            return this.g;
        }
        set nativeElement(a) {
            this.g = a;
            this.d.o = this.g;
        }
        setNativeElement(a) {
            this.nativeElement = a;
        }
        get b() {
            let ret_ = this.e.m("sourcePane", (a) => DockManagerPane.createFromNative(a));
            return ret_;
        }
        set b(a) {
            let value_ = a;
            this.e.w("sourcePane", value_);
        }
        get a() {
            let ret_ = this.e.n("panes");
            return ret_;
        }
        set a(a) {
            let value_ = a;
            this.e.w("panes", value_);
        }
        get c() {
            let ret_ = this.e.m("action", (a) => new PaneDragAction());
            return ret_;
        }
        set c(a) {
            let value_ = a;
            this.e.w("action", value_);
        }
        get f() {
            let ret_ = this.e.n("isValid");
            return ret_;
        }
        set f(a) {
            let value_ = a;
            this.e.w("isValid", value_);
        }
    }
    PaneDragOverEventArgsDetail.$t = /*@__PURE__*/ markType(PaneDragOverEventArgsDetail, 'PaneDragOverEventArgsDetail');
    return PaneDragOverEventArgsDetail;
})();
