// igniteui-react/esm2015/lib/PaneCloseEventArgsDetail.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { WCNativeHelper } from "./WCNativeHelper";
import { DockManagerPane } from "./DockManagerPane";
/**
 * @hidden
 */
export let PaneCloseEventArgsDetail = /*@__PURE__*/ (() => {
    class PaneCloseEventArgsDetail extends Base {
        constructor() {
            super(...arguments);
            this.c = new WCNativeHelper();
            this.e = null;
        }
        get d() {
            return this.c;
        }
        get nativeElement() {
            return this.e;
        }
        set nativeElement(a) {
            this.e = a;
            this.c.o = this.e;
        }
        setNativeElement(a) {
            this.nativeElement = a;
        }
        get b() {
            let ret_ = this.d.m("sourcePane", (a) => DockManagerPane.createFromNative(a));
            return ret_;
        }
        set b(a) {
            let value_ = a;
            this.d.w("sourcePane", value_);
        }
        get a() {
            let ret_ = this.d.n("panes");
            return ret_;
        }
        set a(a) {
            let value_ = a;
            this.d.w("panes", value_);
        }
    }
    PaneCloseEventArgsDetail.$t = /*@__PURE__*/ markType(PaneCloseEventArgsDetail, 'PaneCloseEventArgsDetail');
    return PaneCloseEventArgsDetail;
})();
