// igniteui-react/esm2015/lib/PanePinnedEventArgsDetail.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, enumGetBox, EnumUtil, markType } from "igniteui-react-core";
import { WCNativeHelper } from "./WCNativeHelper";
import { ContentPane } from "./ContentPane";
import { UnpinnedLocation_$type } from "./UnpinnedLocation";
/**
 * @hidden
 */
export let PanePinnedEventArgsDetail = /*@__PURE__*/ (() => {
    class PanePinnedEventArgsDetail extends Base {
        constructor() {
            super(...arguments);
            this.d = new WCNativeHelper();
            this.g = null;
        }
        get e() {
            return this.d;
        }
        get nativeElement() {
            return this.g;
        }
        set nativeElement(a) {
            this.g = a;
            this.d.o = this.g;
        }
        setNativeElement(a) {
            this.nativeElement = a;
        }
        get a() {
            let ret_ = this.e.m("sourcePane", (a) => new ContentPane());
            return ret_;
        }
        set a(a) {
            let value_ = a;
            this.e.w("sourcePane", value_);
        }
        get b() {
            let ret_ = this.e.n("panes");
            return ret_;
        }
        set b(a) {
            let value_ = a;
            this.e.w("panes", value_);
        }
        get f() {
            let ret_ = this.e.n("newValue");
            return ret_;
        }
        set f(a) {
            let value_ = a;
            this.e.w("newValue", value_);
        }
        get c() {
            let ret_ = this.e.n("location");
            if ((ret_ === undefined && this.___rawMode)) {
                return undefined;
            }
            ret_ = enumGetBox(UnpinnedLocation_$type, EnumUtil.getEnumValue(UnpinnedLocation_$type, this.e.f(UnpinnedLocation_$type, ret_)));
            return EnumUtil.getEnumValue(UnpinnedLocation_$type, ret_);
        }
        set c(a) {
            let value_ = enumGetBox(UnpinnedLocation_$type, a);
            value_ = WCNativeHelper.camel(value_.toString());
            this.e.w("location", value_);
        }
    }
    PanePinnedEventArgsDetail.$t = /*@__PURE__*/ markType(PanePinnedEventArgsDetail, 'PanePinnedEventArgsDetail');
    return PanePinnedEventArgsDetail;
})();
