// igniteui-react/esm2015/lib/ProgressBase.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { BaseElement } from "./BaseElement";
import { WCNativeHelper } from "./WCNativeHelper";
import { enumGetBox, EnumUtil, markType } from "igniteui-react-core";
import { StyleVariant_$type } from "./StyleVariant";
/**
 * @hidden
 */
export let ProgressBase = /*@__PURE__*/ (() => {
    class ProgressBase extends BaseElement {
        constructor() {
            super(...arguments);
            this.b = new WCNativeHelper();
            this.i = null;
        }
        get c() {
            return this.b;
        }
        get nativeElement() {
            return this.i;
        }
        set nativeElement(a) {
            this.i = a;
            this.b.o = this.i;
        }
        setNativeElement(a) {
            this.nativeElement = a;
        }
        get g() {
            let ret_ = this.c.n("max");
            return ret_;
        }
        set g(a) {
            let value_ = a;
            this.c.w("max", value_);
        }
        get h() {
            let ret_ = this.c.n("value");
            return ret_;
        }
        set h(a) {
            let value_ = a;
            this.c.w("value", value_);
        }
        get a() {
            let ret_ = this.c.n("variant");
            if ((ret_ === undefined && this.___rawMode)) {
                return undefined;
            }
            ret_ = enumGetBox(StyleVariant_$type, EnumUtil.getEnumValue(StyleVariant_$type, this.c.f(StyleVariant_$type, ret_)));
            return EnumUtil.getEnumValue(StyleVariant_$type, ret_);
        }
        set a(a) {
            let value_ = enumGetBox(StyleVariant_$type, a);
            value_ = WCNativeHelper.camel(value_.toString());
            this.c.w("variant", value_);
        }
        get f() {
            let ret_ = this.c.n("animationDuration");
            return ret_;
        }
        set f(a) {
            let value_ = a;
            this.c.w("animationDuration", value_);
        }
        get e() {
            let ret_ = this.c.n("indeterminate");
            return ret_;
        }
        set e(a) {
            let value_ = a;
            this.c.w("indeterminate", value_);
        }
        get d() {
            let ret_ = this.c.n("hideLabel");
            return ret_;
        }
        set d(a) {
            let value_ = a;
            this.c.w("hideLabel", value_);
        }
        get k() {
            let ret_ = this.c.n("labelFormat");
            return ret_;
        }
        set k(a) {
            let value_ = a;
            this.c.w("labelFormat", value_);
        }
        l() {
            this.nativeElement.connectedCallback();
        }
    }
    ProgressBase.$t = /*@__PURE__*/ markType(ProgressBase, 'ProgressBase', BaseElement.$);
    return ProgressBase;
})();
