// igniteui-react/esm2015/lib/RadioGroup.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { BaseElement } from "./BaseElement";
import { WCNativeHelper } from "./WCNativeHelper";
import { enumGetBox, EnumUtil, markType } from "igniteui-react-core";
import { RadioGroupAlignment_$type } from "./RadioGroupAlignment";
import { RadioChangeEventArgs } from "./RadioChangeEventArgs";
/**
 * @hidden
 */
export let RadioGroup = /*@__PURE__*/ (() => {
    class RadioGroup extends BaseElement {
        constructor() {
            super(...arguments);
            this.change = null;
            this.c = new WCNativeHelper();
            this.e = null;
        }
        changeChanged() {
            if (this.change != null) {
                this.d.s("change", (a) => this.change(this, RadioChangeEventArgs.a(a)));
            }
            else {
                this.d.v("change");
            }
        }
        get d() {
            return this.c;
        }
        get nativeElement() {
            return this.e;
        }
        set nativeElement(a) {
            this.e = a;
            this.c.o = this.e;
        }
        setNativeElement(a) {
            this.nativeElement = a;
        }
        get b() {
            let ret_ = this.d.n("alignment");
            if ((ret_ === undefined && this.___rawMode)) {
                return undefined;
            }
            ret_ = enumGetBox(RadioGroupAlignment_$type, EnumUtil.getEnumValue(RadioGroupAlignment_$type, this.d.f(RadioGroupAlignment_$type, ret_)));
            return EnumUtil.getEnumValue(RadioGroupAlignment_$type, ret_);
        }
        set b(a) {
            let value_ = enumGetBox(RadioGroupAlignment_$type, a);
            value_ = WCNativeHelper.camel(value_.toString());
            this.d.w("alignment", value_);
        }
        get g() {
            let ret_ = this.d.n("name");
            return ret_;
        }
        set g(a) {
            let value_ = a;
            this.d.w("name", value_);
        }
        get h() {
            let ret_ = this.d.n("value");
            return ret_;
        }
        set h(a) {
            let value_ = a;
            this.d.w("value", value_);
        }
    }
    RadioGroup.$t = /*@__PURE__*/ markType(RadioGroup, 'RadioGroup', BaseElement.$);
    return RadioGroup;
})();
