// igniteui-react/esm2015/lib/RadioChangeEventArgsDetail.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { Base, markType } from "igniteui-react-core";
import { WCNativeHelper } from "./WCNativeHelper";
/**
 * @hidden
 */
export let RadioChangeEventArgsDetail = /*@__PURE__*/ (() => {
    class RadioChangeEventArgsDetail extends Base {
        constructor() {
            super(...arguments);
            this.a = new WCNativeHelper();
            this.d = null;
        }
        get b() {
            return this.a;
        }
        get nativeElement() {
            return this.d;
        }
        set nativeElement(a) {
            this.d = a;
            this.a.o = this.d;
        }
        setNativeElement(a) {
            this.nativeElement = a;
        }
        get c() {
            let ret_ = this.b.n("checked");
            return ret_;
        }
        set c(a) {
            let value_ = a;
            this.b.w("checked", value_);
        }
        get f() {
            let ret_ = this.b.n("value");
            return ret_;
        }
        set f(a) {
            let value_ = a;
            this.b.w("value", value_);
        }
    }
    RadioChangeEventArgsDetail.$t = /*@__PURE__*/ markType(RadioChangeEventArgsDetail, 'RadioChangeEventArgsDetail');
    return RadioChangeEventArgsDetail;
})();
