// igniteui-react/esm2015/lib/RangeSlider.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { SliderBase } from "./SliderBase";
import { markType } from "igniteui-react-core";
import { RangeSliderValueEventArgs } from "./RangeSliderValueEventArgs";
/**
 * @hidden
 */
export let RangeSlider = /*@__PURE__*/ (() => {
    class RangeSlider extends SliderBase {
        constructor() {
            super(...arguments);
            this.input = null;
            this.change = null;
        }
        get z() {
            let ret_ = this.e.n("lower");
            return ret_;
        }
        set z(a) {
            let value_ = a;
            this.e.w("lower", value_);
        }
        get aa() {
            let ret_ = this.e.n("upper");
            return ret_;
        }
        set aa(a) {
            let value_ = a;
            this.e.w("upper", value_);
        }
        get ab() {
            let ret_ = this.e.n("thumbLabelLower");
            return ret_;
        }
        set ab(a) {
            let value_ = a;
            this.e.w("thumbLabelLower", value_);
        }
        get ac() {
            let ret_ = this.e.n("thumbLabelUpper");
            return ret_;
        }
        set ac(a) {
            let value_ = a;
            this.e.w("thumbLabelUpper", value_);
        }
        inputChanged() {
            if (this.input != null) {
                this.e.s("input", (a) => this.input(this, RangeSliderValueEventArgs.b(a)));
            }
            else {
                this.e.v("input");
            }
        }
        changeChanged() {
            if (this.change != null) {
                this.e.s("change", (a) => this.change(this, RangeSliderValueEventArgs.b(a)));
            }
            else {
                this.e.v("change");
            }
        }
    }
    RangeSlider.$t = /*@__PURE__*/ markType(RangeSlider, 'RangeSlider', SliderBase.$);
    return RangeSlider;
})();
