// igniteui-react/esm2015/lib/SelectGroup.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { BaseElement } from "./BaseElement";
import { WCNativeHelper } from "./WCNativeHelper";
import { markType } from "igniteui-react-core";
import { SelectItem } from "./SelectItem";
/**
 * @hidden
 */
export let SelectGroup = /*@__PURE__*/ (() => {
    class SelectGroup extends BaseElement {
        constructor() {
            super(...arguments);
            this.b = new WCNativeHelper();
            this.e = null;
        }
        get c() {
            return this.b;
        }
        get nativeElement() {
            return this.e;
        }
        set nativeElement(a) {
            this.e = a;
            this.b.o = this.e;
        }
        setNativeElement(a) {
            this.nativeElement = a;
        }
        get a() {
            let ret_ = this.c.m("items", (a) => new SelectItem());
            return ret_;
        }
        set a(a) {
            let value_ = a;
            this.c.w("items", value_);
        }
        get d() {
            let ret_ = this.c.n("disabled");
            return ret_;
        }
        set d(a) {
            let value_ = a;
            this.c.w("disabled", value_);
        }
    }
    SelectGroup.$t = /*@__PURE__*/ markType(SelectGroup, 'SelectGroup', BaseElement.$);
    return SelectGroup;
})();
