// igniteui-react/esm2015/lib/StepperTitlePosition.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
export var StepperTitlePosition = /*@__PURE__*/ (function (StepperTitlePosition) {
    StepperTitlePosition[StepperTitlePosition["Bottom"] = 0] = "Bottom";
    StepperTitlePosition[StepperTitlePosition["Top"] = 1] = "Top";
    StepperTitlePosition[StepperTitlePosition["Start"] = 2] = "Start";
    StepperTitlePosition[StepperTitlePosition["End"] = 3] = "End";
    return StepperTitlePosition;
})({});
/**
 * @hidden
 */
export let StepperTitlePosition_$type = /*@__PURE__*/ markEnum('StepperTitlePosition', 'Bottom,0|Top,1|Start,2|End,3');
