// igniteui-react/esm2015/lib/TabHeaderElement.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { BaseElement } from "./BaseElement";
import { WCNativeHelper } from "./WCNativeHelper";
import { markType } from "igniteui-react-core";
import { DragService } from "./DragService";
/**
 * @hidden
 */
export let TabHeaderElement = /*@__PURE__*/ (() => {
    class TabHeaderElement extends BaseElement {
        constructor() {
            super(...arguments);
            this.b = new WCNativeHelper();
            this.d = null;
        }
        get c() {
            return this.b;
        }
        get nativeElement() {
            return this.d;
        }
        set nativeElement(a) {
            this.d = a;
            this.b.o = this.d;
        }
        setNativeElement(a) {
            this.nativeElement = a;
        }
        get a() {
            let ret_ = this.c.m("dragService", (a) => new DragService());
            return ret_;
        }
        set a(a) {
            let value_ = a;
            this.c.w("dragService", value_);
        }
    }
    TabHeaderElement.$t = /*@__PURE__*/ markType(TabHeaderElement, 'TabHeaderElement', BaseElement.$);
    return TabHeaderElement;
})();
