// igniteui-react/esm2015/lib/TextareaInputMode.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
export var TextareaInputMode = /*@__PURE__*/ (function (TextareaInputMode) {
    TextareaInputMode[TextareaInputMode["None"] = 0] = "None";
    TextareaInputMode[TextareaInputMode["Email"] = 1] = "Email";
    TextareaInputMode[TextareaInputMode["Search"] = 2] = "Search";
    TextareaInputMode[TextareaInputMode["Tel"] = 3] = "Tel";
    TextareaInputMode[TextareaInputMode["Text"] = 4] = "Text";
    TextareaInputMode[TextareaInputMode["Url"] = 5] = "Url";
    TextareaInputMode[TextareaInputMode["Decimal"] = 6] = "Decimal";
    TextareaInputMode[TextareaInputMode["Numeric"] = 7] = "Numeric";
    return TextareaInputMode;
})({});
/**
 * @hidden
 */
export let TextareaInputMode_$type = /*@__PURE__*/ markEnum('TextareaInputMode', 'None,0|Email,1|Search,2|Tel,3|Text,4|Url,5|Decimal,6|Numeric,7');
