// igniteui-react/esm2015/lib/ToggleButton.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { BaseElement } from "./BaseElement";
import { WCNativeHelper } from "./WCNativeHelper";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
export let ToggleButton = /*@__PURE__*/ (() => {
    class ToggleButton extends BaseElement {
        constructor() {
            super(...arguments);
            this.a = new WCNativeHelper();
            this.e = null;
        }
        get b() {
            return this.a;
        }
        get nativeElement() {
            return this.e;
        }
        set nativeElement(a) {
            this.e = a;
            this.a.o = this.e;
        }
        setNativeElement(a) {
            this.nativeElement = a;
        }
        get g() {
            let ret_ = this.b.n("value");
            return ret_;
        }
        set g(a) {
            let value_ = a;
            this.b.w("value", value_);
        }
        get d() {
            let ret_ = this.b.n("selected");
            return ret_;
        }
        set d(a) {
            let value_ = a;
            this.b.w("selected", value_);
        }
        get c() {
            let ret_ = this.b.n("disabled");
            return ret_;
        }
        set c(a) {
            let value_ = a;
            this.b.w("disabled", value_);
        }
        j(a) {
            this.nativeElement.focus(this.b.i(a));
        }
        h() {
            this.nativeElement.blur();
        }
        i() {
            this.nativeElement.click();
        }
    }
    ToggleButton.$t = /*@__PURE__*/ markType(ToggleButton, 'ToggleButton', BaseElement.$);
    return ToggleButton;
})();
