// igniteui-webcomponents/components/calendar/base.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { LitElement } from 'lit';
import { property, state } from 'lit/decorators.js';
import { blazorDeepImport } from '../common/decorators/blazorDeepImport.js';
import { blazorIndirectRender } from '../common/decorators/blazorIndirectRender.js';
import { watch } from '../common/decorators/watch.js';
import { dateFromISOString, datesFromISOStrings, getWeekDayNumber, } from './helpers.js';
import { CalendarDay } from './model.js';
let IgcCalendarBaseComponent = class IgcCalendarBaseComponent extends LitElement {
    constructor() {
        super(...arguments);
        this._initialActiveDateSet = false;
        this._firstDayOfWeek = 0;
        this._activeDate = CalendarDay.today;
        this._values = [];
        this._specialDates = [];
        this._disabledDates = [];
        this.selection = 'single';
        this.showWeekNumbers = false;
        this.weekStart = 'sunday';
        this.locale = 'en';
    }
    get _hasValues() {
        return this._values.length > 0;
    }
    get _isSingle() {
        return this.selection === 'single';
    }
    get _isMultiple() {
        return this.selection === 'multiple';
    }
    get _isRange() {
        return this.selection === 'range';
    }
    get value() {
        return this._value ? this._value.native : undefined;
    }
    set value(value) {
        this._value = value ? CalendarDay.from(value) : undefined;
    }
    get values() {
        return this._values ? this._values.map((v) => v.native) : [];
    }
    set values(values) {
        this._values = values ? values.map((v) => CalendarDay.from(v)) : [];
    }
    get activeDate() {
        return this._activeDate.native;
    }
    set activeDate(value) {
        this._initialActiveDateSet = true;
        this._activeDate = value ? CalendarDay.from(value) : CalendarDay.today;
    }
    set specialDates(value) {
        this._specialDates = value ?? [];
    }
    get specialDates() {
        return this._specialDates.length ? this._specialDates : undefined;
    }
    set disabledDates(value) {
        this._disabledDates = value ?? [];
    }
    get disabledDates() {
        return this._disabledDates.length ? this._disabledDates : undefined;
    }
    weekStartChanged() {
        this._firstDayOfWeek = getWeekDayNumber(this.weekStart);
    }
    selectionChanged() {
        this._rangePreviewDate = undefined;
        this._value = undefined;
        this._values = [];
    }
    firstUpdated() {
        if (this._initialActiveDateSet) {
            return;
        }
        if (this._isSingle) {
            this.activeDate = this.value ?? this.activeDate;
        }
        else {
            this.activeDate = this.values[0] ?? this.activeDate;
        }
    }
};
__decorate([
    state()
], IgcCalendarBaseComponent.prototype, "_rangePreviewDate", void 0);
__decorate([
    state()
], IgcCalendarBaseComponent.prototype, "_firstDayOfWeek", void 0);
__decorate([
    state()
], IgcCalendarBaseComponent.prototype, "_activeDate", void 0);
__decorate([
    state()
], IgcCalendarBaseComponent.prototype, "_value", void 0);
__decorate([
    state()
], IgcCalendarBaseComponent.prototype, "_values", void 0);
__decorate([
    state()
], IgcCalendarBaseComponent.prototype, "_specialDates", void 0);
__decorate([
    state()
], IgcCalendarBaseComponent.prototype, "_disabledDates", void 0);
__decorate([
    property({ converter: dateFromISOString })
], IgcCalendarBaseComponent.prototype, "value", null);
__decorate([
    property({ converter: datesFromISOStrings })
], IgcCalendarBaseComponent.prototype, "values", null);
__decorate([
    property({ attribute: 'active-date', converter: dateFromISOString })
], IgcCalendarBaseComponent.prototype, "activeDate", null);
__decorate([
    property()
], IgcCalendarBaseComponent.prototype, "selection", void 0);
__decorate([
    property({ type: Boolean, reflect: true, attribute: 'show-week-numbers' })
], IgcCalendarBaseComponent.prototype, "showWeekNumbers", void 0);
__decorate([
    property({ attribute: 'week-start' })
], IgcCalendarBaseComponent.prototype, "weekStart", void 0);
__decorate([
    property()
], IgcCalendarBaseComponent.prototype, "locale", void 0);
__decorate([
    property({ attribute: false })
], IgcCalendarBaseComponent.prototype, "specialDates", null);
__decorate([
    property({ attribute: false })
], IgcCalendarBaseComponent.prototype, "disabledDates", null);
__decorate([
    watch('weekStart')
], IgcCalendarBaseComponent.prototype, "weekStartChanged", null);
__decorate([
    watch('selection', { waitUntilFirstUpdate: true })
], IgcCalendarBaseComponent.prototype, "selectionChanged", null);
IgcCalendarBaseComponent = __decorate([
    blazorIndirectRender,
    blazorDeepImport
], IgcCalendarBaseComponent);
export { IgcCalendarBaseComponent };
