// igniteui-react/esm5/lib/Accordion.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { BaseElement } from "./BaseElement";
import { WCNativeHelper } from "./WCNativeHelper";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
var Accordion = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(Accordion, _super);
    function Accordion() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.a = new WCNativeHelper();
        _this.d = null;
        return _this;
    }
    Object.defineProperty(Accordion.prototype, "b", {
        get: function () {
            return this.a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Accordion.prototype, "nativeElement", {
        get: function () {
            return this.d;
        },
        set: function (a) {
            this.d = a;
            this.a.o = this.d;
        },
        enumerable: false,
        configurable: true
    });
    Accordion.prototype.setNativeElement = function (a) {
        this.nativeElement = a;
    };
    Object.defineProperty(Accordion.prototype, "c", {
        get: function () {
            var ret_ = this.b.n("singleExpand");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.b.w("singleExpand", value_);
        },
        enumerable: false,
        configurable: true
    });
    Accordion.prototype.f = function () {
        this.nativeElement.hideAll();
    };
    Accordion.prototype.h = function () {
        this.nativeElement.showAll();
    };
    Accordion.$t = markType(Accordion, 'Accordion', BaseElement.$);
    return Accordion;
}(BaseElement));
export { Accordion };
