// igniteui-react/esm5/lib/Banner.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { BaseElement } from "./BaseElement";
import { WCNativeHelper } from "./WCNativeHelper";
import { markType } from "igniteui-react-core";
import { VoidEventArgs } from "./VoidEventArgs";
/**
 * @hidden
 */
var Banner = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(Banner, _super);
    function Banner() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.c = new WCNativeHelper();
        _this.i = null;
        _this.closing = null;
        _this.closed = null;
        return _this;
    }
    Object.defineProperty(Banner.prototype, "d", {
        get: function () {
            return this.c;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Banner.prototype, "nativeElement", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.c.o = this.i;
        },
        enumerable: false,
        configurable: true
    });
    Banner.prototype.setNativeElement = function (a) {
        this.nativeElement = a;
    };
    Object.defineProperty(Banner.prototype, "f", {
        get: function () {
            var ret_ = this.d.n("open");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.d.w("open", value_);
        },
        enumerable: false,
        configurable: true
    });
    Banner.prototype.closingChanged = function () {
        var _this = this;
        if (this.closing != null) {
            this.d.s("closing", function (a) { return _this.closing(_this, VoidEventArgs.b(a)); });
        }
        else {
            this.d.v("closing");
        }
    };
    Banner.prototype.closedChanged = function () {
        var _this = this;
        if (this.closed != null) {
            this.d.s("closed", function (a) { return _this.closed(_this, VoidEventArgs.b(a)); });
        }
        else {
            this.d.v("closed");
        }
    };
    Banner.prototype.g = function () {
        return (this.nativeElement.show());
    };
    Banner.prototype.e = function () {
        return (this.nativeElement.hide());
    };
    Banner.prototype.h = function () {
        return (this.nativeElement.toggle());
    };
    Banner.$t = markType(Banner, 'Banner', BaseElement.$);
    return Banner;
}(BaseElement));
export { Banner };
