// igniteui-react/esm5/lib/BaseComboBoxLike.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { BaseElement } from "./BaseElement";
import { WCNativeHelper } from "./WCNativeHelper";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
var BaseComboBoxLike = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(BaseComboBoxLike, _super);
    function BaseComboBoxLike() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.a = new WCNativeHelper();
        _this.i = null;
        return _this;
    }
    Object.defineProperty(BaseComboBoxLike.prototype, "b", {
        get: function () {
            return this.a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseComboBoxLike.prototype, "nativeElement", {
        get: function () {
            return this.i;
        },
        set: function (a) {
            this.i = a;
            this.a.o = this.i;
        },
        enumerable: false,
        configurable: true
    });
    BaseComboBoxLike.prototype.setNativeElement = function (a) {
        this.nativeElement = a;
    };
    Object.defineProperty(BaseComboBoxLike.prototype, "j", {
        get: function () {
            var ret_ = this.b.n("emitEvent");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.b.w("emitEvent", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseComboBoxLike.prototype, "e", {
        get: function () {
            var ret_ = this.b.n("keepOpenOnSelect");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.b.w("keepOpenOnSelect", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseComboBoxLike.prototype, "d", {
        get: function () {
            var ret_ = this.b.n("keepOpenOnOutsideClick");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.b.w("keepOpenOnOutsideClick", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseComboBoxLike.prototype, "f", {
        get: function () {
            var ret_ = this.b.n("open");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.b.w("open", value_);
        },
        enumerable: false,
        configurable: true
    });
    BaseComboBoxLike.prototype.g = function () {
        return (this.nativeElement.show());
    };
    BaseComboBoxLike.prototype.c = function () {
        return (this.nativeElement.hide());
    };
    BaseComboBoxLike.prototype.h = function () {
        return (this.nativeElement.toggle());
    };
    BaseComboBoxLike.$t = markType(BaseComboBoxLike, 'BaseComboBoxLike', BaseElement.$);
    return BaseComboBoxLike;
}(BaseElement));
export { BaseComboBoxLike };
