// igniteui-react/esm5/lib/BaseAlertLike.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { BaseElement } from "./BaseElement";
import { WCNativeHelper } from "./WCNativeHelper";
import { enumGetBox, EnumUtil, markType } from "igniteui-react-core";
import { BaseAlertLikePosition_$type } from "./BaseAlertLikePosition";
/**
 * @hidden
 */
var BaseAlertLike = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(BaseAlertLike, _super);
    function BaseAlertLike() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.b = new WCNativeHelper();
        _this.j = null;
        return _this;
    }
    Object.defineProperty(BaseAlertLike.prototype, "c", {
        get: function () {
            return this.b;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseAlertLike.prototype, "nativeElement", {
        get: function () {
            return this.j;
        },
        set: function (a) {
            this.j = a;
            this.b.o = this.j;
        },
        enumerable: false,
        configurable: true
    });
    BaseAlertLike.prototype.setNativeElement = function (a) {
        this.nativeElement = a;
    };
    Object.defineProperty(BaseAlertLike.prototype, "f", {
        get: function () {
            var ret_ = this.c.n("open");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.c.w("open", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseAlertLike.prototype, "i", {
        get: function () {
            var ret_ = this.c.n("displayTime");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.c.w("displayTime", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseAlertLike.prototype, "e", {
        get: function () {
            var ret_ = this.c.n("keepOpen");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.c.w("keepOpen", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseAlertLike.prototype, "a", {
        get: function () {
            var ret_ = this.c.n("position");
            if ((ret_ === undefined && this.___rawMode)) {
                return undefined;
            }
            ret_ = enumGetBox(BaseAlertLikePosition_$type, EnumUtil.getEnumValue(BaseAlertLikePosition_$type, this.c.f(BaseAlertLikePosition_$type, ret_)));
            return EnumUtil.getEnumValue(BaseAlertLikePosition_$type, ret_);
        },
        set: function (a) {
            var value_ = enumGetBox(BaseAlertLikePosition_$type, a);
            value_ = WCNativeHelper.camel(value_.toString());
            this.c.w("position", value_);
        },
        enumerable: false,
        configurable: true
    });
    BaseAlertLike.prototype.g = function () {
        return (this.nativeElement.show());
    };
    BaseAlertLike.prototype.d = function () {
        return (this.nativeElement.hide());
    };
    BaseAlertLike.prototype.h = function () {
        return (this.nativeElement.toggle());
    };
    BaseAlertLike.$t = markType(BaseAlertLike, 'BaseAlertLike', BaseElement.$);
    return BaseAlertLike;
}(BaseElement));
export { BaseAlertLike };
