// igniteui-react/esm5/lib/DockingIndicatorPosition.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
export var DockingIndicatorPosition = /*@__PURE__*/ (function (DockingIndicatorPosition) {
    DockingIndicatorPosition[DockingIndicatorPosition["Left"] = 0] = "Left";
    DockingIndicatorPosition[DockingIndicatorPosition["OuterLeft"] = 1] = "OuterLeft";
    DockingIndicatorPosition[DockingIndicatorPosition["Right"] = 2] = "Right";
    DockingIndicatorPosition[DockingIndicatorPosition["OuterRight"] = 3] = "OuterRight";
    DockingIndicatorPosition[DockingIndicatorPosition["Top"] = 4] = "Top";
    DockingIndicatorPosition[DockingIndicatorPosition["OuterTop"] = 5] = "OuterTop";
    DockingIndicatorPosition[DockingIndicatorPosition["Bottom"] = 6] = "Bottom";
    DockingIndicatorPosition[DockingIndicatorPosition["OuterBottom"] = 7] = "OuterBottom";
    DockingIndicatorPosition[DockingIndicatorPosition["Center"] = 8] = "Center";
    return DockingIndicatorPosition;
})({});
/**
 * @hidden
 */
export var DockingIndicatorPosition_$type = /*@__PURE__*/ markEnum('DockingIndicatorPosition', 'Left:left,0|OuterLeft:outerLeft,1|Right:right,2|OuterRight:outerRight,3|Top:top,4|OuterTop:outerTop,5|Bottom:bottom,6|OuterBottom:outerBottom,7|Center:center,8');
