// igniteui-react/esm5/lib/DockManagerPaneType.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
export var DockManagerPaneType = /*@__PURE__*/ (function (DockManagerPaneType) {
    DockManagerPaneType[DockManagerPaneType["SplitPane"] = 0] = "SplitPane";
    DockManagerPaneType[DockManagerPaneType["ContentPane"] = 1] = "ContentPane";
    DockManagerPaneType[DockManagerPaneType["TabGroupPane"] = 2] = "TabGroupPane";
    DockManagerPaneType[DockManagerPaneType["DocumentHost"] = 3] = "DocumentHost";
    return DockManagerPaneType;
})({});
/**
 * @hidden
 */
export var DockManagerPaneType_$type = /*@__PURE__*/ markEnum('DockManagerPaneType', 'SplitPane,0|ContentPane,1|TabGroupPane,2|DocumentHost,3');
