// igniteui-react/esm5/lib/Dropdown.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { BaseComboBoxLike } from "./BaseComboBoxLike";
import { DropdownPlacement_$type } from "./DropdownPlacement";
import { enumGetBox, EnumUtil, markType } from "igniteui-react-core";
import { WCNativeHelper } from "./WCNativeHelper";
import { DropdownScrollStrategy_$type } from "./DropdownScrollStrategy";
import { DropdownItem } from "./DropdownItem";
import { DropdownGroup } from "./DropdownGroup";
import { VoidEventArgs } from "./VoidEventArgs";
import { DropdownItemComponentEventArgs } from "./DropdownItemComponentEventArgs";
/**
 * @hidden
 */
var Dropdown = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(Dropdown, _super);
    function Dropdown() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.opening = null;
        _this.opened = null;
        _this.closing = null;
        _this.closed = null;
        _this.change = null;
        return _this;
    }
    Dropdown.prototype.showTarget = function (target_) {
        this.nativeElement.show(target_);
    };
    Dropdown.prototype.toggleTarget = function (target_) {
        this.nativeElement.toggle(target_);
    };
    Dropdown.prototype.o = function (a) {
        return (this.nativeElement.navigateTo(a));
    };
    Dropdown.prototype.p = function (a) {
        return (this.nativeElement.select(a));
    };
    Object.defineProperty(Dropdown.prototype, "s", {
        get: function () {
            var ret_ = this.b.n("placement");
            if ((ret_ === undefined && this.___rawMode)) {
                return undefined;
            }
            ret_ = enumGetBox(DropdownPlacement_$type, EnumUtil.getEnumValue(DropdownPlacement_$type, this.b.f(DropdownPlacement_$type, ret_)));
            return EnumUtil.getEnumValue(DropdownPlacement_$type, ret_);
        },
        set: function (a) {
            var value_ = enumGetBox(DropdownPlacement_$type, a);
            value_ = WCNativeHelper.camel(value_.toString());
            this.b.w("placement", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Dropdown.prototype, "t", {
        get: function () {
            var ret_ = this.b.n("scrollStrategy");
            if ((ret_ === undefined && this.___rawMode)) {
                return undefined;
            }
            ret_ = enumGetBox(DropdownScrollStrategy_$type, EnumUtil.getEnumValue(DropdownScrollStrategy_$type, this.b.f(DropdownScrollStrategy_$type, ret_)));
            return EnumUtil.getEnumValue(DropdownScrollStrategy_$type, ret_);
        },
        set: function (a) {
            var value_ = enumGetBox(DropdownScrollStrategy_$type, a);
            value_ = WCNativeHelper.camel(value_.toString());
            this.b.w("scrollStrategy", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Dropdown.prototype, "y", {
        get: function () {
            var ret_ = this.b.n("flip");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.b.w("flip", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Dropdown.prototype, "aa", {
        get: function () {
            var ret_ = this.b.n("distance");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.b.w("distance", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Dropdown.prototype, "z", {
        get: function () {
            var ret_ = this.b.n("sameWidth");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.b.w("sameWidth", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Dropdown.prototype, "n", {
        get: function () {
            var ret_ = this.b.m("items", function (a) { return new DropdownItem(); });
            return ret_;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Dropdown.prototype, "m", {
        get: function () {
            var ret_ = this.b.m("groups", function (a) { return new DropdownGroup(); });
            return ret_;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Dropdown.prototype, "q", {
        get: function () {
            var ret_ = this.b.m("selectedItem", function (a) { return new DropdownItem(); });
            return ret_;
        },
        enumerable: false,
        configurable: true
    });
    Dropdown.prototype.openingChanged = function () {
        var _this = this;
        if (this.opening != null) {
            this.b.s("opening", function (a) { return _this.opening(_this, VoidEventArgs.b(a)); });
        }
        else {
            this.b.v("opening");
        }
    };
    Dropdown.prototype.openedChanged = function () {
        var _this = this;
        if (this.opened != null) {
            this.b.s("opened", function (a) { return _this.opened(_this, VoidEventArgs.b(a)); });
        }
        else {
            this.b.v("opened");
        }
    };
    Dropdown.prototype.closingChanged = function () {
        var _this = this;
        if (this.closing != null) {
            this.b.s("closing", function (a) { return _this.closing(_this, VoidEventArgs.b(a)); });
        }
        else {
            this.b.v("closing");
        }
    };
    Dropdown.prototype.closedChanged = function () {
        var _this = this;
        if (this.closed != null) {
            this.b.s("closed", function (a) { return _this.closed(_this, VoidEventArgs.b(a)); });
        }
        else {
            this.b.v("closed");
        }
    };
    Dropdown.prototype.changeChanged = function () {
        var _this = this;
        if (this.change != null) {
            this.b.s("change", function (a) { return _this.change(_this, DropdownItemComponentEventArgs.b(a)); });
        }
        else {
            this.b.v("change");
        }
    };
    Dropdown.prototype.af = function () {
        this.nativeElement.disconnectedCallback();
    };
    Dropdown.prototype.ac = function () {
        this.nativeElement.clearSelection();
    };
    Dropdown.$t = markType(Dropdown, 'Dropdown', BaseComboBoxLike.$);
    return Dropdown;
}(BaseComboBoxLike));
export { Dropdown };
