// igniteui-react/esm5/lib/DropdownPlacement.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
export var DropdownPlacement = /*@__PURE__*/ (function (DropdownPlacement) {
    DropdownPlacement[DropdownPlacement["Top"] = 0] = "Top";
    DropdownPlacement[DropdownPlacement["TopStart"] = 1] = "TopStart";
    DropdownPlacement[DropdownPlacement["TopEnd"] = 2] = "TopEnd";
    DropdownPlacement[DropdownPlacement["Bottom"] = 3] = "Bottom";
    DropdownPlacement[DropdownPlacement["BottomStart"] = 4] = "BottomStart";
    DropdownPlacement[DropdownPlacement["BottomEnd"] = 5] = "BottomEnd";
    DropdownPlacement[DropdownPlacement["Right"] = 6] = "Right";
    DropdownPlacement[DropdownPlacement["RightStart"] = 7] = "RightStart";
    DropdownPlacement[DropdownPlacement["RightEnd"] = 8] = "RightEnd";
    DropdownPlacement[DropdownPlacement["Left"] = 9] = "Left";
    DropdownPlacement[DropdownPlacement["LeftStart"] = 10] = "LeftStart";
    DropdownPlacement[DropdownPlacement["LeftEnd"] = 11] = "LeftEnd";
    return DropdownPlacement;
})({});
/**
 * @hidden
 */
export var DropdownPlacement_$type = /*@__PURE__*/ markEnum('DropdownPlacement', 'Top,0|TopStart:top-start,1|TopEnd:top-end,2|Bottom,3|BottomStart:bottom-start,4|BottomEnd:bottom-end,5|Right,6|RightStart:right-start,7|RightEnd:right-end,8|Left,9|LeftStart:left-start,10|LeftEnd:left-end,11');
