// igniteui-react/esm5/lib/igr-button-group.js
import { __assign, __extends, __values } from "tslib";
import * as React from 'react';
import { delegateCombine, delegateRemove } from "igniteui-react-core";
import { ButtonGroupAlignment_$type } from "./ButtonGroupAlignment";
import { ButtonGroupSelection_$type } from "./ButtonGroupSelection";
import { IgrComponentValueChangedEventArgs } from "./igr-component-value-changed-event-args";
import { ButtonGroup } from "./ButtonGroup";
import { TypeRegistrar } from "igniteui-react-core";
import { ReactRenderer, PortalManager } from "igniteui-react-core";
import { IgcButtonGroupComponent, defineComponents } from "igniteui-webcomponents";
import { NamePatcher, getModifiedProps, isValidProp, ensureBool, ensureEnum, toSpinal, initializePropertiesFromCss } from "igniteui-react-core";
/**
 * The `igc-button-group` groups a series of `igc-toggle-button`s together, exposing features such as layout and selection.
*/
var IgrButtonGroup = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgrButtonGroup, _super);
    function IgrButtonGroup(props) {
        var _this = _super.call(this, props) || this;
        _this.mounted = false;
        _this.__p = null;
        _this._hasUserValues = new Set();
        _this._stylingContainer = null;
        _this._stylingParent = null;
        _this._inStyling = false;
        _this._select = null;
        _this._select_wrapped = null;
        _this._deselect = null;
        _this._deselect_wrapped = null;
        if (_this._styling) {
            NamePatcher.ensureStylablePatched(Object.getPrototypeOf(_this));
        }
        _this._getMainRef = _this._getMainRef.bind(_this);
        defineComponents(IgcButtonGroupComponent);
        _this._implementation = _this.createImplementation();
        _this._portalManager = new PortalManager("templates", function () {
            if (_this.mounted) {
                _this.setState({});
            }
        });
        if (typeof window !== 'undefined' && typeof document !== 'undefined') {
            _this._renderer = new ReactRenderer(_this._implementation.nativeElement, document, false, null, _this._portalManager);
        }
        _this._implementation.externalObject = _this;
        _this.onImplementationCreated();
        if (_this._initializeAdapters) {
            _this._initializeAdapters();
        }
        return _this;
    }
    IgrButtonGroup.prototype.createImplementation = function () {
        return new ButtonGroup();
    };
    Object.defineProperty(IgrButtonGroup.prototype, "nativeElement", {
        get: function () {
            return this._implementation.nativeElement;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrButtonGroup.prototype, "i", {
        /**
         * @hidden
         */
        get: function () {
            return this._implementation;
        } /**
                                     * @hidden
                                     */,
        enumerable: false,
        configurable: true
    });
    IgrButtonGroup._createFromInternal = function (internal) {
        if (!internal) {
            return null;
        }
        if (!internal.$type) {
            return null;
        }
        var name = internal.$type.name;
        var externalName = "Igr" + name;
        if (!TypeRegistrar.isRegistered(externalName)) {
            return null;
        }
        return TypeRegistrar.create(externalName);
    };
    IgrButtonGroup.prototype.onImplementationCreated = function () {
    };
    IgrButtonGroup.prototype.componentDidMount = function () {
        var e_1, _a;
        this.i.nativeElement = this._elRef;
        this.mounted = true;
        try {
            for (var _b = __values(Object.keys(this.props)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var p = _c.value;
                if (isValidProp(this, p)) {
                    this[p] = this.props[p];
                }
            }
        }
        catch (e_1_1) {
            e_1 = { error: e_1_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_1)
                    throw e_1.error;
            }
        }
    };
    IgrButtonGroup.prototype.shouldComponentUpdate = function (nextProps, nextState) {
        var e_2, _a;
        var mod = getModifiedProps(this.props, nextProps);
        try {
            for (var _b = __values(Object.keys(mod)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var p = _c.value;
                if (isValidProp(this, p)) {
                    this[p] = mod[p];
                }
            }
        }
        catch (e_2_1) {
            e_2 = { error: e_2_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_2)
                    throw e_2.error;
            }
        }
        return true;
    };
    IgrButtonGroup.prototype.render = function () {
        var _this = this;
        var nativePropsName = Object.keys(this.props).filter(function (prop) { return !isValidProp(_this, prop) && prop !== "originalRef"; });
        var nativeProps = {};
        nativePropsName.forEach(function (propName) {
            nativeProps[propName] = _this.props[propName];
        });
        var propChildren = this.props.children;
        var children = [];
        React.Children.forEach(propChildren, function (ch) {
            children.push(React.cloneElement(ch));
        });
        this._portalManager.onRender(children);
        var style = {};
        if (this.props.style) {
            style = this.props.style;
        }
        var div = React.createElement("igc-button-group", __assign(__assign({}, nativeProps), { ref: this._getMainRef, id: this.props.id, class: this.props.className, style: style, children: children }));
        return div;
    };
    IgrButtonGroup.prototype._getMainRef = function (ref) {
        this._elRef = ref;
    };
    Object.defineProperty(IgrButtonGroup.prototype, "disabled", {
        /**
         * Disables all buttons inside the group.
        */
        get: function () {
            return this.i.h;
        },
        set: function (v) {
            this.i.h = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrButtonGroup.prototype, "alignment", {
        /**
         * Sets the orientation of the buttons in the group.
        */
        get: function () {
            return this.i.b;
        },
        set: function (v) {
            this.i.b = ensureEnum(ButtonGroupAlignment_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrButtonGroup.prototype, "selection", {
        /**
         * Controls the mode of selection for the button group.
        */
        get: function () {
            return this.i.c;
        },
        set: function (v) {
            this.i.c = ensureEnum(ButtonGroupSelection_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrButtonGroup.prototype, "selectedItems", {
        get: function () {
            return this.i.a;
        },
        set: function (v) {
            if (v && !Array.isArray(v) && typeof (v) == "string") {
                var re = /\s*(?:,|\s|$)\s*/gm;
                v = v.split(re);
            }
            this.i.a = v;
        },
        enumerable: false,
        configurable: true
    });
    IgrButtonGroup.prototype.findByName = function (name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        return null;
    };
    Object.defineProperty(IgrButtonGroup.prototype, "hasUserValues", {
        get: function () {
            return this._hasUserValues;
        },
        enumerable: false,
        configurable: true
    });
    IgrButtonGroup.prototype.__m = function (propertyName) {
        if (!this._inStyling) {
            this._hasUserValues.add(propertyName);
        }
    };
    IgrButtonGroup.prototype._styling = function (container, component, parent) {
        if (this._inStyling) {
            return;
        }
        this._inStyling = true;
        this._stylingContainer = container;
        this._stylingParent = component;
        var genericPrefix = "";
        var typeName = this.i.$type.name;
        if (typeName.indexOf("Xam") === 0) {
            typeName = typeName.substring(3);
        }
        genericPrefix = toSpinal("ButtonGroup");
        var additionalPrefixes = [];
        var prefix = toSpinal(typeName);
        additionalPrefixes.push(prefix + "-");
        var b = this.i.$type.baseType;
        while (b && b.name != "Object" &&
            b.name != "Base" &&
            b.name != "Control" &&
            b.Name != "DependencyObject" &&
            b.Name != "FrameworkElement") {
            typeName = b.name;
            if (typeName.indexOf("Xam") === 0) {
                typeName = typeName.substring(3);
            }
            var basePrefix = toSpinal(typeName);
            additionalPrefixes.push(basePrefix + "-");
            b = b.baseType;
        }
        if (parent) {
            var parentTypeName = parent.i.$type.name;
            if (parentTypeName.indexOf("Xam") === 0) {
                parentTypeName = parentTypeName.substring(3);
            }
            var parentPrefix = toSpinal(parentTypeName);
            additionalPrefixes.push(parentPrefix + "-" + genericPrefix + "-");
            additionalPrefixes.push(parentPrefix + "-" + prefix + "-");
        }
        initializePropertiesFromCss(container, this, genericPrefix + "-", this.hasUserValues, false, additionalPrefixes);
        if (this._otherStyling) {
            this._otherStyling(container, component, parent);
        }
        this._inStyling = false;
    };
    IgrButtonGroup.prototype.setNativeElement = function (element) {
        this.i.setNativeElement(element);
    };
    Object.defineProperty(IgrButtonGroup.prototype, "select", {
        get: function () {
            return this._select;
        },
        set: function (ev) {
            var _this = this;
            if (this._select_wrapped !== null) {
                this.i.select = delegateRemove(this.i.select, this._select_wrapped);
                this._select_wrapped = null;
                this._select = null;
            }
            this._select = ev;
            this._select_wrapped = function (o, e) {
                var outerArgs = new IgrComponentValueChangedEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforeSelect) {
                    _this.beforeSelect(_this, outerArgs);
                }
                if (_this._select) {
                    _this._select(_this, outerArgs);
                }
            };
            this.i.select = delegateCombine(this.i.select, this._select_wrapped);
            if (this.i.selectChanged) {
                this.i.selectChanged();
            }
            ;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrButtonGroup.prototype, "deselect", {
        get: function () {
            return this._deselect;
        },
        set: function (ev) {
            var _this = this;
            if (this._deselect_wrapped !== null) {
                this.i.deselect = delegateRemove(this.i.deselect, this._deselect_wrapped);
                this._deselect_wrapped = null;
                this._deselect = null;
            }
            this._deselect = ev;
            this._deselect_wrapped = function (o, e) {
                var outerArgs = new IgrComponentValueChangedEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforeDeselect) {
                    _this.beforeDeselect(_this, outerArgs);
                }
                if (_this._deselect) {
                    _this._deselect(_this, outerArgs);
                }
            };
            this.i.deselect = delegateCombine(this.i.deselect, this._deselect_wrapped);
            if (this.i.deselectChanged) {
                this.i.deselectChanged();
            }
            ;
        },
        enumerable: false,
        configurable: true
    });
    return IgrButtonGroup;
}(React.Component));
export { IgrButtonGroup };
