// igniteui-react/esm5/lib/igr-docking-indicator.js
import { DockingIndicatorPosition_$type } from "./DockingIndicatorPosition";
import { DockingIndicator as DockingIndicator_internal } from "./DockingIndicator";
import { ensureEnum, ensureBool } from "igniteui-react-core";
var IgrDockingIndicator = /** @class */ /*@__PURE__*/ (function () {
    function IgrDockingIndicator() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    IgrDockingIndicator.prototype.createImplementation = function () {
        return new DockingIndicator_internal();
    };
    Object.defineProperty(IgrDockingIndicator.prototype, "nativeElement", {
        get: function () {
            return this._implementation.nativeElement;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrDockingIndicator.prototype, "i", {
        /**
         * @hidden
         */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    IgrDockingIndicator.prototype.onImplementationCreated = function () {
    };
    IgrDockingIndicator.prototype._provideImplementation = function (i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    };
    Object.defineProperty(IgrDockingIndicator.prototype, "position", {
        /**
         * Gets the position of the docking indicator.
        */
        get: function () {
            return this.i.a;
        },
        set: function (v) {
            this.i.a = ensureEnum(DockingIndicatorPosition_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrDockingIndicator.prototype, "isRoot", {
        /**
         * Gets a value indicating whether the docking indicator is a root one.
        */
        get: function () {
            return this.i.d;
        },
        set: function (v) {
            this.i.d = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrDockingIndicator.prototype, "direction", {
        /**
         * Gets a value indicating the direction of the docking indicator.
        */
        get: function () {
            return this.i.g;
        },
        set: function (v) {
            this.i.g = v;
        },
        enumerable: false,
        configurable: true
    });
    IgrDockingIndicator.prototype.findByName = function (name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        return null;
    };
    IgrDockingIndicator.prototype.setNativeElement = function (element) {
        this.i.setNativeElement(element);
    };
    return IgrDockingIndicator;
}());
export { IgrDockingIndicator };
