// igniteui-react/esm5/lib/igr-radio-change-event-args-detail.js
import { RadioChangeEventArgsDetail as RadioChangeEventArgsDetail_internal } from "./RadioChangeEventArgsDetail";
import { ensureBool } from "igniteui-react-core";
var IgrRadioChangeEventArgsDetail = /** @class */ /*@__PURE__*/ (function () {
    function IgrRadioChangeEventArgsDetail() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    IgrRadioChangeEventArgsDetail.prototype.createImplementation = function () {
        return new RadioChangeEventArgsDetail_internal();
    };
    Object.defineProperty(IgrRadioChangeEventArgsDetail.prototype, "nativeElement", {
        get: function () {
            return this._implementation.nativeElement;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrRadioChangeEventArgsDetail.prototype, "i", {
        /**
         * @hidden
         */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    IgrRadioChangeEventArgsDetail.prototype.onImplementationCreated = function () {
    };
    IgrRadioChangeEventArgsDetail.prototype._provideImplementation = function (i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    };
    Object.defineProperty(IgrRadioChangeEventArgsDetail.prototype, "checked", {
        get: function () {
            return this.i.c;
        },
        set: function (v) {
            this.i.c = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrRadioChangeEventArgsDetail.prototype, "value", {
        get: function () {
            return this.i.f;
        },
        set: function (v) {
            this.i.f = v;
        },
        enumerable: false,
        configurable: true
    });
    IgrRadioChangeEventArgsDetail.prototype.findByName = function (name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        return null;
    };
    IgrRadioChangeEventArgsDetail.prototype.setNativeElement = function (element) {
        this.i.setNativeElement(element);
    };
    return IgrRadioChangeEventArgsDetail;
}());
export { IgrRadioChangeEventArgsDetail };
