// igniteui-react/esm5/lib/igr-pane-header-connection-event-args-detail.js
import { IgrContentPane } from "./igr-content-pane";
import { IgrPaneHeaderElement } from "./igr-pane-header-element";
import { PaneHeaderConnectionEventArgsDetail as PaneHeaderConnectionEventArgsDetail_internal } from "./PaneHeaderConnectionEventArgsDetail";
var IgrPaneHeaderConnectionEventArgsDetail = /** @class */ /*@__PURE__*/ (function () {
    function IgrPaneHeaderConnectionEventArgsDetail() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    IgrPaneHeaderConnectionEventArgsDetail.prototype.createImplementation = function () {
        return new PaneHeaderConnectionEventArgsDetail_internal();
    };
    Object.defineProperty(IgrPaneHeaderConnectionEventArgsDetail.prototype, "nativeElement", {
        get: function () {
            return this._implementation.nativeElement;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrPaneHeaderConnectionEventArgsDetail.prototype, "i", {
        /**
         * @hidden
         */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    IgrPaneHeaderConnectionEventArgsDetail.prototype.onImplementationCreated = function () {
    };
    IgrPaneHeaderConnectionEventArgsDetail.prototype._provideImplementation = function (i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    };
    Object.defineProperty(IgrPaneHeaderConnectionEventArgsDetail.prototype, "pane", {
        /**
         * The pane whose header is being connected/disconnected.
        */
        get: function () {
            var r = this.i.a;
            if (r == null) {
                return null;
            }
            if (!r.externalObject) {
                var e = IgrContentPane._createFromInternal(r);
                if (e) {
                    e._implementation = r;
                }
                r.externalObject = e;
            }
            return r.externalObject;
        },
        set: function (v) {
            v == null ? this.i.a = null : this.i.a = v.i;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrPaneHeaderConnectionEventArgsDetail.prototype, "element", {
        /**
         * The header element that is being connected/disconnected.
        */
        get: function () {
            var r = this.i.b;
            if (r == null) {
                return null;
            }
            if (!r.externalObject) {
                var e = new IgrPaneHeaderElement();
                if (r.$type) {
                    e._implementation = r;
                }
                else {
                    if (e.i.setNativeElement) {
                        e.i.setNativeElement(r);
                    }
                }
                r.externalObject = e;
            }
            return r.externalObject;
        },
        set: function (v) {
            v == null ? this.i.b = null : this.i.b = v.i;
        },
        enumerable: false,
        configurable: true
    });
    IgrPaneHeaderConnectionEventArgsDetail.prototype.findByName = function (name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        if (this.pane && this.pane.name && this.pane.name == name) {
            return this.pane;
        }
        if (this.element && this.element.name && this.element.name == name) {
            return this.element;
        }
        return null;
    };
    IgrPaneHeaderConnectionEventArgsDetail.prototype.setNativeElement = function (element) {
        this.i.setNativeElement(element);
    };
    return IgrPaneHeaderConnectionEventArgsDetail;
}());
export { IgrPaneHeaderConnectionEventArgsDetail };
