// igniteui-react/esm5/lib/igr-stepper.js
import { __assign, __extends, __values } from "tslib";
import * as React from 'react';
import { delegateCombine, delegateRemove } from "igniteui-react-core";
import { StepperOrientation_$type } from "./StepperOrientation";
import { StepperStepType_$type } from "./StepperStepType";
import { StepperVerticalAnimation_$type } from "./StepperVerticalAnimation";
import { StepperHorizontalAnimation_$type } from "./StepperHorizontalAnimation";
import { StepperTitlePosition_$type } from "./StepperTitlePosition";
import { IgrActiveStepChangingArgsEventArgs } from "./igr-active-step-changing-args-event-args";
import { IgrActiveStepChangedArgsEventArgs } from "./igr-active-step-changed-args-event-args";
import { Stepper } from "./Stepper";
import { TypeRegistrar } from "igniteui-react-core";
import { ReactRenderer, PortalManager } from "igniteui-react-core";
import { IgcStepperComponent, defineComponents } from "igniteui-webcomponents";
import { NamePatcher, getModifiedProps, isValidProp, ensureEnum, ensureBool, toSpinal, initializePropertiesFromCss } from "igniteui-react-core";
import { IgrStep } from "./igr-step";
/**
 * IgxStepper provides a wizard-like workflow by dividing content into logical steps.
 * @remarks
 * The stepper component allows the user to navigate between multiple steps.
 * It supports horizontal and vertical orientation as well as keyboard navigation and provides API methods to control the active step.
*/
var IgrStepper = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgrStepper, _super);
    function IgrStepper(props) {
        var _this = _super.call(this, props) || this;
        _this.mounted = false;
        _this.__p = null;
        _this._hasUserValues = new Set();
        _this._stylingContainer = null;
        _this._stylingParent = null;
        _this._inStyling = false;
        _this._activeStepChanging = null;
        _this._activeStepChanging_wrapped = null;
        _this._activeStepChanged = null;
        _this._activeStepChanged_wrapped = null;
        if (_this._styling) {
            NamePatcher.ensureStylablePatched(Object.getPrototypeOf(_this));
        }
        _this._getMainRef = _this._getMainRef.bind(_this);
        defineComponents(IgcStepperComponent);
        _this._implementation = _this.createImplementation();
        _this._portalManager = new PortalManager("templates", function () {
            if (_this.mounted) {
                _this.setState({});
            }
        });
        if (typeof window !== 'undefined' && typeof document !== 'undefined') {
            _this._renderer = new ReactRenderer(_this._implementation.nativeElement, document, false, null, _this._portalManager);
        }
        _this._implementation.externalObject = _this;
        _this.onImplementationCreated();
        if (_this._initializeAdapters) {
            _this._initializeAdapters();
        }
        return _this;
    }
    IgrStepper.prototype.createImplementation = function () {
        return new Stepper();
    };
    Object.defineProperty(IgrStepper.prototype, "nativeElement", {
        get: function () {
            return this._implementation.nativeElement;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrStepper.prototype, "i", {
        /**
         * @hidden
         */
        get: function () {
            return this._implementation;
        } /**
                                     * @hidden
                                     */,
        enumerable: false,
        configurable: true
    });
    IgrStepper._createFromInternal = function (internal) {
        if (!internal) {
            return null;
        }
        if (!internal.$type) {
            return null;
        }
        var name = internal.$type.name;
        var externalName = "Igr" + name;
        if (!TypeRegistrar.isRegistered(externalName)) {
            return null;
        }
        return TypeRegistrar.create(externalName);
    };
    IgrStepper.prototype.onImplementationCreated = function () {
    };
    IgrStepper.prototype.componentDidMount = function () {
        var e_1, _a;
        this.i.nativeElement = this._elRef;
        this.mounted = true;
        try {
            for (var _b = __values(Object.keys(this.props)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var p = _c.value;
                if (isValidProp(this, p)) {
                    this[p] = this.props[p];
                }
            }
        }
        catch (e_1_1) {
            e_1 = { error: e_1_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_1)
                    throw e_1.error;
            }
        }
    };
    IgrStepper.prototype.shouldComponentUpdate = function (nextProps, nextState) {
        var e_2, _a;
        var mod = getModifiedProps(this.props, nextProps);
        try {
            for (var _b = __values(Object.keys(mod)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var p = _c.value;
                if (isValidProp(this, p)) {
                    this[p] = mod[p];
                }
            }
        }
        catch (e_2_1) {
            e_2 = { error: e_2_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_2)
                    throw e_2.error;
            }
        }
        return true;
    };
    IgrStepper.prototype.render = function () {
        var _this = this;
        var nativePropsName = Object.keys(this.props).filter(function (prop) { return !isValidProp(_this, prop) && prop !== "originalRef"; });
        var nativeProps = {};
        nativePropsName.forEach(function (propName) {
            nativeProps[propName] = _this.props[propName];
        });
        var propChildren = this.props.children;
        var children = [];
        React.Children.forEach(propChildren, function (ch) {
            children.push(React.cloneElement(ch));
        });
        this._portalManager.onRender(children);
        var style = {};
        if (this.props.style) {
            style = this.props.style;
        }
        var div = React.createElement("igc-stepper", __assign(__assign({}, nativeProps), { ref: this._getMainRef, id: this.props.id, class: this.props.className, style: style, children: children }));
        return div;
    };
    IgrStepper.prototype._getMainRef = function (ref) {
        this._elRef = ref;
    };
    Object.defineProperty(IgrStepper.prototype, "steps", {
        /**
         * Returns all of the stepper's steps.
        */
        get: function () {
            if (!this.i.a) {
                return undefined;
            }
            var ret = [];
            for (var i = 0; i < this.i.a.length; i++) {
                var impl = this.i.a[i];
                if (!impl.externalObject) {
                    if (impl instanceof IgrStep) {
                        ret.push(impl);
                        continue;
                    }
                    var e = IgrStep._createFromInternal(impl);
                    e._implementation = impl;
                    impl.externalObject = e;
                }
                ret.push(impl.externalObject);
            }
            return ret;
        },
        set: function (v) {
            var arr = [];
            for (var i = 0; i < v.length; i++) {
                arr.push(v[i].i);
            }
            this.i.a = arr;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrStepper.prototype, "orientation", {
        /**
         * Gets/Sets the orientation of the stepper.
         * @remarks
         * Default value is `horizontal`.
        */
        get: function () {
            return this.i.e;
        },
        set: function (v) {
            this.i.e = ensureEnum(StepperOrientation_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrStepper.prototype, "stepType", {
        /**
         * Get/Set the type of the steps.
         * @remarks
         * Default value is `full`.
        */
        get: function () {
            return this.i.f;
        },
        set: function (v) {
            this.i.f = ensureEnum(StepperStepType_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrStepper.prototype, "linear", {
        /**
         * Get/Set whether the stepper is linear.
         * @remarks
         * If the stepper is in linear mode and if the active step is valid only then the user is able to move forward.
        */
        get: function () {
            return this.i.l;
        },
        set: function (v) {
            this.i.l = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrStepper.prototype, "contentTop", {
        /**
         * Get/Set whether the content is displayed above the steps.
         * @remarks
         * Default value is `false` and the content is below the steps.
        */
        get: function () {
            return this.i.k;
        },
        set: function (v) {
            this.i.k = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrStepper.prototype, "verticalAnimation", {
        /**
         * The animation type when in vertical mode.
        */
        get: function () {
            return this.i.h;
        },
        set: function (v) {
            this.i.h = ensureEnum(StepperVerticalAnimation_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrStepper.prototype, "horizontalAnimation", {
        /**
         * The animation type when in horizontal mode.
        */
        get: function () {
            return this.i.d;
        },
        set: function (v) {
            this.i.d = ensureEnum(StepperHorizontalAnimation_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrStepper.prototype, "animationDuration", {
        /**
         * The animation duration in either vertical or horizontal mode.
        */
        get: function () {
            return this.i.m;
        },
        set: function (v) {
            this.i.m = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrStepper.prototype, "titlePosition", {
        /**
         * Get/Set the position of the steps title.
         * @remarks
         * The default value is undefined.
         * When the stepper is horizontally orientated the title is positioned below the indicator.
         * When the stepper is horizontally orientated the title is positioned on the right side of the indicator.
        */
        get: function () {
            return this.i.g;
        },
        set: function (v) {
            this.i.g = ensureEnum(StepperTitlePosition_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    IgrStepper.prototype.findByName = function (name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        return null;
    };
    Object.defineProperty(IgrStepper.prototype, "hasUserValues", {
        get: function () {
            return this._hasUserValues;
        },
        enumerable: false,
        configurable: true
    });
    IgrStepper.prototype.__m = function (propertyName) {
        if (!this._inStyling) {
            this._hasUserValues.add(propertyName);
        }
    };
    IgrStepper.prototype._styling = function (container, component, parent) {
        if (this._inStyling) {
            return;
        }
        this._inStyling = true;
        this._stylingContainer = container;
        this._stylingParent = component;
        var genericPrefix = "";
        var typeName = this.i.$type.name;
        if (typeName.indexOf("Xam") === 0) {
            typeName = typeName.substring(3);
        }
        genericPrefix = toSpinal("Stepper");
        var additionalPrefixes = [];
        var prefix = toSpinal(typeName);
        additionalPrefixes.push(prefix + "-");
        var b = this.i.$type.baseType;
        while (b && b.name != "Object" &&
            b.name != "Base" &&
            b.name != "Control" &&
            b.Name != "DependencyObject" &&
            b.Name != "FrameworkElement") {
            typeName = b.name;
            if (typeName.indexOf("Xam") === 0) {
                typeName = typeName.substring(3);
            }
            var basePrefix = toSpinal(typeName);
            additionalPrefixes.push(basePrefix + "-");
            b = b.baseType;
        }
        if (parent) {
            var parentTypeName = parent.i.$type.name;
            if (parentTypeName.indexOf("Xam") === 0) {
                parentTypeName = parentTypeName.substring(3);
            }
            var parentPrefix = toSpinal(parentTypeName);
            additionalPrefixes.push(parentPrefix + "-" + genericPrefix + "-");
            additionalPrefixes.push(parentPrefix + "-" + prefix + "-");
        }
        initializePropertiesFromCss(container, this, genericPrefix + "-", this.hasUserValues, false, additionalPrefixes);
        if (this._otherStyling) {
            this._otherStyling(container, component, parent);
        }
        this._inStyling = false;
    };
    IgrStepper.prototype.setNativeElement = function (element) {
        this.i.setNativeElement(element);
    };
    IgrStepper.prototype.connectedCallback = function () {
        this.i.r();
    };
    /**
     * Activates the step at a given index.
    
    */
    IgrStepper.prototype.navigateTo = function (index) {
        this.i.s(index);
    };
    /**
     * Activates the next enabled step.
    
    */
    IgrStepper.prototype.next = function () {
        this.i.t();
    };
    /**
     * Activates the previous enabled step.
    
    */
    IgrStepper.prototype.prev = function () {
        this.i.u();
    };
    /**
     * Resets the stepper to its initial state i.e. activates the first step.
     * @remarks
     * The steps' content will not be automatically reset.
    
    */
    IgrStepper.prototype.reset = function () {
        this.i.v();
    };
    Object.defineProperty(IgrStepper.prototype, "activeStepChanging", {
        get: function () {
            return this._activeStepChanging;
        },
        set: function (ev) {
            var _this = this;
            if (this._activeStepChanging_wrapped !== null) {
                this.i.activeStepChanging = delegateRemove(this.i.activeStepChanging, this._activeStepChanging_wrapped);
                this._activeStepChanging_wrapped = null;
                this._activeStepChanging = null;
            }
            this._activeStepChanging = ev;
            this._activeStepChanging_wrapped = function (o, e) {
                var outerArgs = new IgrActiveStepChangingArgsEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforeActiveStepChanging) {
                    _this.beforeActiveStepChanging(_this, outerArgs);
                }
                if (_this._activeStepChanging) {
                    _this._activeStepChanging(_this, outerArgs);
                }
            };
            this.i.activeStepChanging = delegateCombine(this.i.activeStepChanging, this._activeStepChanging_wrapped);
            if (this.i.activeStepChangingChanged) {
                this.i.activeStepChangingChanged();
            }
            ;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrStepper.prototype, "activeStepChanged", {
        get: function () {
            return this._activeStepChanged;
        },
        set: function (ev) {
            var _this = this;
            if (this._activeStepChanged_wrapped !== null) {
                this.i.activeStepChanged = delegateRemove(this.i.activeStepChanged, this._activeStepChanged_wrapped);
                this._activeStepChanged_wrapped = null;
                this._activeStepChanged = null;
            }
            this._activeStepChanged = ev;
            this._activeStepChanged_wrapped = function (o, e) {
                var outerArgs = new IgrActiveStepChangedArgsEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforeActiveStepChanged) {
                    _this.beforeActiveStepChanged(_this, outerArgs);
                }
                if (_this._activeStepChanged) {
                    _this._activeStepChanged(_this, outerArgs);
                }
            };
            this.i.activeStepChanged = delegateCombine(this.i.activeStepChanged, this._activeStepChanged_wrapped);
            if (this.i.activeStepChangedChanged) {
                this.i.activeStepChangedChanged();
            }
            ;
        },
        enumerable: false,
        configurable: true
    });
    return IgrStepper;
}(React.Component));
export { IgrStepper };
