// igniteui-react/esm5/lib/igr-dock-manager-point.js
import { DockManagerPoint as DockManagerPoint_internal } from "./DockManagerPoint";
var IgrDockManagerPoint = /** @class */ /*@__PURE__*/ (function () {
    function IgrDockManagerPoint() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    IgrDockManagerPoint.prototype.createImplementation = function () {
        return new DockManagerPoint_internal();
    };
    Object.defineProperty(IgrDockManagerPoint.prototype, "nativeElement", {
        get: function () {
            return this._implementation.nativeElement;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrDockManagerPoint.prototype, "i", {
        /**
         * @hidden
         */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    IgrDockManagerPoint.prototype.onImplementationCreated = function () {
    };
    IgrDockManagerPoint.prototype._provideImplementation = function (i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    };
    Object.defineProperty(IgrDockManagerPoint.prototype, "x", {
        get: function () {
            return this.i.c;
        },
        set: function (v) {
            this.i.c = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrDockManagerPoint.prototype, "y", {
        get: function () {
            return this.i.d;
        },
        set: function (v) {
            this.i.d = +v;
        },
        enumerable: false,
        configurable: true
    });
    IgrDockManagerPoint.prototype.findByName = function (name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        return null;
    };
    IgrDockManagerPoint.prototype.setNativeElement = function (element) {
        this.i.setNativeElement(element);
    };
    return IgrDockManagerPoint;
}());
export { IgrDockManagerPoint };
