// igniteui-react/esm5/lib/igr-combo-change-event-args-detail.js
import { ComboChangeType_$type } from "./ComboChangeType";
import { ComboChangeEventArgsDetail as ComboChangeEventArgsDetail_internal } from "./ComboChangeEventArgsDetail";
import { ensureEnum } from "igniteui-react-core";
var IgrComboChangeEventArgsDetail = /** @class */ /*@__PURE__*/ (function () {
    function IgrComboChangeEventArgsDetail() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    IgrComboChangeEventArgsDetail.prototype.createImplementation = function () {
        return new ComboChangeEventArgsDetail_internal();
    };
    Object.defineProperty(IgrComboChangeEventArgsDetail.prototype, "nativeElement", {
        get: function () {
            return this._implementation.nativeElement;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrComboChangeEventArgsDetail.prototype, "i", {
        /**
         * @hidden
         */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    IgrComboChangeEventArgsDetail.prototype.onImplementationCreated = function () {
    };
    IgrComboChangeEventArgsDetail.prototype._provideImplementation = function (i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    };
    Object.defineProperty(IgrComboChangeEventArgsDetail.prototype, "newValue", {
        get: function () {
            return this.i.b;
        },
        set: function (v) {
            if (v && !Array.isArray(v) && typeof (v) == "string") {
                var re = /\s*(?:,|\s|$)\s*/gm;
                v = v.split(re);
            }
            this.i.b = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrComboChangeEventArgsDetail.prototype, "items", {
        get: function () {
            return this.i.a;
        },
        set: function (v) {
            if (v && !Array.isArray(v) && typeof (v) == "string") {
                var re = /\s*(?:,|\s|$)\s*/gm;
                v = v.split(re);
            }
            this.i.a = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrComboChangeEventArgsDetail.prototype, "type", {
        get: function () {
            return this.i.c;
        },
        set: function (v) {
            this.i.c = ensureEnum(ComboChangeType_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    IgrComboChangeEventArgsDetail.prototype.findByName = function (name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        return null;
    };
    IgrComboChangeEventArgsDetail.prototype.setNativeElement = function (element) {
        this.i.setNativeElement(element);
    };
    return IgrComboChangeEventArgsDetail;
}());
export { IgrComboChangeEventArgsDetail };
