// igniteui-react/esm5/lib/igr-pane-drag-start-event-args-detail.js
import { IgrDockManagerPane } from "./igr-dock-manager-pane";
import { IgrContentPaneCollection } from "./igr-content-pane-collection";
import { PaneDragStartEventArgsDetail as PaneDragStartEventArgsDetail_internal } from "./PaneDragStartEventArgsDetail";
import { ContentPaneCollection as ContentPaneCollection_internal } from "./ContentPaneCollection";
import { ContentPane } from "./ContentPane";
import { SyncableObservableCollection$1 } from "igniteui-react-core";
var IgrPaneDragStartEventArgsDetail = /** @class */ /*@__PURE__*/ (function () {
    function IgrPaneDragStartEventArgsDetail() {
        this.mounted = false;
        this._panes = null;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    IgrPaneDragStartEventArgsDetail.prototype.createImplementation = function () {
        return new PaneDragStartEventArgsDetail_internal();
    };
    Object.defineProperty(IgrPaneDragStartEventArgsDetail.prototype, "nativeElement", {
        get: function () {
            return this._implementation.nativeElement;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrPaneDragStartEventArgsDetail.prototype, "i", {
        /**
         * @hidden
         */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    IgrPaneDragStartEventArgsDetail.prototype.onImplementationCreated = function () {
    };
    IgrPaneDragStartEventArgsDetail.prototype._provideImplementation = function (i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    };
    Object.defineProperty(IgrPaneDragStartEventArgsDetail.prototype, "sourcePane", {
        /**
         * Gets the source pane that triggers the drag start.
        */
        get: function () {
            var r = this.i.b;
            if (r == null) {
                return null;
            }
            if (!r.externalObject) {
                var e = IgrDockManagerPane._createFromInternal(r);
                if (e) {
                    e._implementation = r;
                }
                r.externalObject = e;
            }
            return r.externalObject;
        },
        set: function (v) {
            v == null ? this.i.b = null : this.i.b = v.i;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrPaneDragStartEventArgsDetail.prototype, "panes", {
        /**
         * Gets the panes that are about to get dragged.
        */
        get: function () {
            if (this._panes === null) {
                var coll = new IgrContentPaneCollection();
                var innerColl = this.i.a;
                if (!innerColl) {
                    innerColl = new ContentPaneCollection_internal();
                }
                this._panes = coll._fromInner(innerColl);
                this.i.a = innerColl;
            }
            return this._panes;
        },
        set: function (v) {
            if (this._panes !== null) {
                this._panes._setSyncTarget(null);
                this._panes = null;
            }
            var coll = new IgrContentPaneCollection();
            this._panes = coll._fromOuter(v);
            var syncColl = new SyncableObservableCollection$1(ContentPane.$type);
            var innerColl = this.i.a;
            if (!innerColl) {
                innerColl = new ContentPaneCollection_internal();
            }
            syncColl._inner = innerColl;
            syncColl.clear();
            this._panes._setSyncTarget(syncColl);
            this.i.a = innerColl;
        },
        enumerable: false,
        configurable: true
    });
    IgrPaneDragStartEventArgsDetail.prototype.findByName = function (name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        if (this.sourcePane && this.sourcePane.name && this.sourcePane.name == name) {
            return this.sourcePane;
        }
        if (this.panes != null && this.panes.findByName && this.panes.findByName(name)) {
            return this.panes.findByName(name);
        }
        return null;
    };
    IgrPaneDragStartEventArgsDetail.prototype.setNativeElement = function (element) {
        this.i.setNativeElement(element);
    };
    return IgrPaneDragStartEventArgsDetail;
}());
export { IgrPaneDragStartEventArgsDetail };
