// igniteui-react/esm5/lib/igr-floating-pane-resize-event-args-detail.js
import { IgrSplitPane } from "./igr-split-pane";
import { ResizerLocation_$type } from "./ResizerLocation";
import { FloatingPaneResizeEventArgsDetail as FloatingPaneResizeEventArgsDetail_internal } from "./FloatingPaneResizeEventArgsDetail";
import { ensureEnum } from "igniteui-react-core";
var IgrFloatingPaneResizeEventArgsDetail = /** @class */ /*@__PURE__*/ (function () {
    function IgrFloatingPaneResizeEventArgsDetail() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    IgrFloatingPaneResizeEventArgsDetail.prototype.createImplementation = function () {
        return new FloatingPaneResizeEventArgsDetail_internal();
    };
    Object.defineProperty(IgrFloatingPaneResizeEventArgsDetail.prototype, "nativeElement", {
        get: function () {
            return this._implementation.nativeElement;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrFloatingPaneResizeEventArgsDetail.prototype, "i", {
        /**
         * @hidden
         */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    IgrFloatingPaneResizeEventArgsDetail.prototype.onImplementationCreated = function () {
    };
    IgrFloatingPaneResizeEventArgsDetail.prototype._provideImplementation = function (i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    };
    Object.defineProperty(IgrFloatingPaneResizeEventArgsDetail.prototype, "sourcePane", {
        /**
         * Gets the source pane that triggers the resize operation.
        */
        get: function () {
            var r = this.i.b;
            if (r == null) {
                return null;
            }
            if (!r.externalObject) {
                var e = IgrSplitPane._createFromInternal(r);
                if (e) {
                    e._implementation = r;
                }
                r.externalObject = e;
            }
            return r.externalObject;
        },
        set: function (v) {
            v == null ? this.i.b = null : this.i.b = v.i;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrFloatingPaneResizeEventArgsDetail.prototype, "resizerLocation", {
        /**
         * Gets the edge/corner that is being dragged.
        */
        get: function () {
            return this.i.a;
        },
        set: function (v) {
            this.i.a = ensureEnum(ResizerLocation_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    IgrFloatingPaneResizeEventArgsDetail.prototype.findByName = function (name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        if (this.sourcePane && this.sourcePane.name && this.sourcePane.name == name) {
            return this.sourcePane;
        }
        return null;
    };
    IgrFloatingPaneResizeEventArgsDetail.prototype.setNativeElement = function (element) {
        this.i.setNativeElement(element);
    };
    return IgrFloatingPaneResizeEventArgsDetail;
}());
export { IgrFloatingPaneResizeEventArgsDetail };
