// igniteui-react/esm5/lib/igr-pane-header-element.js
import { IgrDragService } from "./igr-drag-service";
import { PaneHeaderElement as PaneHeaderElement_internal } from "./PaneHeaderElement";
/**
 * Describes a pane header element.
*/
var IgrPaneHeaderElement = /** @class */ /*@__PURE__*/ (function () {
    function IgrPaneHeaderElement() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    IgrPaneHeaderElement.prototype.createImplementation = function () {
        return new PaneHeaderElement_internal();
    };
    Object.defineProperty(IgrPaneHeaderElement.prototype, "nativeElement", {
        get: function () {
            return this._implementation.nativeElement;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrPaneHeaderElement.prototype, "i", {
        /**
         * @hidden
         */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    IgrPaneHeaderElement.prototype.onImplementationCreated = function () {
    };
    IgrPaneHeaderElement.prototype._provideImplementation = function (i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    };
    Object.defineProperty(IgrPaneHeaderElement.prototype, "dragService", {
        /**
         * Gets/sets the drag service.
        */
        get: function () {
            var r = this.i.a;
            if (r == null) {
                return null;
            }
            if (!r.externalObject) {
                var e = new IgrDragService();
                if (r.$type) {
                    e._implementation = r;
                }
                else {
                    if (e.i.setNativeElement) {
                        e.i.setNativeElement(r);
                    }
                }
                r.externalObject = e;
            }
            return r.externalObject;
        },
        set: function (v) {
            v == null ? this.i.a = null : this.i.a = v.i;
        },
        enumerable: false,
        configurable: true
    });
    IgrPaneHeaderElement.prototype.findByName = function (name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        if (this.dragService && this.dragService.name && this.dragService.name == name) {
            return this.dragService;
        }
        return null;
    };
    IgrPaneHeaderElement.prototype.setNativeElement = function (element) {
        this.i.setNativeElement(element);
    };
    return IgrPaneHeaderElement;
}());
export { IgrPaneHeaderElement };
