// igniteui-react/esm5/lib/igr-input-base.js
import { __extends, __values } from "tslib";
import * as React from 'react';
import { delegateCombine, delegateRemove } from "igniteui-react-core";
import { IgrComponentValueChangedEventArgs } from "./igr-component-value-changed-event-args";
import { IgrVoidEventArgs } from "./igr-void-event-args";
import { TypeRegistrar } from "igniteui-react-core";
import { ReactRenderer, PortalManager } from "igniteui-react-core";
import { NamePatcher, getModifiedProps, isValidProp, ensureBool, toSpinal, initializePropertiesFromCss } from "igniteui-react-core";
var IgrInputBase = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgrInputBase, _super);
    function IgrInputBase(props) {
        var _this = _super.call(this, props) || this;
        _this.mounted = false;
        _this.__p = null;
        _this._hasUserValues = new Set();
        _this._stylingContainer = null;
        _this._stylingParent = null;
        _this._inStyling = false;
        _this._inputOcurred = null;
        _this._inputOcurred_wrapped = null;
        _this._focus = null;
        _this._focus_wrapped = null;
        _this._blur = null;
        _this._blur_wrapped = null;
        if (_this._styling) {
            NamePatcher.ensureStylablePatched(Object.getPrototypeOf(_this));
        }
        _this._implementation = _this.createImplementation();
        _this._portalManager = new PortalManager("templates", function () {
            if (_this.mounted) {
                _this.setState({});
            }
        });
        if (typeof window !== 'undefined' && typeof document !== 'undefined') {
            _this._renderer = new ReactRenderer(_this._implementation.nativeElement, document, false, null, _this._portalManager);
        }
        _this._implementation.externalObject = _this;
        _this.onImplementationCreated();
        if (_this._initializeAdapters) {
            _this._initializeAdapters();
        }
        return _this;
    }
    IgrInputBase.prototype.createImplementation = function () {
        return null;
    };
    Object.defineProperty(IgrInputBase.prototype, "nativeElement", {
        get: function () {
            return this._implementation.nativeElement;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrInputBase.prototype, "i", {
        /**
         * @hidden
         */
        get: function () {
            return this._implementation;
        } /**
                                     * @hidden
                                     */,
        enumerable: false,
        configurable: true
    });
    IgrInputBase._createFromInternal = function (internal) {
        if (!internal) {
            return null;
        }
        if (!internal.$type) {
            return null;
        }
        var name = internal.$type.name;
        var externalName = "Igr" + name;
        if (!TypeRegistrar.isRegistered(externalName)) {
            return null;
        }
        return TypeRegistrar.create(externalName);
    };
    IgrInputBase.prototype.onImplementationCreated = function () {
    };
    IgrInputBase.prototype.componentDidMount = function () {
        var e_1, _a;
        this.mounted = true;
        try {
            for (var _b = __values(Object.keys(this.props)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var p = _c.value;
                if (isValidProp(this, p)) {
                    this[p] = this.props[p];
                }
            }
        }
        catch (e_1_1) {
            e_1 = { error: e_1_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_1)
                    throw e_1.error;
            }
        }
    };
    IgrInputBase.prototype.shouldComponentUpdate = function (nextProps, nextState) {
        var e_2, _a;
        var mod = getModifiedProps(this.props, nextProps);
        try {
            for (var _b = __values(Object.keys(mod)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var p = _c.value;
                if (isValidProp(this, p)) {
                    this[p] = mod[p];
                }
            }
        }
        catch (e_2_1) {
            e_2 = { error: e_2_1 };
        }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return))
                    _a.call(_b);
            }
            finally {
                if (e_2)
                    throw e_2.error;
            }
        }
        return true;
    };
    IgrInputBase.prototype.render = function () {
        return null;
    };
    Object.defineProperty(IgrInputBase.prototype, "outlined", {
        /**
         * Whether the control will have outlined appearance.
        */
        get: function () {
            return this.i.h;
        },
        set: function (v) {
            this.i.h = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrInputBase.prototype, "readOnly", {
        /**
         * Makes the control a readonly field.
        */
        get: function () {
            return this.i.i;
        },
        set: function (v) {
            this.i.i = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrInputBase.prototype, "placeholder", {
        /**
         * The placeholder attribute of the control.
        */
        get: function () {
            return this.i.o;
        },
        set: function (v) {
            this.i.o = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrInputBase.prototype, "label", {
        /**
         * The label for the control.
        */
        get: function () {
            return this.i.m;
        },
        set: function (v) {
            this.i.m = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrInputBase.prototype, "disabled", {
        /**
         * The disabled state of the component
        */
        get: function () {
            return this.i.f;
        },
        set: function (v) {
            this.i.f = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrInputBase.prototype, "required", {
        /**
         * Makes the control a required field in a form context.
        */
        get: function () {
            return this.i.j;
        },
        set: function (v) {
            this.i.j = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrInputBase.prototype, "name", {
        /**
         * The name attribute of the control.
        */
        get: function () {
            return this.i.n;
        },
        set: function (v) {
            this.i.n = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrInputBase.prototype, "invalid", {
        /**
         * Control the validity of the control.
        */
        get: function () {
            return this.i.g;
        },
        set: function (v) {
            this.i.g = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    IgrInputBase.prototype.findByName = function (name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        return null;
    };
    Object.defineProperty(IgrInputBase.prototype, "hasUserValues", {
        get: function () {
            return this._hasUserValues;
        },
        enumerable: false,
        configurable: true
    });
    IgrInputBase.prototype.__m = function (propertyName) {
        if (!this._inStyling) {
            this._hasUserValues.add(propertyName);
        }
    };
    IgrInputBase.prototype._styling = function (container, component, parent) {
        if (this._inStyling) {
            return;
        }
        this._inStyling = true;
        this._stylingContainer = container;
        this._stylingParent = component;
        var genericPrefix = "";
        var typeName = this.i.$type.name;
        if (typeName.indexOf("Xam") === 0) {
            typeName = typeName.substring(3);
        }
        genericPrefix = toSpinal("InputBase");
        var additionalPrefixes = [];
        var prefix = toSpinal(typeName);
        additionalPrefixes.push(prefix + "-");
        var b = this.i.$type.baseType;
        while (b && b.name != "Object" &&
            b.name != "Base" &&
            b.name != "Control" &&
            b.Name != "DependencyObject" &&
            b.Name != "FrameworkElement") {
            typeName = b.name;
            if (typeName.indexOf("Xam") === 0) {
                typeName = typeName.substring(3);
            }
            var basePrefix = toSpinal(typeName);
            additionalPrefixes.push(basePrefix + "-");
            b = b.baseType;
        }
        if (parent) {
            var parentTypeName = parent.i.$type.name;
            if (parentTypeName.indexOf("Xam") === 0) {
                parentTypeName = parentTypeName.substring(3);
            }
            var parentPrefix = toSpinal(parentTypeName);
            additionalPrefixes.push(parentPrefix + "-" + genericPrefix + "-");
            additionalPrefixes.push(parentPrefix + "-" + prefix + "-");
        }
        initializePropertiesFromCss(container, this, genericPrefix + "-", this.hasUserValues, false, additionalPrefixes);
        if (this._otherStyling) {
            this._otherStyling(container, component, parent);
        }
        this._inStyling = false;
    };
    IgrInputBase.prototype.setNativeElement = function (element) {
        this.i.setNativeElement(element);
    };
    /**
     * Sets focus on the control.
    
    */
    IgrInputBase.prototype.focusComponent = function (options) {
        this.i.t((options == null ? null : options.i));
    };
    /**
     * Removes focus from the control.
    
    */
    IgrInputBase.prototype.blurComponent = function () {
        this.i.q();
    };
    IgrInputBase.prototype.setSelectionRange = function (start, end, direction) {
        this.i.z(start, end, (direction == null ? null : direction));
    };
    IgrInputBase.prototype.setRangeText = function (replacement, start, end, selectMode) {
        this.i.y(replacement, start, end, (selectMode == null ? null : selectMode));
    };
    /**
     * Checks for validity of the control and shows the browser message if it invalid.
    
    */
    IgrInputBase.prototype.reportValidity = function () {
        this.i.v();
    };
    /**
     * Checks for validity of the control and emits the invalid event if it invalid.
    
    */
    IgrInputBase.prototype.checkValidity = function () {
        this.i.r();
    };
    /**
     * Sets a custom validation message for the control.
     * As long as `message` is not empty, the control is considered invalid.
    
    */
    IgrInputBase.prototype.setCustomValidity = function (message) {
        this.i.w(message);
    };
    Object.defineProperty(IgrInputBase.prototype, "inputOcurred", {
        get: function () {
            return this._inputOcurred;
        },
        set: function (ev) {
            var _this = this;
            if (this._inputOcurred_wrapped !== null) {
                this.i.inputOcurred = delegateRemove(this.i.inputOcurred, this._inputOcurred_wrapped);
                this._inputOcurred_wrapped = null;
                this._inputOcurred = null;
            }
            this._inputOcurred = ev;
            this._inputOcurred_wrapped = function (o, e) {
                var outerArgs = new IgrComponentValueChangedEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforeInputOcurred) {
                    _this.beforeInputOcurred(_this, outerArgs);
                }
                if (_this._inputOcurred) {
                    _this._inputOcurred(_this, outerArgs);
                }
            };
            this.i.inputOcurred = delegateCombine(this.i.inputOcurred, this._inputOcurred_wrapped);
            if (this.i.inputOcurredChanged) {
                this.i.inputOcurredChanged();
            }
            ;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrInputBase.prototype, "focus", {
        get: function () {
            return this._focus;
        },
        set: function (ev) {
            var _this = this;
            if (this._focus_wrapped !== null) {
                this.i.focus = delegateRemove(this.i.focus, this._focus_wrapped);
                this._focus_wrapped = null;
                this._focus = null;
            }
            this._focus = ev;
            this._focus_wrapped = function (o, e) {
                var outerArgs = new IgrVoidEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforeFocus) {
                    _this.beforeFocus(_this, outerArgs);
                }
                if (_this._focus) {
                    _this._focus(_this, outerArgs);
                }
            };
            this.i.focus = delegateCombine(this.i.focus, this._focus_wrapped);
            if (this.i.focusChanged) {
                this.i.focusChanged();
            }
            ;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrInputBase.prototype, "blur", {
        get: function () {
            return this._blur;
        },
        set: function (ev) {
            var _this = this;
            if (this._blur_wrapped !== null) {
                this.i.blur = delegateRemove(this.i.blur, this._blur_wrapped);
                this._blur_wrapped = null;
                this._blur = null;
            }
            this._blur = ev;
            this._blur_wrapped = function (o, e) {
                var outerArgs = new IgrVoidEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforeBlur) {
                    _this.beforeBlur(_this, outerArgs);
                }
                if (_this._blur) {
                    _this._blur(_this, outerArgs);
                }
            };
            this.i.blur = delegateCombine(this.i.blur, this._blur_wrapped);
            if (this.i.blurChanged) {
                this.i.blurChanged();
            }
            ;
        },
        enumerable: false,
        configurable: true
    });
    return IgrInputBase;
}(React.Component));
export { IgrInputBase };
