// igniteui-react/esm5/lib/igr-active-pane-event-args-detail.js
import { IgrContentPane } from "./igr-content-pane";
import { ActivePaneEventArgsDetail as ActivePaneEventArgsDetail_internal } from "./ActivePaneEventArgsDetail";
var IgrActivePaneEventArgsDetail = /** @class */ /*@__PURE__*/ (function () {
    function IgrActivePaneEventArgsDetail() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    IgrActivePaneEventArgsDetail.prototype.createImplementation = function () {
        return new ActivePaneEventArgsDetail_internal();
    };
    Object.defineProperty(IgrActivePaneEventArgsDetail.prototype, "nativeElement", {
        get: function () {
            return this._implementation.nativeElement;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrActivePaneEventArgsDetail.prototype, "i", {
        /**
         * @hidden
         */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    IgrActivePaneEventArgsDetail.prototype.onImplementationCreated = function () {
    };
    IgrActivePaneEventArgsDetail.prototype._provideImplementation = function (i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    };
    Object.defineProperty(IgrActivePaneEventArgsDetail.prototype, "newPane", {
        /**
         * Gets the new active pane.
        */
        get: function () {
            var r = this.i.a;
            if (r == null) {
                return null;
            }
            if (!r.externalObject) {
                var e = IgrContentPane._createFromInternal(r);
                if (e) {
                    e._implementation = r;
                }
                r.externalObject = e;
            }
            return r.externalObject;
        },
        set: function (v) {
            v == null ? this.i.a = null : this.i.a = v.i;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrActivePaneEventArgsDetail.prototype, "oldPane", {
        /**
         * Gets the old active pane.
        */
        get: function () {
            var r = this.i.b;
            if (r == null) {
                return null;
            }
            if (!r.externalObject) {
                var e = IgrContentPane._createFromInternal(r);
                if (e) {
                    e._implementation = r;
                }
                r.externalObject = e;
            }
            return r.externalObject;
        },
        set: function (v) {
            v == null ? this.i.b = null : this.i.b = v.i;
        },
        enumerable: false,
        configurable: true
    });
    IgrActivePaneEventArgsDetail.prototype.findByName = function (name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        if (this.newPane && this.newPane.name && this.newPane.name == name) {
            return this.newPane;
        }
        if (this.oldPane && this.oldPane.name && this.oldPane.name == name) {
            return this.oldPane;
        }
        return null;
    };
    IgrActivePaneEventArgsDetail.prototype.setNativeElement = function (element) {
        this.i.setNativeElement(element);
    };
    return IgrActivePaneEventArgsDetail;
}());
export { IgrActivePaneEventArgsDetail };
