// igniteui-react/esm5/lib/igr-pane-pinned-event-args-detail.js
import { IgrContentPane } from "./igr-content-pane";
import { IgrContentPaneCollection } from "./igr-content-pane-collection";
import { UnpinnedLocation_$type } from "./UnpinnedLocation";
import { PanePinnedEventArgsDetail as PanePinnedEventArgsDetail_internal } from "./PanePinnedEventArgsDetail";
import { ensureBool, ensureEnum } from "igniteui-react-core";
import { ContentPaneCollection as ContentPaneCollection_internal } from "./ContentPaneCollection";
import { ContentPane } from "./ContentPane";
import { SyncableObservableCollection$1 } from "igniteui-react-core";
var IgrPanePinnedEventArgsDetail = /** @class */ /*@__PURE__*/ (function () {
    function IgrPanePinnedEventArgsDetail() {
        this.mounted = false;
        this._panes = null;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    IgrPanePinnedEventArgsDetail.prototype.createImplementation = function () {
        return new PanePinnedEventArgsDetail_internal();
    };
    Object.defineProperty(IgrPanePinnedEventArgsDetail.prototype, "nativeElement", {
        get: function () {
            return this._implementation.nativeElement;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrPanePinnedEventArgsDetail.prototype, "i", {
        /**
         * @hidden
         */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    IgrPanePinnedEventArgsDetail.prototype.onImplementationCreated = function () {
    };
    IgrPanePinnedEventArgsDetail.prototype._provideImplementation = function (i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    };
    Object.defineProperty(IgrPanePinnedEventArgsDetail.prototype, "sourcePane", {
        /**
         * Gets the source pane that triggers the pinned state change.
        */
        get: function () {
            var r = this.i.a;
            if (r == null) {
                return null;
            }
            if (!r.externalObject) {
                var e = IgrContentPane._createFromInternal(r);
                if (e) {
                    e._implementation = r;
                }
                r.externalObject = e;
            }
            return r.externalObject;
        },
        set: function (v) {
            v == null ? this.i.a = null : this.i.a = v.i;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrPanePinnedEventArgsDetail.prototype, "panes", {
        /**
         * Gets/sets the panes that are about to get pinned/unpinned.
        */
        get: function () {
            if (this._panes === null) {
                var coll = new IgrContentPaneCollection();
                var innerColl = this.i.b;
                if (!innerColl) {
                    innerColl = new ContentPaneCollection_internal();
                }
                this._panes = coll._fromInner(innerColl);
                this.i.b = innerColl;
            }
            return this._panes;
        },
        set: function (v) {
            if (this._panes !== null) {
                this._panes._setSyncTarget(null);
                this._panes = null;
            }
            var coll = new IgrContentPaneCollection();
            this._panes = coll._fromOuter(v);
            var syncColl = new SyncableObservableCollection$1(ContentPane.$type);
            var innerColl = this.i.b;
            if (!innerColl) {
                innerColl = new ContentPaneCollection_internal();
            }
            syncColl._inner = innerColl;
            syncColl.clear();
            this._panes._setSyncTarget(syncColl);
            this.i.b = innerColl;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrPanePinnedEventArgsDetail.prototype, "newValue", {
        /**
         * Gets the new pinned value.
        */
        get: function () {
            return this.i.f;
        },
        set: function (v) {
            this.i.f = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrPanePinnedEventArgsDetail.prototype, "location", {
        /**
         * Gets the unpinned location.
        */
        get: function () {
            return this.i.c;
        },
        set: function (v) {
            this.i.c = ensureEnum(UnpinnedLocation_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    IgrPanePinnedEventArgsDetail.prototype.findByName = function (name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        if (this.sourcePane && this.sourcePane.name && this.sourcePane.name == name) {
            return this.sourcePane;
        }
        if (this.panes != null && this.panes.findByName && this.panes.findByName(name)) {
            return this.panes.findByName(name);
        }
        return null;
    };
    IgrPanePinnedEventArgsDetail.prototype.setNativeElement = function (element) {
        this.i.setNativeElement(element);
    };
    return IgrPanePinnedEventArgsDetail;
}());
export { IgrPanePinnedEventArgsDetail };
