// igniteui-react/esm5/lib/igr-checkbox-change-event-args-detail.js
import { CheckboxChangeEventArgsDetail as CheckboxChangeEventArgsDetail_internal } from "./CheckboxChangeEventArgsDetail";
import { ensureBool } from "igniteui-react-core";
var IgrCheckboxChangeEventArgsDetail = /** @class */ /*@__PURE__*/ (function () {
    function IgrCheckboxChangeEventArgsDetail() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    IgrCheckboxChangeEventArgsDetail.prototype.createImplementation = function () {
        return new CheckboxChangeEventArgsDetail_internal();
    };
    Object.defineProperty(IgrCheckboxChangeEventArgsDetail.prototype, "nativeElement", {
        get: function () {
            return this._implementation.nativeElement;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrCheckboxChangeEventArgsDetail.prototype, "i", {
        /**
         * @hidden
         */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    IgrCheckboxChangeEventArgsDetail.prototype.onImplementationCreated = function () {
    };
    IgrCheckboxChangeEventArgsDetail.prototype._provideImplementation = function (i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    };
    Object.defineProperty(IgrCheckboxChangeEventArgsDetail.prototype, "checked", {
        get: function () {
            return this.i.c;
        },
        set: function (v) {
            this.i.c = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrCheckboxChangeEventArgsDetail.prototype, "value", {
        get: function () {
            return this.i.f;
        },
        set: function (v) {
            this.i.f = v;
        },
        enumerable: false,
        configurable: true
    });
    IgrCheckboxChangeEventArgsDetail.prototype.findByName = function (name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        return null;
    };
    IgrCheckboxChangeEventArgsDetail.prototype.setNativeElement = function (element) {
        this.i.setNativeElement(element);
    };
    return IgrCheckboxChangeEventArgsDetail;
}());
export { IgrCheckboxChangeEventArgsDetail };
