// igniteui-react/esm5/lib/igr-input.js
import { __assign, __extends } from "tslib";
import * as React from 'react';
import { delegateCombine, delegateRemove } from "igniteui-react-core";
import { InputType_$type } from "./InputType";
import { IgrComponentDataValueChangedEventArgs } from "./igr-component-data-value-changed-event-args";
import { IgrInputBase } from "./igr-input-base";
import { Input } from "./Input";
import { IgcInputComponent, defineComponents } from "igniteui-webcomponents";
import { isValidProp, ensureEnum, ensureBool } from "igniteui-react-core";
/**

*/
var IgrInput = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgrInput, _super);
    function IgrInput(props) {
        var _this = _super.call(this, props) || this;
        _this._change = null;
        _this._change_wrapped = null;
        _this._getMainRef = _this._getMainRef.bind(_this);
        defineComponents(IgcInputComponent);
        return _this;
    }
    IgrInput.prototype.createImplementation = function () {
        return new Input();
    };
    Object.defineProperty(IgrInput.prototype, "i", {
        /**
                                     * @hidden
                                     */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    IgrInput.prototype.componentDidMount = function () {
        this.i.nativeElement = this._elRef;
        this.mounted = true;
        _super.prototype.componentDidMount.call(this);
    };
    IgrInput.prototype.render = function () {
        var _this = this;
        var nativePropsName = Object.keys(this.props).filter(function (prop) { return !isValidProp(_this, prop) && prop !== "originalRef"; });
        var nativeProps = {};
        nativePropsName.forEach(function (propName) {
            nativeProps[propName] = _this.props[propName];
        });
        var propChildren = this.props.children;
        var children = [];
        React.Children.forEach(propChildren, function (ch) {
            children.push(React.cloneElement(ch));
        });
        this._portalManager.onRender(children);
        var style = {};
        if (this.props.style) {
            style = this.props.style;
        }
        var div = React.createElement("igc-input", __assign(__assign({}, nativeProps), { ref: this._getMainRef, id: this.props.id, class: this.props.className, style: style, children: children }));
        return div;
    };
    IgrInput.prototype._getMainRef = function (ref) {
        this._elRef = ref;
    };
    Object.defineProperty(IgrInput.prototype, "value", {
        /**
         * The value of the control.
        */
        get: function () {
            return this.i.am;
        },
        set: function (v) {
            this.i.am = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrInput.prototype, "type", {
        /**
         * The type attribute of the control.
        */
        get: function () {
            return this.i.ab;
        },
        set: function (v) {
            this.i.ab = ensureEnum(InputType_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrInput.prototype, "inputMode", {
        /**
         * The input mode attribute of the control.
         * See [relevant MDN article](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/inputmode)
        */
        get: function () {
            return this.i.ai;
        },
        set: function (v) {
            this.i.ai = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrInput.prototype, "pattern", {
        /**
         * The pattern attribute of the control.
        */
        get: function () {
            return this.i.al;
        },
        set: function (v) {
            this.i.al = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrInput.prototype, "minLength", {
        /**
         * The minimum string length required by the control.
        */
        get: function () {
            return this.i.af;
        },
        set: function (v) {
            this.i.af = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrInput.prototype, "maxLength", {
        /**
         * The maximum string length of the control.
        */
        get: function () {
            return this.i.ae;
        },
        set: function (v) {
            this.i.ae = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrInput.prototype, "min", {
        /**
         * The min attribute of the control.
        */
        get: function () {
            return this.i.ak;
        },
        set: function (v) {
            this.i.ak = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrInput.prototype, "max", {
        /**
         * The max attribute of the control.
        */
        get: function () {
            return this.i.aj;
        },
        set: function (v) {
            this.i.aj = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrInput.prototype, "step", {
        /**
         * The step attribute of the control.
        */
        get: function () {
            return this.i.ag;
        },
        set: function (v) {
            this.i.ag = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrInput.prototype, "autofocus", {
        /**
         * The autofocus attribute of the control.
        */
        get: function () {
            return this.i.ac;
        },
        set: function (v) {
            this.i.ac = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrInput.prototype, "autocomplete", {
        /**
         * The autocomplete attribute of the control.
        */
        get: function () {
            return this.i.ah;
        },
        set: function (v) {
            this.i.ah = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrInput.prototype, "validateOnly", {
        /**
         * Enables validation rules to be evaluated without restricting user input. This applies to the `maxLength` property for
         * string-type inputs or allows spin buttons to exceed the predefined `min/max` limits for number-type inputs.
        */
        get: function () {
            return this.i.ad;
        },
        set: function (v) {
            this.i.ad = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Selects all text within the input.
    
    */
    IgrInput.prototype.select = function () {
        this.i.ao();
    };
    /**
     * Increments the numeric value of the input by one or more steps.
    
    */
    IgrInput.prototype.stepUp = function (n) {
        this.i.aq(n);
    };
    /**
     * Decrements the numeric value of the input by one or more steps.
    
    */
    IgrInput.prototype.stepDown = function (n) {
        this.i.ap(n);
    };
    Object.defineProperty(IgrInput.prototype, "change", {
        get: function () {
            return this._change;
        },
        set: function (ev) {
            var _this = this;
            if (this._change_wrapped !== null) {
                this.i.change = delegateRemove(this.i.change, this._change_wrapped);
                this._change_wrapped = null;
                this._change = null;
            }
            this._change = ev;
            this._change_wrapped = function (o, e) {
                var outerArgs = new IgrComponentDataValueChangedEventArgs();
                outerArgs._provideImplementation(e);
                if (_this.beforeChange) {
                    _this.beforeChange(_this, outerArgs);
                }
                if (_this._change) {
                    _this._change(_this, outerArgs);
                }
            };
            this.i.change = delegateCombine(this.i.change, this._change_wrapped);
            if (this.i.changeChanged) {
                this.i.changeChanged();
            }
            ;
        },
        enumerable: false,
        configurable: true
    });
    return IgrInput;
}(IgrInputBase));
export { IgrInput };
