// igniteui-react/esm5/lib/igr-split-pane.js
import { __assign, __extends } from "tslib";
import * as React from 'react';
import { DockManagerPaneType_$type } from "./DockManagerPaneType";
import { SplitPaneOrientation_$type } from "./SplitPaneOrientation";
import { IgrDockManagerPaneCollection } from "./igr-dock-manager-pane-collection";
import { IgrDockManagerPoint } from "./igr-dock-manager-point";
import { IgrDockManagerPane } from "./igr-dock-manager-pane";
import { SplitPane } from "./SplitPane";
import { defineCustomElements } from "igniteui-dockmanager/loader";
import { isValidProp, ensureEnum, ensureBool } from "igniteui-react-core";
import { DockManagerPaneCollection as DockManagerPaneCollection_internal } from "./DockManagerPaneCollection";
import { DockManagerPane } from "./DockManagerPane";
import { SyncableObservableCollection$1 } from "igniteui-react-core";
var IgrSplitPane = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(IgrSplitPane, _super);
    function IgrSplitPane(props) {
        var _this = _super.call(this, props) || this;
        _this._panes = null;
        _this._getMainRef = _this._getMainRef.bind(_this);
        defineCustomElements();
        return _this;
    }
    IgrSplitPane.prototype.createImplementation = function () {
        return new SplitPane();
    };
    Object.defineProperty(IgrSplitPane.prototype, "i", {
        /**
                                     * @hidden
                                     */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    IgrSplitPane.prototype.componentDidMount = function () {
        this.i.nativeElement = this._elRef;
        this.mounted = true;
        _super.prototype.componentDidMount.call(this);
    };
    IgrSplitPane.prototype.render = function () {
        var _this = this;
        var nativePropsName = Object.keys(this.props).filter(function (prop) { return !isValidProp(_this, prop) && prop !== "originalRef"; });
        var nativeProps = {};
        nativePropsName.forEach(function (propName) {
            nativeProps[propName] = _this.props[propName];
        });
        var propChildren = this.props.children;
        var children = [];
        React.Children.forEach(propChildren, function (ch) {
            children.push(React.cloneElement(ch));
        });
        this._portalManager.onRender(children);
        var style = {};
        if (this.props.style) {
            style = this.props.style;
        }
        var div = React.createElement("", __assign(__assign({}, nativeProps), { ref: this._getMainRef, class: this.props.className, style: style, children: children }));
        return div;
    };
    IgrSplitPane.prototype._getMainRef = function (ref) {
        this._elRef = ref;
    };
    Object.defineProperty(IgrSplitPane.prototype, "id", {
        /**
         * The id of the pane. If not set the Dock Manager generates it automatically.
        */
        get: function () {
            return this.i.s;
        },
        set: function (v) {
            this.i.s = v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSplitPane.prototype, "type", {
        /**
         * The type of the pane.
        */
        get: function () {
            return this.i.f;
        },
        set: function (v) {
            this.i.f = ensureEnum(DockManagerPaneType_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSplitPane.prototype, "orientation", {
        /**
         * The orientation of the split pane.
        */
        get: function () {
            return this.i.h;
        },
        set: function (v) {
            this.i.h = ensureEnum(SplitPaneOrientation_$type, v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSplitPane.prototype, "panes", {
        /**
         * The child panes of the split pane.
        */
        get: function () {
            if (this._panes === null) {
                var coll = new IgrDockManagerPaneCollection();
                var innerColl = this.i.e;
                if (!innerColl) {
                    innerColl = new DockManagerPaneCollection_internal();
                }
                this._panes = coll._fromInner(innerColl);
            }
            return this._panes;
        },
        set: function (v) {
            if (this._panes !== null) {
                this._panes._setSyncTarget(null);
                this._panes = null;
            }
            var coll = new IgrDockManagerPaneCollection();
            this._panes = coll._fromOuter(v);
            var syncColl = new SyncableObservableCollection$1(DockManagerPane.$type);
            var innerColl = this.i.e;
            if (!innerColl) {
                innerColl = new DockManagerPaneCollection_internal();
            }
            syncColl._inner = innerColl;
            syncColl.clear();
            this._panes._setSyncTarget(syncColl);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSplitPane.prototype, "size", {
        /**
         * The size of the pane relative to its sibling panes' sizes. Defaults to 100.
        */
        get: function () {
            return this.i.p;
        },
        set: function (v) {
            this.i.p = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSplitPane.prototype, "floatingLocation", {
        /**
         * The absolute location point of the pane. Applies only for floating panes.
        */
        get: function () {
            var r = this.i.g;
            if (r == null) {
                return null;
            }
            if (!r.externalObject) {
                var e = new IgrDockManagerPoint();
                if (r.$type) {
                    e._implementation = r;
                }
                else {
                    if (e.i.setNativeElement) {
                        e.i.setNativeElement(r);
                    }
                }
                r.externalObject = e;
            }
            return r.externalObject;
        },
        set: function (v) {
            v == null ? this.i.g = null : this.i.g = v.i;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSplitPane.prototype, "floatingWidth", {
        /**
         * The absolute width of the pane. Applies only for floating panes. Defaults to 100.
        */
        get: function () {
            return this.i.o;
        },
        set: function (v) {
            this.i.o = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSplitPane.prototype, "floatingHeight", {
        /**
         * The absolute height of the pane. Applies only for floating panes. Defaults to 100.
        */
        get: function () {
            return this.i.n;
        },
        set: function (v) {
            this.i.n = +v;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSplitPane.prototype, "floatingResizable", {
        /**
         * Determines whether floating pane resizing is allowed. Applies only for floating panes.
        */
        get: function () {
            return this.i.l;
        },
        set: function (v) {
            this.i.l = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSplitPane.prototype, "allowEmpty", {
        /**
         * Determines whether the pane should present in the UI when empty.
        */
        get: function () {
            return this.i.k;
        },
        set: function (v) {
            this.i.k = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrSplitPane.prototype, "isMaximized", {
        /**
         * Determines whether a split pane is maximized or not. Defaults to false.
         * @deprecated Use the `isMaximized` property of TabGroupPane and/or ContentPane instead.  Having isMaximized set to true on a split pane level has no real effect as split panes serve as containers only, meaning they have no actual content to be shown maximized.
        */
        get: function () {
            return this.i.m;
        },
        set: function (v) {
            this.i.m = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    IgrSplitPane.prototype.findByName = function (name) {
        var baseResult = _super.prototype.findByName.call(this, name);
        if (baseResult) {
            return baseResult;
        }
        if (this.panes != null && this.panes.findByName && this.panes.findByName(name)) {
            return this.panes.findByName(name);
        }
        if (this.floatingLocation && this.floatingLocation.name && this.floatingLocation.name == name) {
            return this.floatingLocation;
        }
        return null;
    };
    IgrSplitPane.prototype.setNativeElement = function (element) {
        this.i.setNativeElement(element);
    };
    return IgrSplitPane;
}(IgrDockManagerPane));
export { IgrSplitPane };
