// igniteui-react/esm5/lib/igr-pane-drag-over-event-args-detail.js
import { IgrDockManagerPane } from "./igr-dock-manager-pane";
import { IgrContentPaneCollection } from "./igr-content-pane-collection";
import { IgrPaneDragAction } from "./igr-pane-drag-action";
import { PaneDragOverEventArgsDetail as PaneDragOverEventArgsDetail_internal } from "./PaneDragOverEventArgsDetail";
import { ensureBool } from "igniteui-react-core";
import { ContentPaneCollection as ContentPaneCollection_internal } from "./ContentPaneCollection";
import { ContentPane } from "./ContentPane";
import { SyncableObservableCollection$1 } from "igniteui-react-core";
var IgrPaneDragOverEventArgsDetail = /** @class */ /*@__PURE__*/ (function () {
    function IgrPaneDragOverEventArgsDetail() {
        this.mounted = false;
        this._panes = null;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    IgrPaneDragOverEventArgsDetail.prototype.createImplementation = function () {
        return new PaneDragOverEventArgsDetail_internal();
    };
    Object.defineProperty(IgrPaneDragOverEventArgsDetail.prototype, "nativeElement", {
        get: function () {
            return this._implementation.nativeElement;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrPaneDragOverEventArgsDetail.prototype, "i", {
        /**
         * @hidden
         */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    IgrPaneDragOverEventArgsDetail.prototype.onImplementationCreated = function () {
    };
    IgrPaneDragOverEventArgsDetail.prototype._provideImplementation = function (i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    };
    Object.defineProperty(IgrPaneDragOverEventArgsDetail.prototype, "sourcePane", {
        /**
         * Gets the source pane that triggers the drag over.
        */
        get: function () {
            var r = this.i.b;
            if (r == null) {
                return null;
            }
            if (!r.externalObject) {
                var e = IgrDockManagerPane._createFromInternal(r);
                if (e) {
                    e._implementation = r;
                }
                r.externalObject = e;
            }
            return r.externalObject;
        },
        set: function (v) {
            v == null ? this.i.b = null : this.i.b = v.i;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrPaneDragOverEventArgsDetail.prototype, "panes", {
        /**
         * Gets the panes that are dragged over.
        */
        get: function () {
            if (this._panes === null) {
                var coll = new IgrContentPaneCollection();
                var innerColl = this.i.a;
                if (!innerColl) {
                    innerColl = new ContentPaneCollection_internal();
                }
                this._panes = coll._fromInner(innerColl);
                this.i.a = innerColl;
            }
            return this._panes;
        },
        set: function (v) {
            if (this._panes !== null) {
                this._panes._setSyncTarget(null);
                this._panes = null;
            }
            var coll = new IgrContentPaneCollection();
            this._panes = coll._fromOuter(v);
            var syncColl = new SyncableObservableCollection$1(ContentPane.$type);
            var innerColl = this.i.a;
            if (!innerColl) {
                innerColl = new ContentPaneCollection_internal();
            }
            syncColl._inner = innerColl;
            syncColl.clear();
            this._panes._setSyncTarget(syncColl);
            this.i.a = innerColl;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrPaneDragOverEventArgsDetail.prototype, "action", {
        /**
         * Gets information about the action being performed.
        */
        get: function () {
            var r = this.i.c;
            if (r == null) {
                return null;
            }
            if (!r.externalObject) {
                var e = new IgrPaneDragAction({});
                if (r.$type) {
                    e._implementation = r;
                }
                else {
                    if (e.i.setNativeElement) {
                        e.i.setNativeElement(r);
                    }
                }
                r.externalObject = e;
            }
            return r.externalObject;
        },
        set: function (v) {
            v == null ? this.i.c = null : this.i.c = v.i;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrPaneDragOverEventArgsDetail.prototype, "isValid", {
        /**
         * Gets/sets whether the action is valid.
        */
        get: function () {
            return this.i.f;
        },
        set: function (v) {
            this.i.f = ensureBool(v);
        },
        enumerable: false,
        configurable: true
    });
    IgrPaneDragOverEventArgsDetail.prototype.findByName = function (name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        if (this.sourcePane && this.sourcePane.name && this.sourcePane.name == name) {
            return this.sourcePane;
        }
        if (this.panes != null && this.panes.findByName && this.panes.findByName(name)) {
            return this.panes.findByName(name);
        }
        if (this.action && this.action.name && this.action.name == name) {
            return this.action;
        }
        return null;
    };
    IgrPaneDragOverEventArgsDetail.prototype.setNativeElement = function (element) {
        this.i.setNativeElement(element);
    };
    return IgrPaneDragOverEventArgsDetail;
}());
export { IgrPaneDragOverEventArgsDetail };
