// igniteui-react/esm5/lib/igr-drag-service.js
import { DragService as DragService_internal } from "./DragService";
var IgrDragService = /** @class */ /*@__PURE__*/ (function () {
    function IgrDragService() {
        this.mounted = false;
        this._implementation = this.createImplementation();
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    }
    IgrDragService.prototype.createImplementation = function () {
        return new DragService_internal();
    };
    Object.defineProperty(IgrDragService.prototype, "nativeElement", {
        get: function () {
            return this._implementation.nativeElement;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IgrDragService.prototype, "i", {
        /**
         * @hidden
         */
        get: function () {
            return this._implementation;
        },
        enumerable: false,
        configurable: true
    });
    IgrDragService.prototype.onImplementationCreated = function () {
    };
    IgrDragService.prototype._provideImplementation = function (i) {
        this._implementation = i;
        this._implementation.externalObject = this;
        this.onImplementationCreated();
        if (this._initializeAdapters) {
            this._initializeAdapters();
        }
    };
    IgrDragService.prototype.findByName = function (name) {
        if (this.findEphemera) {
            if (name && name.indexOf("@@e:") == 0) {
                return this.findEphemera(name);
            }
        }
        return null;
    };
    IgrDragService.prototype.setNativeElement = function (element) {
        this.i.setNativeElement(element);
    };
    return IgrDragService;
}());
export { IgrDragService };
