// igniteui-react/esm5/lib/InputType.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
export var InputType = /*@__PURE__*/ (function (InputType) {
    InputType[InputType["Number"] = 0] = "Number";
    InputType[InputType["Email"] = 1] = "Email";
    InputType[InputType["Password"] = 2] = "Password";
    InputType[InputType["Search"] = 3] = "Search";
    InputType[InputType["Tel"] = 4] = "Tel";
    InputType[InputType["Text"] = 5] = "Text";
    InputType[InputType["Url"] = 6] = "Url";
    return InputType;
})({});
/**
 * @hidden
 */
export var InputType_$type = /*@__PURE__*/ markEnum('InputType', 'Number,0|Email,1|Password,2|Search,3|Tel,4|Text,5|Url,6');
