// igniteui-react/esm5/lib/LinearProgressLabelAlign.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
export var LinearProgressLabelAlign = /*@__PURE__*/ (function (LinearProgressLabelAlign) {
    LinearProgressLabelAlign[LinearProgressLabelAlign["Bottom"] = 0] = "Bottom";
    LinearProgressLabelAlign[LinearProgressLabelAlign["Top"] = 1] = "Top";
    LinearProgressLabelAlign[LinearProgressLabelAlign["TopStart"] = 2] = "TopStart";
    LinearProgressLabelAlign[LinearProgressLabelAlign["TopEnd"] = 3] = "TopEnd";
    LinearProgressLabelAlign[LinearProgressLabelAlign["BottomStart"] = 4] = "BottomStart";
    LinearProgressLabelAlign[LinearProgressLabelAlign["BottomEnd"] = 5] = "BottomEnd";
    return LinearProgressLabelAlign;
})({});
/**
 * @hidden
 */
export var LinearProgressLabelAlign_$type = /*@__PURE__*/ markEnum('LinearProgressLabelAlign', 'Bottom,0|Top,1|TopStart:top-start,2|TopEnd:top-end,3|BottomStart:bottom-start,4|BottomEnd:bottom-end,5');
