// igniteui-react/esm5/lib/NavDrawer.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { BaseElement } from "./BaseElement";
import { WCNativeHelper } from "./WCNativeHelper";
import { enumGetBox, EnumUtil, markType } from "igniteui-react-core";
import { NavDrawerPosition_$type } from "./NavDrawerPosition";
/**
 * @hidden
 */
var NavDrawer = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(NavDrawer, _super);
    function NavDrawer() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.b = new WCNativeHelper();
        _this.h = null;
        return _this;
    }
    Object.defineProperty(NavDrawer.prototype, "c", {
        get: function () {
            return this.b;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NavDrawer.prototype, "nativeElement", {
        get: function () {
            return this.h;
        },
        set: function (a) {
            this.h = a;
            this.b.o = this.h;
        },
        enumerable: false,
        configurable: true
    });
    NavDrawer.prototype.setNativeElement = function (a) {
        this.nativeElement = a;
    };
    Object.defineProperty(NavDrawer.prototype, "a", {
        get: function () {
            var ret_ = this.c.n("position");
            if ((ret_ === undefined && this.___rawMode)) {
                return undefined;
            }
            ret_ = enumGetBox(NavDrawerPosition_$type, EnumUtil.getEnumValue(NavDrawerPosition_$type, this.c.f(NavDrawerPosition_$type, ret_)));
            return EnumUtil.getEnumValue(NavDrawerPosition_$type, ret_);
        },
        set: function (a) {
            var value_ = enumGetBox(NavDrawerPosition_$type, a);
            value_ = WCNativeHelper.camel(value_.toString());
            this.c.w("position", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NavDrawer.prototype, "e", {
        get: function () {
            var ret_ = this.c.n("open");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.c.w("open", value_);
        },
        enumerable: false,
        configurable: true
    });
    NavDrawer.prototype.f = function () {
        return (this.nativeElement.show());
    };
    NavDrawer.prototype.d = function () {
        return (this.nativeElement.hide());
    };
    NavDrawer.prototype.g = function () {
        return (this.nativeElement.toggle());
    };
    NavDrawer.$t = markType(NavDrawer, 'NavDrawer', BaseElement.$);
    return NavDrawer;
}(BaseElement));
export { NavDrawer };
