// igniteui-react/esm5/lib/RangeTextSelectMode.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
export var RangeTextSelectMode = /*@__PURE__*/ (function (RangeTextSelectMode) {
    RangeTextSelectMode[RangeTextSelectMode["Preserve"] = 0] = "Preserve";
    RangeTextSelectMode[RangeTextSelectMode["Select"] = 1] = "Select";
    RangeTextSelectMode[RangeTextSelectMode["Start"] = 2] = "Start";
    RangeTextSelectMode[RangeTextSelectMode["End"] = 3] = "End";
    return RangeTextSelectMode;
})({});
/**
 * @hidden
 */
export var RangeTextSelectMode_$type = /*@__PURE__*/ markEnum('RangeTextSelectMode', 'Preserve,0|Select,1|Start,2|End,3');
