// igniteui-react/esm5/lib/Rating.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { BaseElement } from "./BaseElement";
import { WCNativeHelper } from "./WCNativeHelper";
import { markType } from "igniteui-react-core";
import { NumberEventArgs } from "./NumberEventArgs";
/**
 * @hidden
 */
var Rating = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(Rating, _super);
    function Rating() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.c = new WCNativeHelper();
        _this.n = null;
        _this.change = null;
        _this.hover = null;
        return _this;
    }
    Object.defineProperty(Rating.prototype, "d", {
        get: function () {
            return this.c;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Rating.prototype, "nativeElement", {
        get: function () {
            return this.n;
        },
        set: function (a) {
            this.n = a;
            this.c.o = this.n;
        },
        enumerable: false,
        configurable: true
    });
    Rating.prototype.setNativeElement = function (a) {
        this.nativeElement = a;
    };
    Object.defineProperty(Rating.prototype, "k", {
        get: function () {
            var ret_ = this.d.n("max");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.d.w("max", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Rating.prototype, "l", {
        get: function () {
            var ret_ = this.d.n("step");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.d.w("step", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Rating.prototype, "p", {
        get: function () {
            var ret_ = this.d.n("label");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.d.w("label", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Rating.prototype, "r", {
        get: function () {
            var ret_ = this.d.n("valueFormat");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.d.w("valueFormat", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Rating.prototype, "m", {
        get: function () {
            var ret_ = this.d.n("value");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.d.w("value", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Rating.prototype, "g", {
        get: function () {
            var ret_ = this.d.n("hoverPreview");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.d.w("hoverPreview", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Rating.prototype, "i", {
        get: function () {
            var ret_ = this.d.n("readOnly");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.d.w("readOnly", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Rating.prototype, "j", {
        get: function () {
            var ret_ = this.d.n("single");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.d.w("single", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Rating.prototype, "e", {
        get: function () {
            var ret_ = this.d.n("allowReset");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.d.w("allowReset", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Rating.prototype, "f", {
        get: function () {
            var ret_ = this.d.n("disabled");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.d.w("disabled", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Rating.prototype, "q", {
        get: function () {
            var ret_ = this.d.n("name");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.d.w("name", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Rating.prototype, "h", {
        get: function () {
            var ret_ = this.d.n("invalid");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.d.w("invalid", value_);
        },
        enumerable: false,
        configurable: true
    });
    Rating.prototype.changeChanged = function () {
        var _this = this;
        if (this.change != null) {
            this.d.s("change", function (a) { return _this.change(_this, NumberEventArgs.a(a)); });
        }
        else {
            this.d.v("change");
        }
    };
    Rating.prototype.hoverChanged = function () {
        var _this = this;
        if (this.hover != null) {
            this.d.s("hover", function (a) { return _this.hover(_this, NumberEventArgs.a(a)); });
        }
        else {
            this.d.v("hover");
        }
    };
    Rating.prototype.z = function (a) {
        if (a === void 0) {
            a = -1;
        }
        this.nativeElement.stepUp(this.d.i(a));
    };
    Rating.prototype.y = function (a) {
        if (a === void 0) {
            a = -1;
        }
        this.nativeElement.stepDown(this.d.i(a));
    };
    Rating.prototype.v = function () {
        this.nativeElement.reportValidity();
    };
    Rating.prototype.t = function () {
        this.nativeElement.checkValidity();
    };
    Rating.prototype.w = function (a) {
        this.nativeElement.setCustomValidity(this.d.i(a));
    };
    Rating.$t = markType(Rating, 'Rating', BaseElement.$);
    return Rating;
}(BaseElement));
export { Rating };
