// igniteui-react/esm5/lib/ResizerLocation.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
export var ResizerLocation = /*@__PURE__*/ (function (ResizerLocation) {
    ResizerLocation[ResizerLocation["Top"] = 0] = "Top";
    ResizerLocation[ResizerLocation["Bottom"] = 1] = "Bottom";
    ResizerLocation[ResizerLocation["Left"] = 2] = "Left";
    ResizerLocation[ResizerLocation["Right"] = 3] = "Right";
    ResizerLocation[ResizerLocation["TopLeft"] = 4] = "TopLeft";
    ResizerLocation[ResizerLocation["TopRight"] = 5] = "TopRight";
    ResizerLocation[ResizerLocation["BottomLeft"] = 6] = "BottomLeft";
    ResizerLocation[ResizerLocation["BottomRight"] = 7] = "BottomRight";
    return ResizerLocation;
})({});
/**
 * @hidden
 */
export var ResizerLocation_$type = /*@__PURE__*/ markEnum('ResizerLocation', 'Top:top,0|Bottom:bottom,1|Left:left,2|Right:right,3|TopLeft:topLeft,4|TopRight:topRight,5|BottomLeft:bottomLeft,6|BottomRight:bottomRight,7');
