// igniteui-react/esm5/lib/Select.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { BaseComboBoxLike } from "./BaseComboBoxLike";
import { enumGetBox, EnumUtil, markType } from "igniteui-react-core";
import { WCNativeHelper } from "./WCNativeHelper";
import { DropdownPlacement_$type } from "./DropdownPlacement";
import { SelectScrollStrategy_$type } from "./SelectScrollStrategy";
import { SelectItem } from "./SelectItem";
import { SelectGroup } from "./SelectGroup";
import { SelectItemComponentEventArgs } from "./SelectItemComponentEventArgs";
import { VoidEventArgs } from "./VoidEventArgs";
/**
 * @hidden
 */
var Select = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(Select, _super);
    function Select() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.change = null;
        _this.focus = null;
        _this.blur = null;
        _this.opening = null;
        _this.opened = null;
        _this.closing = null;
        _this.closed = null;
        return _this;
    }
    Object.defineProperty(Select.prototype, "ah", {
        get: function () {
            var ret_ = this.b.n("value");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.b.w("value", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Select.prototype, "ab", {
        get: function () {
            var ret_ = this.b.n("outlined");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.b.w("outlined", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Select.prototype, "y", {
        get: function () {
            var ret_ = this.b.n("autofocus");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.b.w("autofocus", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Select.prototype, "ad", {
        get: function () {
            var ret_ = this.b.n("distance");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.b.w("distance", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Select.prototype, "ae", {
        get: function () {
            var ret_ = this.b.n("label");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.b.w("label", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Select.prototype, "ag", {
        get: function () {
            var ret_ = this.b.n("placeholder");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.b.w("placeholder", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Select.prototype, "o", {
        get: function () {
            var ret_ = this.b.n("placement");
            if ((ret_ === undefined && this.___rawMode)) {
                return undefined;
            }
            ret_ = enumGetBox(DropdownPlacement_$type, EnumUtil.getEnumValue(DropdownPlacement_$type, this.b.f(DropdownPlacement_$type, ret_)));
            return EnumUtil.getEnumValue(DropdownPlacement_$type, ret_);
        },
        set: function (a) {
            var value_ = enumGetBox(DropdownPlacement_$type, a);
            value_ = WCNativeHelper.camel(value_.toString());
            this.b.w("placement", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Select.prototype, "r", {
        get: function () {
            var ret_ = this.b.n("scrollStrategy");
            if ((ret_ === undefined && this.___rawMode)) {
                return undefined;
            }
            ret_ = enumGetBox(SelectScrollStrategy_$type, EnumUtil.getEnumValue(SelectScrollStrategy_$type, this.b.f(SelectScrollStrategy_$type, ret_)));
            return EnumUtil.getEnumValue(SelectScrollStrategy_$type, ret_);
        },
        set: function (a) {
            var value_ = enumGetBox(SelectScrollStrategy_$type, a);
            value_ = WCNativeHelper.camel(value_.toString());
            this.b.w("scrollStrategy", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Select.prototype, "n", {
        get: function () {
            var ret_ = this.b.m("items", function (a) { return new SelectItem(); });
            return ret_;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Select.prototype, "m", {
        get: function () {
            var ret_ = this.b.m("groups", function (a) { return new SelectGroup(); });
            return ret_;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Select.prototype, "p", {
        get: function () {
            var ret_ = this.b.m("selectedItem", function (a) { return new SelectItem(); });
            return ret_;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Select.prototype, "z", {
        get: function () {
            var ret_ = this.b.n("disabled");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.b.w("disabled", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Select.prototype, "ac", {
        get: function () {
            var ret_ = this.b.n("required");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.b.w("required", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Select.prototype, "af", {
        get: function () {
            var ret_ = this.b.n("name");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.b.w("name", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Select.prototype, "aa", {
        get: function () {
            var ret_ = this.b.n("invalid");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.b.w("invalid", value_);
        },
        enumerable: false,
        configurable: true
    });
    Select.prototype.changeChanged = function () {
        var _this = this;
        if (this.change != null) {
            this.b.s("change", function (a) { return _this.change(_this, SelectItemComponentEventArgs.b(a)); });
        }
        else {
            this.b.v("change");
        }
    };
    Select.prototype.focusChanged = function () {
        var _this = this;
        if (this.focus != null) {
            this.b.s("focus", function (a) { return _this.focus(_this, VoidEventArgs.b(a)); }, false);
        }
        else {
            this.b.v("focus");
        }
    };
    Select.prototype.blurChanged = function () {
        var _this = this;
        if (this.blur != null) {
            this.b.s("blur", function (a) { return _this.blur(_this, VoidEventArgs.b(a)); }, false);
        }
        else {
            this.b.v("blur");
        }
    };
    Select.prototype.openingChanged = function () {
        var _this = this;
        if (this.opening != null) {
            this.b.s("opening", function (a) { return _this.opening(_this, VoidEventArgs.b(a)); });
        }
        else {
            this.b.v("opening");
        }
    };
    Select.prototype.openedChanged = function () {
        var _this = this;
        if (this.opened != null) {
            this.b.s("opened", function (a) { return _this.opened(_this, VoidEventArgs.b(a)); });
        }
        else {
            this.b.v("opened");
        }
    };
    Select.prototype.closingChanged = function () {
        var _this = this;
        if (this.closing != null) {
            this.b.s("closing", function (a) { return _this.closing(_this, VoidEventArgs.b(a)); });
        }
        else {
            this.b.v("closing");
        }
    };
    Select.prototype.closedChanged = function () {
        var _this = this;
        if (this.closed != null) {
            this.b.s("closed", function (a) { return _this.closed(_this, VoidEventArgs.b(a)); });
        }
        else {
            this.b.v("closed");
        }
    };
    Select.prototype.aq = function (a) {
        this.nativeElement.focus(this.b.i(a));
    };
    Select.prototype.aj = function () {
        this.nativeElement.blur();
    };
    Select.prototype.at = function () {
        this.nativeElement.reportValidity();
    };
    Select.prototype.am = function () {
        this.nativeElement.clearSelection();
    };
    Select.prototype.al = function () {
        this.nativeElement.checkValidity();
    };
    Select.prototype.au = function (a) {
        this.nativeElement.setCustomValidity(this.b.i(a));
    };
    Select.$t = markType(Select, 'Select', BaseComboBoxLike.$);
    return Select;
}(BaseComboBoxLike));
export { Select };
