// igniteui-react/esm5/lib/Slider.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { SliderBase } from "./SliderBase";
import { markType } from "igniteui-react-core";
import { NumberEventArgs } from "./NumberEventArgs";
/**
 * @hidden
 */
var Slider = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(Slider, _super);
    function Slider() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.input = null;
        _this.change = null;
        return _this;
    }
    Object.defineProperty(Slider.prototype, "aa", {
        get: function () {
            var ret_ = this.e.n("value");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.e.w("value", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Slider.prototype, "ab", {
        get: function () {
            var ret_ = this.e.n("name");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.e.w("name", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Slider.prototype, "z", {
        get: function () {
            var ret_ = this.e.n("invalid");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.e.w("invalid", value_);
        },
        enumerable: false,
        configurable: true
    });
    Slider.prototype.inputChanged = function () {
        var _this = this;
        if (this.input != null) {
            this.e.s("input", function (a) { return _this.input(_this, NumberEventArgs.a(a)); });
        }
        else {
            this.e.v("input");
        }
    };
    Slider.prototype.changeChanged = function () {
        var _this = this;
        if (this.change != null) {
            this.e.s("change", function (a) { return _this.change(_this, NumberEventArgs.a(a)); });
        }
        else {
            this.e.v("change");
        }
    };
    Slider.prototype.ai = function (a) {
        if (a === void 0) {
            a = -1;
        }
        this.nativeElement.stepUp(this.e.i(a));
    };
    Slider.prototype.ah = function (a) {
        if (a === void 0) {
            a = -1;
        }
        this.nativeElement.stepDown(this.e.i(a));
    };
    Slider.prototype.af = function () {
        this.nativeElement.reportValidity();
    };
    Slider.prototype.ad = function () {
        this.nativeElement.checkValidity();
    };
    Slider.prototype.ag = function (a) {
        this.nativeElement.setCustomValidity(this.e.i(a));
    };
    Slider.$t = markType(Slider, 'Slider', SliderBase.$);
    return Slider;
}(SliderBase));
export { Slider };
