// igniteui-react/esm5/lib/StyleVariant.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
export var StyleVariant = /*@__PURE__*/ (function (StyleVariant) {
    StyleVariant[StyleVariant["Primary"] = 0] = "Primary";
    StyleVariant[StyleVariant["Info"] = 1] = "Info";
    StyleVariant[StyleVariant["Success"] = 2] = "Success";
    StyleVariant[StyleVariant["Warning"] = 3] = "Warning";
    StyleVariant[StyleVariant["Danger"] = 4] = "Danger";
    return StyleVariant;
})({});
/**
 * @hidden
 */
export var StyleVariant_$type = /*@__PURE__*/ markEnum('StyleVariant', 'Primary,0|Info,1|Success,2|Warning,3|Danger,4');
