// igniteui-react/esm5/lib/Tab.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { BaseElement } from "./BaseElement";
import { WCNativeHelper } from "./WCNativeHelper";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
var Tab = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(Tab, _super);
    function Tab() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.a = new WCNativeHelper();
        _this.e = null;
        return _this;
    }
    Object.defineProperty(Tab.prototype, "b", {
        get: function () {
            return this.a;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Tab.prototype, "nativeElement", {
        get: function () {
            return this.e;
        },
        set: function (a) {
            this.e = a;
            this.a.o = this.e;
        },
        enumerable: false,
        configurable: true
    });
    Tab.prototype.setNativeElement = function (a) {
        this.nativeElement = a;
    };
    Object.defineProperty(Tab.prototype, "g", {
        get: function () {
            var ret_ = this.b.n("panel");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.b.w("panel", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Tab.prototype, "d", {
        get: function () {
            var ret_ = this.b.n("selected");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.b.w("selected", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Tab.prototype, "c", {
        get: function () {
            var ret_ = this.b.n("disabled");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.b.w("disabled", value_);
        },
        enumerable: false,
        configurable: true
    });
    Tab.prototype.i = function () {
        this.nativeElement.connectedCallback();
    };
    Tab.prototype.j = function (a) {
        this.nativeElement.focus(this.b.i(a));
    };
    Tab.prototype.h = function () {
        this.nativeElement.blur();
    };
    Tab.$t = markType(Tab, 'Tab', BaseElement.$);
    return Tab;
}(BaseElement));
export { Tab };
