// igniteui-react/esm5/lib/Tabs.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { BaseElement } from "./BaseElement";
import { WCNativeHelper } from "./WCNativeHelper";
import { enumGetBox, EnumUtil, markType } from "igniteui-react-core";
import { TabsAlignment_$type } from "./TabsAlignment";
import { TabsActivation_$type } from "./TabsActivation";
import { TabComponentEventArgs } from "./TabComponentEventArgs";
/**
 * @hidden
 */
var Tabs = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(Tabs, _super);
    function Tabs() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.d = new WCNativeHelper();
        _this.f = null;
        _this.change = null;
        return _this;
    }
    Object.defineProperty(Tabs.prototype, "e", {
        get: function () {
            return this.d;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Tabs.prototype, "nativeElement", {
        get: function () {
            return this.f;
        },
        set: function (a) {
            this.f = a;
            this.d.o = this.f;
        },
        enumerable: false,
        configurable: true
    });
    Tabs.prototype.setNativeElement = function (a) {
        this.nativeElement = a;
    };
    Object.defineProperty(Tabs.prototype, "h", {
        get: function () {
            var ret_ = this.e.n("selected");
            return ret_;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Tabs.prototype, "c", {
        get: function () {
            var ret_ = this.e.n("alignment");
            if ((ret_ === undefined && this.___rawMode)) {
                return undefined;
            }
            ret_ = enumGetBox(TabsAlignment_$type, EnumUtil.getEnumValue(TabsAlignment_$type, this.e.f(TabsAlignment_$type, ret_)));
            return EnumUtil.getEnumValue(TabsAlignment_$type, ret_);
        },
        set: function (a) {
            var value_ = enumGetBox(TabsAlignment_$type, a);
            value_ = WCNativeHelper.camel(value_.toString());
            this.e.w("alignment", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Tabs.prototype, "b", {
        get: function () {
            var ret_ = this.e.n("activation");
            if ((ret_ === undefined && this.___rawMode)) {
                return undefined;
            }
            ret_ = enumGetBox(TabsActivation_$type, EnumUtil.getEnumValue(TabsActivation_$type, this.e.f(TabsActivation_$type, ret_)));
            return EnumUtil.getEnumValue(TabsActivation_$type, ret_);
        },
        set: function (a) {
            var value_ = enumGetBox(TabsActivation_$type, a);
            value_ = WCNativeHelper.camel(value_.toString());
            this.e.w("activation", value_);
        },
        enumerable: false,
        configurable: true
    });
    Tabs.prototype.changeChanged = function () {
        var _this = this;
        if (this.change != null) {
            this.e.s("change", function (a) { return _this.change(_this, TabComponentEventArgs.b(a)); });
        }
        else {
            this.e.v("change");
        }
    };
    Tabs.prototype.j = function () {
        this.nativeElement.disconnectedCallback();
    };
    Tabs.prototype.k = function (a) {
        this.nativeElement.select(this.e.i(a));
    };
    Tabs.$t = markType(Tabs, 'Tabs', BaseElement.$);
    return Tabs;
}(BaseElement));
export { Tabs };
