// igniteui-react/esm5/lib/TabsAlignment.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
export var TabsAlignment = /*@__PURE__*/ (function (TabsAlignment) {
    TabsAlignment[TabsAlignment["Start"] = 0] = "Start";
    TabsAlignment[TabsAlignment["End"] = 1] = "End";
    TabsAlignment[TabsAlignment["Center"] = 2] = "Center";
    TabsAlignment[TabsAlignment["Justify"] = 3] = "Justify";
    return TabsAlignment;
})({});
/**
 * @hidden
 */
export var TabsAlignment_$type = /*@__PURE__*/ markEnum('TabsAlignment', 'Start,0|End,1|Center,2|Justify,3');
