// igniteui-react/esm5/lib/TextareaAutocapitalize.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
export var TextareaAutocapitalize = /*@__PURE__*/ (function (TextareaAutocapitalize) {
    TextareaAutocapitalize[TextareaAutocapitalize["None"] = 0] = "None";
    TextareaAutocapitalize[TextareaAutocapitalize["Off"] = 1] = "Off";
    TextareaAutocapitalize[TextareaAutocapitalize["On"] = 2] = "On";
    TextareaAutocapitalize[TextareaAutocapitalize["Sentences"] = 3] = "Sentences";
    TextareaAutocapitalize[TextareaAutocapitalize["Words"] = 4] = "Words";
    TextareaAutocapitalize[TextareaAutocapitalize["Characters"] = 5] = "Characters";
    return TextareaAutocapitalize;
})({});
/**
 * @hidden
 */
export var TextareaAutocapitalize_$type = /*@__PURE__*/ markEnum('TextareaAutocapitalize', 'None,0|Off,1|On,2|Sentences,3|Words,4|Characters,5');
