// igniteui-react/esm5/lib/TreeItem.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { __extends } from "tslib";
import { BaseElement } from "./BaseElement";
import { WCNativeHelper } from "./WCNativeHelper";
import { markType } from "igniteui-react-core";
/**
 * @hidden
 */
var TreeItem = /** @class */ /*@__PURE__*/ (function (_super) {
    __extends(TreeItem, _super);
    function TreeItem() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.c = new WCNativeHelper();
        _this.k = null;
        return _this;
    }
    Object.defineProperty(TreeItem.prototype, "d", {
        get: function () {
            return this.c;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreeItem.prototype, "nativeElement", {
        get: function () {
            return this.k;
        },
        set: function (a) {
            this.k = a;
            this.c.o = this.k;
        },
        enumerable: false,
        configurable: true
    });
    TreeItem.prototype.setNativeElement = function (a) {
        this.nativeElement = a;
    };
    Object.defineProperty(TreeItem.prototype, "b", {
        get: function () {
            var ret_ = this.d.m("parent", function (a) { return new TreeItem(); });
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.d.w("parent", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreeItem.prototype, "j", {
        get: function () {
            var ret_ = this.d.n("level");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.d.w("level", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreeItem.prototype, "n", {
        get: function () {
            var ret_ = this.d.n("label");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.d.w("label", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreeItem.prototype, "g", {
        get: function () {
            var ret_ = this.d.n("expanded");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.d.w("expanded", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreeItem.prototype, "e", {
        get: function () {
            var ret_ = this.d.n("active");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.d.w("active", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreeItem.prototype, "f", {
        get: function () {
            var ret_ = this.d.n("disabled");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.d.w("disabled", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreeItem.prototype, "i", {
        get: function () {
            var ret_ = this.d.n("selected");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.d.w("selected", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreeItem.prototype, "h", {
        get: function () {
            var ret_ = this.d.n("loading");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.d.w("loading", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreeItem.prototype, "m", {
        get: function () {
            var ret_ = this.d.n("value");
            return ret_;
        },
        set: function (a) {
            var value_ = a;
            this.d.w("value", value_);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TreeItem.prototype, "a", {
        get: function () {
            var ret_ = this.d.m("path", function (a) { return new TreeItem(); });
            return ret_;
        },
        enumerable: false,
        configurable: true
    });
    TreeItem.prototype.q = function () {
        this.nativeElement.connectedCallback();
    };
    TreeItem.prototype.r = function () {
        this.nativeElement.disconnectedCallback();
    };
    TreeItem.prototype.t = function () {
        this.nativeElement.expandWithEvent();
    };
    TreeItem.prototype.p = function () {
        this.nativeElement.collapseWithEvent();
    };
    TreeItem.prototype.v = function () {
        this.nativeElement.toggle();
    };
    TreeItem.prototype.s = function () {
        this.nativeElement.expand();
    };
    TreeItem.prototype.o = function () {
        this.nativeElement.collapse();
    };
    TreeItem.$t = markType(TreeItem, 'TreeItem', BaseElement.$);
    return TreeItem;
}(BaseElement));
export { TreeItem };
