// igniteui-react/esm5/lib/UnpinnedLocation.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { markEnum } from "igniteui-react-core";
export var UnpinnedLocation = /*@__PURE__*/ (function (UnpinnedLocation) {
    UnpinnedLocation[UnpinnedLocation["Top"] = 0] = "Top";
    UnpinnedLocation[UnpinnedLocation["Bottom"] = 1] = "Bottom";
    UnpinnedLocation[UnpinnedLocation["Left"] = 2] = "Left";
    UnpinnedLocation[UnpinnedLocation["Right"] = 3] = "Right";
    return UnpinnedLocation;
})({});
/**
 * @hidden
 */
export var UnpinnedLocation_$type = /*@__PURE__*/ markEnum('UnpinnedLocation', 'Top:top,0|Bottom:bottom,1|Left:left,2|Right:right,3');
